// @replyke/comments-threaded-react-js/dist/cjs/components/MentionSuggestions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MentionSuggestions = MentionSuggestions;
const jsx_runtime_1 = require("react/jsx-runtime");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
function MentionSuggestions({ isMentionActive, isLoadingMentions, mentionSuggestions, handleMentionClick, }) {
    if (!isMentionActive)
        return null;
    return ((0, jsx_runtime_1.jsx)("div", { style: {
            position: "absolute",
            bottom: "100%",
            left: 0,
            right: 0,
            zIndex: 50,
            backgroundColor: "#FFFFFF",
            borderRadius: "12px",
            border: "1px solid #E5E7EB",
            boxShadow: "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)",
            maxHeight: "200px",
            overflowY: "auto",
            marginBottom: "8px",
        }, children: (0, jsx_runtime_1.jsx)("div", { style: { padding: "12px" }, children: isLoadingMentions ? ((0, jsx_runtime_1.jsx)("div", { style: { display: "flex", flexDirection: "column", gap: "8px" }, children: [1, 2, 3].map((i) => ((0, jsx_runtime_1.jsx)(ui_core_react_js_1.UserMentionSkeleton, {}, i))) })) : mentionSuggestions.length > 0 ? ((0, jsx_runtime_1.jsx)("div", { style: { display: "flex", flexDirection: "column", gap: "4px" }, children: mentionSuggestions.map((user) => ((0, jsx_runtime_1.jsxs)("div", { onClick: () => handleMentionClick(user), style: {
                        cursor: "pointer",
                        display: "flex",
                        alignItems: "center",
                        gap: "12px",
                        padding: "8px",
                        borderRadius: "8px",
                        transition: "background-color 150ms ease-in-out",
                    }, onMouseEnter: (e) => {
                        e.currentTarget.style.backgroundColor = "#F3F4F6";
                    }, onMouseLeave: (e) => {
                        e.currentTarget.style.backgroundColor = "transparent";
                    }, children: [(0, jsx_runtime_1.jsx)(ui_core_react_js_1.UserAvatar, { user: user, size: 32 }), (0, jsx_runtime_1.jsxs)("div", { style: {
                                display: "flex",
                                flexDirection: "column",
                                gap: "2px",
                            }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                        fontSize: "14px",
                                        fontWeight: "500",
                                        color: "#111827",
                                    }, children: ["@", user.username] }), user.name && ((0, jsx_runtime_1.jsx)("div", { style: {
                                        fontSize: "12px",
                                        fontWeight: "400",
                                        color: "#6B7280",
                                    }, children: user.name }))] })] }, user.id))) })) : ((0, jsx_runtime_1.jsx)("div", { style: {
                    padding: "16px",
                    textAlign: "center",
                    color: "#6B7280",
                    fontSize: "14px",
                }, children: "No users found" })) }) }));
}
