// @replyke/comments-threaded-react-js/dist/cjs/components/NewCommentForm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewCommentForm = NewCommentForm;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_js_1 = require("@replyke/react-js");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const comments_threaded_core_1 = require("@replyke/comments-threaded-core");
const MentionSuggestions_1 = require("./MentionSuggestions");
function NewCommentForm() {
    const { user } = (0, react_js_1.useUser)();
    const { project } = (0, react_js_1.useProject)();
    const giphyApiKey = project?.integrations.find((int) => int.name === "giphy")
        ?.data.apiKey;
    const textAreaRef = (0, react_1.useRef)(null);
    const [content, setContent] = (0, react_1.useState)("");
    const [isSubmitting, setIsSubmitting] = (0, react_1.useState)(false);
    const [isGiphyVisible, setIsGiphyVisible] = (0, react_1.useState)(false);
    const { createComment, callbacks } = (0, react_js_1.useCommentSection)();
    const { styleConfig } = (0, comments_threaded_core_1.useThreadedStyleConfig)();
    const { backgroundColor, itemsGap, verticalPadding, paddingLeft, paddingRight, placeholderText, textareaTextSize, textareaTextColor, textareaBackgroundColor, postButtonText, postButtonFontSize, postButtonFontWeight, postButtonFontColor, } = styleConfig.newCommentFormProps;
    const hasContent = content.trim().length > 0;
    const { cursorPosition, isSelectionActive } = (0, ui_core_react_js_1.useTextareaCursorIndicator)({
        textAreaRef,
    });
    const { isMentionActive, loading, mentionSuggestions, handleMentionClick, mentions, addMention, resetMentions, } = (0, react_js_1.useMentions)({
        content: textAreaRef.current?.value || "",
        setContent: (value) => {
            if (textAreaRef.current) {
                textAreaRef.current.value = value;
                setContent(value);
            }
        },
        focus: () => textAreaRef.current?.focus(),
        cursorPosition,
        isSelectionActive,
    });
    const handleSubmit = (0, react_1.useCallback)(async (e) => {
        e?.preventDefault();
        const textArea = textAreaRef.current;
        if (!textArea || !hasContent || isSubmitting)
            return;
        if (!user) {
            callbacks?.loginRequiredCallback();
            return;
        }
        const tempContent = textArea.value.trim();
        setIsSubmitting(true);
        try {
            await createComment?.({ content: tempContent, mentions });
            textArea.value = "";
            setContent("");
            resetMentions();
        }
        catch (error) {
            console.error("Error creating comment:", error);
        }
        finally {
            setIsSubmitting(false);
        }
    }, [hasContent, isSubmitting, user, createComment, mentions, resetMentions, callbacks]);
    const handleCreateGif = (0, react_1.useCallback)(async (gif) => {
        const textArea = textAreaRef.current;
        if (!textArea)
            throw new Error("Can not find textarea");
        textArea.value = "";
        setContent("");
        resetMentions();
        setIsGiphyVisible(false);
        try {
            await createComment({ gif, mentions });
        }
        catch (err) {
            (0, react_js_1.handleError)(err, "Creating comment failed: ");
        }
    }, [createComment, mentions, resetMentions]);
    // Add keyboard event handler for Enter key
    (0, react_1.useEffect)(() => {
        const handleKeyDown = async (event) => {
            if (event.key === "Enter" && !event.ctrlKey && !event.shiftKey) {
                event.preventDefault();
                handleSubmit();
            }
        };
        const textArea = textAreaRef.current;
        textArea?.addEventListener("keydown", handleKeyDown);
        return () => {
            textArea?.removeEventListener("keydown", handleKeyDown);
        };
    }, [handleSubmit]);
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [giphyApiKey ? ((0, jsx_runtime_1.jsx)(ui_core_react_js_1.GiphyContainer, { giphyApiKey: giphyApiKey, onClickBack: () => setIsGiphyVisible(false), onSelectGif: (selected) => handleCreateGif(selected), visible: isGiphyVisible })) : null, (0, jsx_runtime_1.jsxs)("form", { onSubmit: handleSubmit, style: { position: "relative" }, children: [(0, jsx_runtime_1.jsx)(MentionSuggestions_1.MentionSuggestions, { isMentionActive: isMentionActive, isLoadingMentions: loading, mentionSuggestions: mentionSuggestions, handleMentionClick: handleMentionClick }), (0, jsx_runtime_1.jsxs)("div", { style: {
                            display: "flex",
                            alignItems: "flex-end",
                            backgroundColor: backgroundColor,
                            borderRadius: "16px",
                            border: `1px solid ${hasContent ? "#BFDBFE" : "#E5E7EB"}`,
                            boxShadow: hasContent
                                ? "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)"
                                : "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
                            transition: "all 300ms ease-in-out",
                            padding: `${verticalPadding}px`,
                        }, onMouseEnter: (e) => {
                            if (!hasContent) {
                                e.currentTarget.style.boxShadow =
                                    "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)";
                            }
                        }, onMouseLeave: (e) => {
                            if (!hasContent) {
                                e.currentTarget.style.boxShadow =
                                    "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)";
                            }
                        }, children: [(0, jsx_runtime_1.jsx)("textarea", { ref: textAreaRef, onChange: (e) => setContent(e.target.value), placeholder: placeholderText, style: {
                                    flex: 1,
                                    padding: `${verticalPadding}px`,
                                    backgroundColor: textareaBackgroundColor,
                                    color: textareaTextColor,
                                    fontSize: `${textareaTextSize}px`,
                                    lineHeight: "1.625",
                                    outline: "none",
                                    resize: "none",
                                    border: "none",
                                }, rows: 2 }), !hasContent && giphyApiKey ? ((0, jsx_runtime_1.jsx)("button", { type: "button", onClick: () => setIsGiphyVisible(true), disabled: isSubmitting, style: {
                                    flexShrink: 0,
                                    padding: `${verticalPadding}px`,
                                    border: "none",
                                    outline: "none",
                                    fontWeight: postButtonFontWeight,
                                    fontSize: postButtonFontSize,
                                    color: postButtonFontColor,
                                    cursor: "pointer",
                                    backgroundColor: "transparent",
                                }, children: "GIF" })) : ((0, jsx_runtime_1.jsx)("button", { type: "submit", disabled: !hasContent || isSubmitting, style: {
                                    flexShrink: 0,
                                    padding: `${verticalPadding}px`,
                                    borderRadius: "50%",
                                    backgroundColor: hasContent && !isSubmitting ? "#2563EB" : "#E5E7EB",
                                    color: hasContent && !isSubmitting ? "#FFFFFF" : "#9CA3AF",
                                    boxShadow: "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
                                    transition: "all 200ms ease-in-out",
                                    border: "none",
                                    cursor: !hasContent || isSubmitting ? "not-allowed" : "pointer",
                                    outline: "none",
                                }, onMouseEnter: (e) => {
                                    if (hasContent && !isSubmitting) {
                                        e.currentTarget.style.backgroundColor = "#1D4ED8";
                                        e.currentTarget.style.boxShadow =
                                            "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)";
                                    }
                                }, onMouseLeave: (e) => {
                                    if (hasContent && !isSubmitting) {
                                        e.currentTarget.style.backgroundColor = "#2563EB";
                                        e.currentTarget.style.boxShadow =
                                            "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)";
                                    }
                                }, onFocus: (e) => {
                                    e.currentTarget.style.outline = "2px solid #3B82F6";
                                    e.currentTarget.style.outlineOffset = "2px";
                                }, onBlur: (e) => {
                                    e.currentTarget.style.outline = "none";
                                    e.currentTarget.style.outlineOffset = "0";
                                }, children: (0, jsx_runtime_1.jsx)("svg", { style: {
                                        height: `${textareaTextSize}px`,
                                        width: `${textareaTextSize}px`,
                                        transition: "transform 200ms ease-in-out",
                                        transform: hasContent ? "scale(1)" : "scale(1)",
                                    }, onMouseEnter: (e) => {
                                        if (hasContent) {
                                            e.currentTarget.style.transform = "scale(1.1)";
                                        }
                                    }, onMouseLeave: (e) => {
                                        if (hasContent) {
                                            e.currentTarget.style.transform = "scale(1)";
                                        }
                                    }, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", children: (0, jsx_runtime_1.jsx)("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M5 10l7-7m0 0l7 7m-7-7v18" }) }) }))] })] })] }));
}
