// @replyke/comments-threaded-react-js/dist/cjs/components/ThreadedCommentSection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importDefault(require("react"));
const comments_threaded_core_1 = require("@replyke/comments-threaded-core");
const useThreadedComments_1 = __importDefault(require("../hooks/useThreadedComments"));
const CommentsFeed_1 = require("./CommentsFeed");
const NewCommentForm_1 = require("./NewCommentForm");
const propComparison_1 = require("../utils/propComparison");
// Custom comparison function to prevent unnecessary re-renders
const arePropsEqual = (prevProps, nextProps) => {
    // Add development warnings for unnecessary prop changes
    (0, propComparison_1.warnPropChanges)("ThreadedCommentSection", prevProps, nextProps, [
        "entity",
        "callbacks",
        "styleConfig",
    ]);
    // Compare primitive values
    if (prevProps.entityId !== nextProps.entityId ||
        prevProps.foreignId !== nextProps.foreignId ||
        prevProps.shortId !== nextProps.shortId ||
        prevProps.isVisible !== nextProps.isVisible) {
        return false;
    }
    // Deep compare entity objects for more accurate comparison
    if (!(0, propComparison_1.deepEqual)(prevProps.entity, nextProps.entity)) {
        return false;
    }
    // Deep compare callbacks and styleConfig to handle cases where
    // parent component creates new objects with same content
    if (!(0, propComparison_1.deepEqual)(prevProps.callbacks, nextProps.callbacks)) {
        return false;
    }
    if (!(0, propComparison_1.deepEqual)(prevProps.styleConfig, nextProps.styleConfig)) {
        return false;
    }
    return true;
};
function ThreadedCommentSectionInner({ isVisible, children, }) {
    const { styleConfig: resolvedStyleConfig } = (0, comments_threaded_core_1.useThreadedStyleConfig)();
    const { backgroundColor } = resolvedStyleConfig.commentFeedProps;
    const { verticalPadding } = resolvedStyleConfig.newCommentFormProps;
    return ((0, jsx_runtime_1.jsxs)("div", { style: { height: "100%", display: "flex", flexDirection: "column" }, children: [(0, jsx_runtime_1.jsx)("div", { style: {
                    flex: 1,
                    overflowY: "auto",
                    backgroundColor: backgroundColor,
                    paddingTop: `${verticalPadding}px`,
                    paddingBottom: `${verticalPadding}px`,
                }, children: (0, jsx_runtime_1.jsx)(CommentsFeed_1.CommentsFeed, { children: children }) }), (0, jsx_runtime_1.jsx)("div", { style: {
                    borderTop: "1px solid #e5e7eb",
                    paddingTop: `${verticalPadding}px`,
                }, children: isVisible && (0, jsx_runtime_1.jsx)(NewCommentForm_1.NewCommentForm, {}) })] }));
}
function ThreadedCommentSection({ entity, entityId, foreignId, shortId, callbacks, styleConfig: styleConfigProp, isVisible = true, highlightedCommentId, children, }) {
    const styleConfig = (0, comments_threaded_core_1.useThreadedStyle)(styleConfigProp);
    const { CommentSectionProvider } = (0, useThreadedComments_1.default)({
        entity,
        entityId,
        foreignId,
        shortId,
        styleConfig,
        callbacks,
        highlightedCommentId,
    });
    return ((0, jsx_runtime_1.jsx)(CommentSectionProvider, { children: (0, jsx_runtime_1.jsx)(ThreadedCommentSectionInner, { isVisible: isVisible, children: children }) }));
}
exports.default = react_1.default.memo(ThreadedCommentSection, arePropsEqual);
