// @replyke/js/dist/index.mjs
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/core/client.ts
import axios from "axios";
var ReplykeHttpClient = class {
  instance;
  constructor({ projectId }) {
    this.instance = axios.create({
      baseURL: `https://api.replyke.com/api/v5/${projectId}`
    });
  }
};

// src/modules/users/index.ts
var users_exports = {};
__export(users_exports, {
  fetchUserByForeignId: () => fetchUserByForeignId,
  fetchUserById: () => fetchUserById
});

// src/modules/users/fetchUserById.ts
async function fetchUserById(client, data) {
  const path = `/users/${data.userId}`;
  const response = await client.instance.get(path);
  return response.data;
}

// src/modules/users/fetchUserByForeignId.ts
async function fetchUserByForeignId(client, data) {
  const path = `/users/by-foreign-id`;
  const params = {
    foreignId: data.foreignId,
    name: data.name,
    username: data.username,
    avatar: data.avatar,
    bio: data.bio,
    // birthdate: data.birthdate,
    // location: data.location,
    metadata: data.metadata ? JSON.stringify(data.metadata) : void 0,
    secureMetadata: data.secureMetadata ? JSON.stringify(data.secureMetadata) : void 0
  };
  const response = await client.instance.get(path, {
    params
  });
  return response.data;
}

// src/modules/entities/index.ts
var entities_exports = {};
__export(entities_exports, {
  createEntity: () => createEntity,
  deleteEntity: () => deleteEntity,
  fetchEntity: () => fetchEntity,
  fetchEntityByForeignId: () => fetchEntityByForeignId,
  fetchEntityByShortId: () => fetchEntityByShortId,
  fetchManyEntities: () => fetchManyEntities,
  updateEntity: () => updateEntity
});

// src/modules/entities/createEntity.ts
async function createEntity(client, data) {
  const path = `/entities`;
  const response = await client.instance.post(path, data);
  return response.data;
}

// src/modules/entities/fetchEntity.ts
async function fetchEntity(client, data) {
  const path = `/entities/${data.entityId}`;
  const response = await client.instance.get(path);
  return response.data;
}

// src/modules/entities/fetchEntityByForeignId.ts
async function fetchEntityByForeignId(client, data) {
  const path = `/entities/by-foreign-id`;
  const response = await client.instance.get(path, { params: data });
  return response.data;
}

// src/modules/entities/fetchEntityByShortId.ts
async function fetchEntityByShortId(client, data) {
  const path = `/entities/by-short-id`;
  const response = await client.instance.get(path, { params: data });
  return response.data;
}

// src/modules/entities/fetchManyEntities.ts
async function fetchManyEntities(client, data) {
  const path = `/entities`;
  const response = await client.instance.get(path, { params: data });
  return response.data;
}

// src/modules/entities/updateEntity.ts
async function updateEntity(client, data) {
  const { entityId, ...restOfProps } = data;
  const path = `/entities/${data.entityId}`;
  const response = await client.instance.patch(path, restOfProps);
  return response.data;
}

// src/modules/entities/deleteEntity.ts
async function deleteEntity(client, data) {
  const path = `/entities/${data.entityId}`;
  const response = await client.instance.delete(path);
  return response.data;
}

// src/modules/comments/index.ts
var comments_exports = {};
__export(comments_exports, {
  fetchComment: () => fetchComment,
  fetchCommentByForeignId: () => fetchCommentByForeignId
});

// src/modules/comments/fetchComment.ts
async function fetchComment(client, data) {
  const path = `/comments/${data.commentId}`;
  const response = await client.instance.get(path);
  return response.data;
}

// src/modules/comments/fetchCommentByForeignId.ts
async function fetchCommentByForeignId(client, data) {
  const path = `/comments/by-foreign-id`;
  const response = await client.instance.get(path, { params: data });
  return response.data;
}

// src/index.ts
var ReplykeClient = class _ReplykeClient {
  http;
  users;
  entities;
  comments;
  constructor(http) {
    this.http = http;
    this.users = bindModule(users_exports, this.http);
    this.entities = bindModule(entities_exports, this.http);
    this.comments = bindModule(comments_exports, this.http);
  }
  static async init(config) {
    const http = new ReplykeHttpClient(config);
    return new _ReplykeClient(http);
  }
};
function bindModule(module, client) {
  const bound = {};
  for (const key in module) {
    bound[key] = (...args) => module[key](client, ...args);
  }
  return bound;
}
export {
  ReplykeClient
};
