// @replyke/comments-threaded-core/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.threadedBaseStyle = exports.useThreadedStyleConfig = exports.useThreadedStyle = exports.ThreadedStyleConfigProvider = exports.mergeThreadedStyleData = exports.isThreadedStyleConfig = void 0;
// Threaded section - helpers
var isThreadedStyleConfig_1 = require("./helpers/isThreadedStyleConfig");
Object.defineProperty(exports, "isThreadedStyleConfig", { enumerable: true, get: function () { return isThreadedStyleConfig_1.isThreadedStyleConfig; } });
var mergeThreadedStyleData_1 = require("./helpers/mergeThreadedStyleData");
Object.defineProperty(exports, "mergeThreadedStyleData", { enumerable: true, get: function () { return mergeThreadedStyleData_1.mergeThreadedStyleData; } });
// Threaded section -context
var threaded_style_config_context_1 = require("./context/threaded-style-config-context");
Object.defineProperty(exports, "ThreadedStyleConfigProvider", { enumerable: true, get: function () { return threaded_style_config_context_1.ThreadedStyleConfigProvider; } });
// Threaded section - hooks
var useThreadedStyle_1 = require("./hooks/useThreadedStyle");
Object.defineProperty(exports, "useThreadedStyle", { enumerable: true, get: function () { return __importDefault(useThreadedStyle_1).default; } });
var useThreadedStyleConfig_1 = require("./hooks/useThreadedStyleConfig");
Object.defineProperty(exports, "useThreadedStyleConfig", { enumerable: true, get: function () { return __importDefault(useThreadedStyleConfig_1).default; } });
// Threaded section - other
var threaded_base_style_1 = require("./threaded-base-style");
Object.defineProperty(exports, "threadedBaseStyle", { enumerable: true, get: function () { return threaded_base_style_1.threadedBaseStyle; } });
