// @replyke/comments-thread-react-js/dist/cjs/components/CommentsFeed/CommentThread/CommentReplies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
function CommentReplies({ depth, isCollapsed, loading, visibleReplies, hiddenRepliesCount, showAllReplies, onShowAllReplies, onDeleteComment, onReportComment, CommentThreadComponent, }) {
    // Don't render anything if collapsed or no replies
    if (isCollapsed || visibleReplies.length === 0) {
        return null;
    }
    return ((0, jsx_runtime_1.jsxs)("div", { style: { marginTop: "8px" }, children: [loading && ((0, jsx_runtime_1.jsx)("div", { style: { fontSize: "12px", color: "#6B7280", marginBottom: "8px" }, children: "Loading replies..." })), visibleReplies.map((reply, index) => ((0, jsx_runtime_1.jsx)(CommentThreadComponent, { comment: reply, depth: depth + 1, isLastReply: index === visibleReplies.length - 1, onDeleteComment: onDeleteComment, onReportComment: onReportComment }, reply.id))), hiddenRepliesCount > 0 && !showAllReplies && ((0, jsx_runtime_1.jsx)("div", { style: { marginTop: "12px", marginLeft: "4px" }, children: (0, jsx_runtime_1.jsxs)("button", { onClick: onShowAllReplies, style: {
                        display: "flex",
                        alignItems: "center",
                        gap: "8px",
                        fontSize: "12px",
                        color: "#2563EB",
                        fontWeight: "500",
                        transition: "all 150ms ease-in-out",
                        padding: "4px 8px",
                        borderRadius: "9999px",
                        background: "transparent",
                        border: "none",
                        cursor: "pointer",
                    }, onMouseEnter: (e) => {
                        e.currentTarget.style.color = "#1D4ED8";
                        e.currentTarget.style.backgroundColor = "#EFF6FF";
                    }, onMouseLeave: (e) => {
                        e.currentTarget.style.color = "#2563EB";
                        e.currentTarget.style.backgroundColor = "transparent";
                    }, children: [(0, jsx_runtime_1.jsx)("svg", { style: { width: "12px", height: "12px" }, fill: "currentColor", viewBox: "0 0 20 20", children: (0, jsx_runtime_1.jsx)("path", { fillRule: "evenodd", d: "M10 3a1 1 0 011 1v5h5a1 1 0 110 2h-5v5a1 1 0 11-2 0v-5H4a1 1 0 110-2h5V4a1 1 0 011-1z", clipRule: "evenodd" }) }), hiddenRepliesCount, " more", " ", hiddenRepliesCount === 1 ? "reply" : "replies"] }) }))] }));
}
exports.default = CommentReplies;
