// @replyke/comments-thread-react-js/dist/cjs/components/CommentsFeed/CommentThread/SingleComment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_js_1 = require("@replyke/react-js");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const VoteButtons_1 = __importDefault(require("./VoteButtons"));
const ActionMenu_1 = __importDefault(require("./ActionMenu"));
const NewReplyForm_1 = __importDefault(require("./NewReplyForm"));
function SingleComment({ comment: commentFromSection, depth, hasReplies, isCollapsed, replyCount, isLastReply = false, onToggleCollapse, }) {
    const [comment, setComment] = (0, react_1.useState)(commentFromSection);
    const { deleteComment } = (0, react_js_1.useCommentSection)();
    const [showReplyForm, setShowReplyForm] = (0, react_1.useState)(false);
    const maxDepth = 6; // Limit visual nesting depth
    const actualDepth = Math.min(depth, maxDepth);
    // Calculate progressive indentation using inline styles for reliability
    const indentationPx = actualDepth * 24; // 40px per level (same as ml-10 = 2.5rem = 40px)
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            position: "relative",
            marginBottom: "8px",
            marginLeft: `${indentationPx}px`,
        }, children: [actualDepth > 0 && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [!isLastReply && ((0, jsx_runtime_1.jsx)("div", { style: {
                            position: "absolute",
                            width: "1px",
                            backgroundColor: "#D1D5DB",
                            zIndex: 0,
                            left: "-12px",
                            top: "0px",
                            height: "100%",
                        } })), (0, jsx_runtime_1.jsx)("div", { style: {
                            position: "absolute",
                            top: "-8px",
                            width: "1px",
                            height: "18px",
                            backgroundColor: "#D1D5DB",
                            zIndex: 0,
                            left: "-12px",
                        } }), (0, jsx_runtime_1.jsx)("div", { style: {
                            position: "absolute",
                            top: "10px",
                            width: "12px",
                            height: "16px",
                            borderLeft: "1.5px solid #D1D5DB",
                            borderBottom: "1.5px solid #D1D5DB",
                            borderTop: "1.5px solid transparent",
                            borderRight: "1.5px solid transparent",
                            borderBottomLeftRadius: "12px",
                            zIndex: 0,
                            left: "-12px",
                        } })] })), (0, jsx_runtime_1.jsx)("div", { style: {
                    padding: "8px 0",
                    borderRadius: "6px",
                    transition: "colors 150ms ease-in-out",
                }, children: (0, jsx_runtime_1.jsxs)("div", { style: { display: "flex" }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                flexShrink: 0,
                                marginRight: "12px",
                                position: "relative",
                                marginTop: "4px",
                            }, children: [(0, jsx_runtime_1.jsx)("div", { style: { position: "relative", zIndex: 10 }, children: (0, jsx_runtime_1.jsx)(ui_core_react_js_1.UserAvatar, { user: comment.user, borderRadius: 24, size: 24 }) }), hasReplies && !isCollapsed && ((0, jsx_runtime_1.jsx)("div", { style: {
                                        position: "absolute",
                                        left: "50%",
                                        top: "20px",
                                        width: "1px",
                                        backgroundColor: "#D1D5DB",
                                        zIndex: 0,
                                        height: "calc(100% + 10px)",
                                    } }))] }), (0, jsx_runtime_1.jsxs)("div", { style: { flex: 1, minWidth: 0 }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                        display: "flex",
                                        alignItems: "center",
                                        justifyContent: "space-between",
                                        marginBottom: "4px",
                                    }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                gap: "8px",
                                                fontSize: "12px",
                                                color: "#6B7280",
                                            }, children: [(0, jsx_runtime_1.jsx)("span", { style: { fontWeight: "500", color: "#374151" }, children: comment.user?.name || "Anonymous" }), (0, jsx_runtime_1.jsx)("span", { children: "\u2022" }), (0, jsx_runtime_1.jsx)("span", { children: new Date(comment.createdAt).toLocaleDateString() }), hasReplies && ((0, jsx_runtime_1.jsx)("button", { onClick: onToggleCollapse, style: {
                                                        marginLeft: "4px",
                                                        width: "16px",
                                                        height: "16px",
                                                        display: "flex",
                                                        alignItems: "center",
                                                        justifyContent: "center",
                                                        color: "#6B7280",
                                                        backgroundColor: "#F3F4F6",
                                                        borderRadius: "2px",
                                                        transition: "all 150ms ease-in-out",
                                                        fontSize: "14px",
                                                        fontWeight: "bold",
                                                        border: isCollapsed ? "1px solid #D1D5DB" : "none",
                                                        cursor: "pointer",
                                                    }, onMouseEnter: (e) => {
                                                        e.currentTarget.style.color = "#374151";
                                                        e.currentTarget.style.backgroundColor = "#E5E7EB";
                                                    }, onMouseLeave: (e) => {
                                                        e.currentTarget.style.color = "#6B7280";
                                                        e.currentTarget.style.backgroundColor = "#F3F4F6";
                                                    }, title: isCollapsed ? "Expand thread" : "Collapse thread", children: isCollapsed ? "+" : "−" }))] }), (0, jsx_runtime_1.jsx)(ActionMenu_1.default, { comment: comment, onDelete: () => {
                                                deleteComment?.({ commentId: comment.id });
                                            } })] }), !isCollapsed && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)("p", { style: {
                                                fontSize: "12px",
                                                color: "#1F2937",
                                                marginBottom: "12px",
                                                lineHeight: "1.625",
                                            }, children: comment.content }), (0, jsx_runtime_1.jsxs)("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                justifyContent: "space-between",
                                            }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                                        display: "flex",
                                                        alignItems: "center",
                                                        gap: "16px",
                                                        fontSize: "12px",
                                                    }, children: [(0, jsx_runtime_1.jsx)("button", { onClick: () => setShowReplyForm((prev) => !prev), style: {
                                                                color: "#6B7280",
                                                                fontWeight: "500",
                                                                padding: "4px 8px",
                                                                borderRadius: "4px",
                                                                marginLeft: "-8px",
                                                                transition: "all 150ms ease-in-out",
                                                                background: "transparent",
                                                                border: "none",
                                                                cursor: "pointer",
                                                            }, onMouseEnter: (e) => {
                                                                e.currentTarget.style.color = "#2563EB";
                                                                e.currentTarget.style.backgroundColor = "#EFF6FF";
                                                            }, onMouseLeave: (e) => {
                                                                e.currentTarget.style.color = "#6B7280";
                                                                e.currentTarget.style.backgroundColor = "transparent";
                                                            }, children: "Reply" }), hasReplies && ((0, jsx_runtime_1.jsxs)("span", { style: { color: "#9CA3AF" }, children: [replyCount, " ", replyCount === 1 ? "reply" : "replies"] }))] }), (0, jsx_runtime_1.jsx)("div", { style: { flexShrink: 0 }, children: (0, jsx_runtime_1.jsx)(VoteButtons_1.default, { comment: comment, setComment: setComment, size: "small" }) })] }), showReplyForm && ((0, jsx_runtime_1.jsx)(NewReplyForm_1.default, { comment: comment, setShowReplyForm: setShowReplyForm }))] }))] })] }) })] }));
}
exports.default = SingleComment;
