// @replyke/comments-thread-react-js/dist/cjs/components/modals/CommentMenuModalOwner/CommentMenuModalOwner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_js_1 = require("@replyke/react-js");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const useModalManager_1 = __importDefault(require("../../../hooks/useModalManager"));
function CommentMenuModalOwner() {
    const { optionsComment, isCommentOptionsModalOwnerOpen, closeCommentOptionsModalOwner, } = (0, useModalManager_1.default)();
    const { deleteComment } = (0, react_js_1.useCommentSection)();
    const handleDeleteComment = async () => {
        const userConfirmed = window.confirm("Are you sure you want to delete this comment?");
        if (userConfirmed && optionsComment) {
            closeCommentOptionsModalOwner?.();
            await deleteComment?.({ commentId: optionsComment.id });
        }
        else {
            closeCommentOptionsModalOwner?.();
        }
    };
    return ((0, jsx_runtime_1.jsx)(ui_core_react_js_1.Modal, { show: !!isCommentOptionsModalOwnerOpen, onClose: closeCommentOptionsModalOwner, children: (0, jsx_runtime_1.jsx)("div", { onClick: (e) => e.stopPropagation(), style: {
                ...ui_core_react_js_1.resetDiv,
                backgroundColor: "white",
                borderRadius: 8,
                width: "100%",
                maxWidth: 240,
                alignSelf: "center",
                paddingTop: 8,
                paddingBottom: 8,
            }, children: (0, jsx_runtime_1.jsxs)("ul", { style: {
                    ...ui_core_react_js_1.resetUl,
                    width: "100%",
                }, children: [(0, jsx_runtime_1.jsx)("li", { style: {
                            display: "flex",
                            justifyContent: "center",
                            justifySelf: "center",
                        }, children: (0, jsx_runtime_1.jsx)("button", { style: {
                                ...ui_core_react_js_1.resetButton,
                                fontWeight: 600,
                                color: "#DC2626",
                                paddingLeft: 24,
                                paddingRight: 24,
                                paddingTop: 8,
                                paddingBottom: 8,
                            }, onClick: handleDeleteComment, children: "Remove" }) }), (0, jsx_runtime_1.jsx)("div", { style: { height: 1, width: "100%", backgroundColor: "#e7e7e7" } }), (0, jsx_runtime_1.jsx)("li", { style: {
                            display: "flex",
                            justifyContent: "center",
                            justifySelf: "center",
                        }, children: (0, jsx_runtime_1.jsx)("button", { style: {
                                ...ui_core_react_js_1.resetButton,
                                paddingLeft: 24,
                                paddingRight: 24,
                                paddingTop: 8,
                                paddingBottom: 8,
                            }, onClick: closeCommentOptionsModalOwner, children: "Cancel" }) })] }) }) }));
}
exports.default = CommentMenuModalOwner;
