// @replyke/core/dist/cjs/hooks/entities/useEntityVotes.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var handleError_1 = require("../../utils/handleError");
var user_1 = require("../user");
// Import the API communication hooks
var useUpvoteEntity_1 = __importDefault(require("./useUpvoteEntity"));
var useRemoveEntityUpvote_1 = __importDefault(require("./useRemoveEntityUpvote"));
var useDownvoteEntity_1 = __importDefault(require("./useDownvoteEntity"));
var useRemoveEntityDownvote_1 = __importDefault(require("./useRemoveEntityDownvote"));
function useEntityVotes(props) {
    var _this = this;
    var user = (0, user_1.useUser)().user;
    var entity = props.entity, setEntity = props.setEntity;
    var _a = (0, react_1.useState)(false), isUpdating = _a[0], setIsUpdating = _a[1];
    var _b = (0, react_1.useState)(entity), previousEntity = _b[0], setPreviousEntity = _b[1];
    // Save the current state for potential reversion on error
    var updateEntityState = function (newEntityState) {
        setPreviousEntity(entity);
        setEntity(newEntityState);
    };
    var revertToPreviousState = (0, react_1.useCallback)(function () {
        setEntity(previousEntity);
    }, [previousEntity, setEntity]);
    // Get the API request functions from the extracted hooks
    var upvoteEntityRequest = (0, useUpvoteEntity_1.default)();
    var removeEntityUpvoteRequest = (0, useRemoveEntityUpvote_1.default)();
    var downvoteEntityRequest = (0, useDownvoteEntity_1.default)();
    var removeEntityDownvoteRequest = (0, useRemoveEntityDownvote_1.default)();
    var handleUpvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedEntity, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (!entity)
                        throw new Error("No entity provided");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: add user's upvote and remove any downvote from the user
                    updateEntityState(__assign(__assign({}, entity), { upvotes: __spreadArray(__spreadArray([], (entity.upvotes || []), true), [user.id], false), downvotes: (entity.downvotes || []).filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, upvoteEntityRequest({
                            entityId: entity.id,
                        })];
                case 2:
                    updatedEntity = _a.sent();
                    setEntity(updatedEntity);
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_1, "Failed to update entity vote: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleRemoveUpvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedEntity, err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (!entity)
                        throw new Error("No entity provided");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: remove user's upvote
                    updateEntityState(__assign(__assign({}, entity), { upvotes: (entity.upvotes || []).filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, removeEntityUpvoteRequest({
                            entityId: entity.id,
                        })];
                case 2:
                    updatedEntity = _a.sent();
                    setEntity(updatedEntity);
                    return [3 /*break*/, 5];
                case 3:
                    err_2 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_2, "Failed to remove upvote from entity: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleDownvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedEntity, err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (!entity)
                        throw new Error("No entity provided");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: add user's downvote and remove any upvote from the user
                    updateEntityState(__assign(__assign({}, entity), { downvotes: __spreadArray(__spreadArray([], (entity.downvotes || []), true), [user.id], false), upvotes: (entity.upvotes || []).filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, downvoteEntityRequest({
                            entityId: entity.id,
                        })];
                case 2:
                    updatedEntity = _a.sent();
                    setEntity(updatedEntity);
                    return [3 /*break*/, 5];
                case 3:
                    err_3 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_3, "Failed to update entity vote: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleRemoveDownvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedEntity, err_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (!entity)
                        throw new Error("No entity provided");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: remove user's downvote
                    updateEntityState(__assign(__assign({}, entity), { downvotes: (entity.downvotes || []).filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, removeEntityDownvoteRequest({
                            entityId: entity.id,
                        })];
                case 2:
                    updatedEntity = _a.sent();
                    setEntity(updatedEntity);
                    return [3 /*break*/, 5];
                case 3:
                    err_4 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_4, "Failed to remove downvote from entity: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    return {
        upvoteEntity: handleUpvote,
        removeEntityUpvote: handleRemoveUpvote,
        downvoteEntity: handleDownvote,
        removeEntityDownvote: handleRemoveDownvote,
    };
}
exports.default = useEntityVotes;
