// @replyke/ui-core-react-js/dist/cjs/constants/reset-styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetIcon = exports.resetImg = exports.resetTextInput = exports.resetP = exports.resetIl = exports.resetUl = exports.resetButton = exports.resetDiv = void 0;
exports.resetDiv = {
    padding: 0,
    margin: 0,
};
exports.resetButton = {
    margin: 0,
    padding: 0,
    borderStyle: "none",
    borderWidth: 0,
    backgroundColor: "transparent",
    outline: "none",
    fontFamily: "inherit",
    cursor: "pointer",
};
exports.resetUl = {
    listStyleType: "none",
    padding: 0,
    margin: 0,
    fontFamily: "inherit",
};
exports.resetIl = {
    padding: 0,
    margin: 0,
    fontFamily: "inherit",
};
exports.resetP = {
    padding: 0,
    margin: 0,
    fontFamily: "inherit",
    cursor: "default",
};
exports.resetTextInput = {
    padding: 0,
    margin: 0,
    fontFamily: "inherit",
    resize: "none",
    border: "none",
    outline: "none",
    overflow: "auto",
    boxShadow: "none",
    WebkitBoxShadow: "none",
};
exports.resetImg = {
    padding: 0,
    margin: 0,
};
exports.resetIcon = {
    padding: 0,
    margin: 0,
};
