// @replyke/comments-threaded-react-js/dist/esm/components/MentionSuggestions.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { UserAvatar, UserMentionSkeleton } from "@replyke/ui-core-react-js";
function MentionSuggestions({ isMentionActive, isLoadingMentions, mentionSuggestions, handleMentionClick, }) {
    if (!isMentionActive)
        return null;
    return (_jsx("div", { style: {
            position: "absolute",
            bottom: "100%",
            left: 0,
            right: 0,
            zIndex: 50,
            backgroundColor: "#FFFFFF",
            borderRadius: "12px",
            border: "1px solid #E5E7EB",
            boxShadow: "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)",
            maxHeight: "200px",
            overflowY: "auto",
            marginBottom: "8px",
        }, children: _jsx("div", { style: { padding: "12px" }, children: isLoadingMentions ? (_jsx("div", { style: { display: "flex", flexDirection: "column", gap: "8px" }, children: [1, 2, 3].map((i) => (_jsx(UserMentionSkeleton, {}, i))) })) : mentionSuggestions.length > 0 ? (_jsx("div", { style: { display: "flex", flexDirection: "column", gap: "4px" }, children: mentionSuggestions.map((user) => (_jsxs("div", { onClick: () => handleMentionClick(user), style: {
                        cursor: "pointer",
                        display: "flex",
                        alignItems: "center",
                        gap: "12px",
                        padding: "8px",
                        borderRadius: "8px",
                        transition: "background-color 150ms ease-in-out",
                    }, onMouseEnter: (e) => {
                        e.currentTarget.style.backgroundColor = "#F3F4F6";
                    }, onMouseLeave: (e) => {
                        e.currentTarget.style.backgroundColor = "transparent";
                    }, children: [_jsx(UserAvatar, { user: user, size: 32 }), _jsxs("div", { style: {
                                display: "flex",
                                flexDirection: "column",
                                gap: "2px",
                            }, children: [_jsxs("div", { style: {
                                        fontSize: "14px",
                                        fontWeight: "500",
                                        color: "#111827",
                                    }, children: ["@", user.username] }), user.name && (_jsx("div", { style: {
                                        fontSize: "12px",
                                        fontWeight: "400",
                                        color: "#6B7280",
                                    }, children: user.name }))] })] }, user.id))) })) : (_jsx("div", { style: {
                    padding: "16px",
                    textAlign: "center",
                    color: "#6B7280",
                    fontSize: "14px",
                }, children: "No users found" })) }) }));
}
export { MentionSuggestions };
