// @replyke/comments-threaded-react-js/dist/esm/components/ThreadedCommentSection.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from "react";
import { useThreadedStyle, useThreadedStyleConfig, } from "@replyke/comments-threaded-core";
import useThreadedComments from "../hooks/useThreadedComments";
import { CommentsFeed } from "./CommentsFeed";
import { NewCommentForm } from "./NewCommentForm";
import { deepEqual, warnPropChanges } from "../utils/propComparison";
// Custom comparison function to prevent unnecessary re-renders
const arePropsEqual = (prevProps, nextProps) => {
    // Add development warnings for unnecessary prop changes
    warnPropChanges("ThreadedCommentSection", prevProps, nextProps, [
        "entity",
        "callbacks",
        "styleConfig",
    ]);
    // Compare primitive values
    if (prevProps.entityId !== nextProps.entityId ||
        prevProps.foreignId !== nextProps.foreignId ||
        prevProps.shortId !== nextProps.shortId ||
        prevProps.isVisible !== nextProps.isVisible) {
        return false;
    }
    // Deep compare entity objects for more accurate comparison
    if (!deepEqual(prevProps.entity, nextProps.entity)) {
        return false;
    }
    // Deep compare callbacks and styleConfig to handle cases where
    // parent component creates new objects with same content
    if (!deepEqual(prevProps.callbacks, nextProps.callbacks)) {
        return false;
    }
    if (!deepEqual(prevProps.styleConfig, nextProps.styleConfig)) {
        return false;
    }
    return true;
};
function ThreadedCommentSectionInner({ isVisible, children, }) {
    const { styleConfig: resolvedStyleConfig } = useThreadedStyleConfig();
    const { backgroundColor } = resolvedStyleConfig.commentFeedProps;
    const { verticalPadding } = resolvedStyleConfig.newCommentFormProps;
    return (_jsxs("div", { style: { height: "100%", display: "flex", flexDirection: "column" }, children: [_jsx("div", { style: {
                    flex: 1,
                    overflowY: "auto",
                    backgroundColor: backgroundColor,
                    paddingTop: `${verticalPadding}px`,
                    paddingBottom: `${verticalPadding}px`,
                }, children: _jsx(CommentsFeed, { children: children }) }), _jsx("div", { style: {
                    borderTop: "1px solid #e5e7eb",
                    paddingTop: `${verticalPadding}px`,
                }, children: isVisible && _jsx(NewCommentForm, {}) })] }));
}
function ThreadedCommentSection({ entity, entityId, foreignId, shortId, callbacks, styleConfig: styleConfigProp, isVisible = true, highlightedCommentId, children, }) {
    const styleConfig = useThreadedStyle(styleConfigProp);
    const { CommentSectionProvider } = useThreadedComments({
        entity,
        entityId,
        foreignId,
        shortId,
        styleConfig,
        callbacks,
        highlightedCommentId,
    });
    return (_jsx(CommentSectionProvider, { children: _jsx(ThreadedCommentSectionInner, { isVisible: isVisible, children: children }) }));
}
export default React.memo(ThreadedCommentSection, arePropsEqual);
