// @replyke/comments-threaded-react-js/dist/esm/components/modals/CommentMenuModal/MainContent.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { resetButton, resetUl } from "@replyke/ui-core-react-js";
import useModalManager from "../../../hooks/useModalManager";
function MainContent({ clickReport }) {
    const { closeCommentOptionsModal } = useModalManager();
    return (_jsxs("ul", { style: {
            ...resetUl,
            width: "100%",
        }, children: [_jsx("li", { style: {
                    display: "flex",
                    justifyContent: "center",
                    justifySelf: "center",
                }, children: _jsx("button", { style: {
                        ...resetButton,
                        fontWeight: 600,
                        color: "#DC2626",
                        paddingLeft: 24,
                        paddingRight: 24,
                        paddingTop: 8,
                        paddingBottom: 8,
                    }, onClick: clickReport, children: "Report" }) }), _jsx("div", { style: { height: 1, width: "100%", backgroundColor: "#e7e7e7" } }), _jsx("li", { style: {
                    display: "flex",
                    justifyContent: "center",
                    justifySelf: "center",
                }, children: _jsx("button", { style: {
                        ...resetButton,
                        paddingLeft: 24,
                        paddingRight: 24,
                        paddingTop: 8,
                        paddingBottom: 8,
                    }, onClick: closeCommentOptionsModal, children: "Cancel" }) })] }));
}
export default MainContent;
