// @giphy/react-components/dist/components/grid.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/components/grid.tsx
var grid_exports = {};
__export(grid_exports, {
  default: () => grid_default
});
module.exports = __toCommonJS(grid_exports);
var import_js_fetch_api = require("@giphy/js-fetch-api");
var import_js_util3 = require("@giphy/js-util");
var import_react12 = __toESM(require("react"));
var import_styled_components8 = __toESM(require("styled-components"));
var import_throttle_debounce = require("throttle-debounce");

// src/util/observer.tsx
var import_intersection_observer = require("intersection-observer");
var import_react = __toESM(require("react"));
var Observer = ({ children, className, onVisibleChange, config }) => {
  const container = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    let io;
    if (container.current) {
      io = new IntersectionObserver(([entry]) => {
        if (onVisibleChange)
          onVisibleChange(entry.isIntersecting);
      }, config);
      io.observe(container.current);
    }
    return () => io == null ? void 0 : io.disconnect();
  }, [onVisibleChange, container, config]);
  return /* @__PURE__ */ import_react.default.createElement("div", { ref: container, className }, children);
};
var observer_default = Observer;

// src/components/fetch-error.tsx
var import_colors = require("@giphy/colors");
var import_react2 = __toESM(require("react"));
var import_styled_components = __toESM(require("styled-components"));
var Message = import_styled_components.default.div`
    color: ${import_colors.giphyLightGrey};
    display: flex;
    justify-content: center;
    margin: 30px 0;
    font-family: interface, Helvetica Neue, helvetica, sans-serif;
    font-size: 16px;
    font-weight: 600;
    a {
        color: ${import_colors.giphyBlue};
        cursor: pointer;
        &:hover {
            color: white;
        }
    }
`;
var FetchError = ({ onClick }) => /* @__PURE__ */ import_react2.default.createElement(Message, null, "Error loading GIFs.\xA0", /* @__PURE__ */ import_react2.default.createElement("a", { onClick }, "Try again?"));
var fetch_error_default = FetchError;

// src/components/gif.tsx
var import_colors3 = require("@giphy/colors");
var import_js_util2 = require("@giphy/js-util");
var import_intersection_observer2 = require("intersection-observer");
var import_react9 = __toESM(require("react"));
var import_styled_components6 = __toESM(require("styled-components"));

// src/util/pingback.ts
var import_js_analytics = require("@giphy/js-analytics");
var import_js_util = require("@giphy/js-util");
var firePingback = (actionType) => (gif, userId, target, attributes = {}) => {
  if (!gif.analytics_response_payload) {
    return;
  }
  (0, import_js_analytics.pingback)({
    analyticsResponsePayload: gif.analytics_response_payload,
    userId,
    actionType,
    attributes: __spreadValues({ position: JSON.stringify((0, import_js_util.getClientRect)(target)) }, attributes)
  });
};
var onGifSeen = (gif, userId, position, attributes = {}) => {
  if (!gif.analytics_response_payload) {
    return;
  }
  (0, import_js_analytics.pingback)({
    analyticsResponsePayload: gif.analytics_response_payload,
    userId,
    actionType: "SEEN",
    attributes: __spreadValues({ position: JSON.stringify(position) }, attributes)
  });
};
var onGifClick = firePingback("CLICK");
var onGifHover = firePingback("HOVER");

// src/components/attribution/overlay.tsx
var import_react7 = __toESM(require("react"));
var import_styled_components5 = __toESM(require("styled-components"));

// src/components/attribution/index.tsx
var import_react6 = __toESM(require("react"));
var import_styled_components4 = __toESM(require("styled-components"));

// src/components/attribution/avatar.tsx
var import_react3 = __toESM(require("react"));
var import_styled_components2 = __toESM(require("styled-components"));
var getSmallAvatar = (avatar) => {
  var _a, _b;
  if (!avatar)
    return "";
  const ext = (_b = (_a = avatar == null ? void 0 : avatar.split(".")) == null ? void 0 : _a.pop()) == null ? void 0 : _b.toLowerCase();
  return avatar.replace(`.${ext}`, `/80h.${ext}`);
};
var Img = import_styled_components2.default.img`
    object-fit: cover;
    width: 32px;
    height: 32px;
    margin-right: 8px;
`;
var Avatar = ({ user, className = "" }) => {
  const defaultAvatarId = (0, import_react3.useRef)(Math.floor(Math.random() * 5) + 1);
  const url = user.avatar_url ? getSmallAvatar(user.avatar_url) : `https://media.giphy.com/avatars/default${defaultAvatarId.current}.gif`;
  return /* @__PURE__ */ import_react3.default.createElement(Img, { src: url, className });
};
var avatar_default = Avatar;

// src/components/attribution/user.tsx
var import_react5 = __toESM(require("react"));
var import_styled_components3 = __toESM(require("styled-components"));

// src/components/attribution/verified-badge.tsx
var import_colors2 = require("@giphy/colors");
var import_react4 = __toESM(require("react"));
var VerifiedBadge = ({ className = "", size = 17, fill = "#15CDFF" }) => /* @__PURE__ */ import_react4.default.createElement("svg", { className: [VerifiedBadge.className, className].join(" "), height: size, width: "19px", viewBox: "0 0 19 17" }, /* @__PURE__ */ import_react4.default.createElement(
  "path",
  {
    className: VerifiedBadge.checkMarkClassName,
    d: "M9.32727273,9.44126709 L9.32727273,3.03016561 L6.55027155,3.03016561 L6.55027155,10.8150746 L6.55027155,12.188882 L12.1042739,12.188882 L12.1042739,9.44126709 L9.32727273,9.44126709 Z",
    fill: import_colors2.giphyBlack,
    transform: "translate(9.327273, 7.609524) scale(-1, 1) rotate(-45.000000) translate(-9.327273, -7.609524) "
  }
), /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(-532.000000, -466.000000)", fill }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(141.000000, 235.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(264.000000, 0.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(10.000000, 224.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(114.000000, 2.500000)" }, /* @__PURE__ */ import_react4.default.createElement("path", { d: "M15.112432,4.80769231 L16.8814194,6.87556817 L19.4157673,7.90116318 L19.6184416,10.6028916 L21.0594951,12.9065042 L19.6184416,15.2101168 L19.4157673,17.9118452 L16.8814194,18.9374402 L15.112432,21.0053161 L12.4528245,20.3611511 L9.79321699,21.0053161 L8.02422954,18.9374402 L5.48988167,17.9118452 L5.28720734,15.2101168 L3.84615385,12.9065042 L5.28720734,10.6028916 L5.48988167,7.90116318 L8.02422954,6.87556817 L9.79321699,4.80769231 L12.4528245,5.4518573 L15.112432,4.80769231 Z M17.8163503,10.8991009 L15.9282384,9.01098901 L11.5681538,13.3696923 L9.68115218,11.4818515 L7.81302031,13.3499833 L9.7011322,15.2380952 L11.5892441,17.1262071 L17.8163503,10.8991009 Z" })))))));
VerifiedBadge.className = "giphy-verified-badge";
VerifiedBadge.checkMarkClassName = "giphy-verified-checkmark";
var verified_badge_default = VerifiedBadge;

// src/components/attribution/user.tsx
var Username = import_styled_components3.default.div`
    color: white;
    font-size: 16px;
    font-weight: 700;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    -webkit-font-smoothing: antialiased;
`;
var VerifiedBadge2 = (0, import_styled_components3.default)(verified_badge_default)`
    margin-left: 4px;
    flex-shrink: 0;
`;
var Container = import_styled_components3.default.div`
    display: flex;
    align-items: center;
    min-width: 0;
`;
var User = ({ user }) => {
  const { display_name, username } = user;
  return /* @__PURE__ */ import_react5.default.createElement(Container, null, /* @__PURE__ */ import_react5.default.createElement(Username, null, display_name || `@${username}`), user.is_verified ? /* @__PURE__ */ import_react5.default.createElement(VerifiedBadge2, { size: 14 }) : null);
};
var user_default = User;

// src/components/attribution/index.tsx
var Container2 = import_styled_components4.default.div`
    display: flex;
    align-items: center;
    font-family: interface, helvetica, arial;
`;
var Avatar2 = (0, import_styled_components4.default)(avatar_default)`
    flex-shrink: 0;
`;
var Attribution = ({ gif, className, onClick }) => {
  const { user } = gif;
  if (!(user == null ? void 0 : user.username) && !(user == null ? void 0 : user.display_name)) {
    return null;
  }
  return /* @__PURE__ */ import_react6.default.createElement(
    Container2,
    {
      className: [Attribution.className, className].join(" "),
      onClick: (e) => {
        e.preventDefault();
        e.stopPropagation();
        if (onClick) {
          onClick(gif);
        } else {
          const url = user.profile_url;
          if (url)
            window.open(url, "_blank");
        }
      }
    },
    /* @__PURE__ */ import_react6.default.createElement(Avatar2, { user }),
    /* @__PURE__ */ import_react6.default.createElement(user_default, { user: gif.user })
  );
};
Attribution.className = "giphy-attribution";
var attribution_default = Attribution;

// src/components/attribution/overlay.tsx
var Background = import_styled_components5.default.div`
    background: linear-gradient(rgba(0, 0, 0, 0), rgba(18, 18, 18, 0.6));
    cursor: default;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    height: 75px;
    pointer-events: none;
`;
var Attribution2 = (0, import_styled_components5.default)(attribution_default)`
    position: absolute;
    bottom: 10px;
    left: 10px;
    right: 10px;
`;
var Container3 = import_styled_components5.default.div`
    transition: opacity 150ms ease-in;
`;
var AttributionOverlay = ({ gif, isHovered, onClick }) => {
  const hasHovered = (0, import_react7.useRef)(isHovered);
  if (isHovered) {
    hasHovered.current = true;
  }
  return gif.user && hasHovered.current ? /* @__PURE__ */ import_react7.default.createElement(Container3, { style: { opacity: isHovered ? 1 : 0 } }, /* @__PURE__ */ import_react7.default.createElement(Background, null), /* @__PURE__ */ import_react7.default.createElement(Attribution2, { gif, onClick })) : null;
};
var overlay_default = AttributionOverlay;

// src/components/pingback-context-manager.tsx
var import_js_analytics2 = require("@giphy/js-analytics");
var import_react8 = __toESM(require("react"));
var PingbackContext = (0, import_react8.createContext)({});
var PingbackContextManager = ({ attributes, children }) => {
  const { attributes: parentAttributes = {} } = (0, import_react8.useContext)(PingbackContext);
  return /* @__PURE__ */ import_react8.default.createElement(PingbackContext.Provider, { value: { attributes: (0, import_js_analytics2.mergeAttributes)(parentAttributes, attributes, "layout_type") } }, children);
};
var pingback_context_manager_default = PingbackContextManager;

// src/components/gif.tsx
var Container4 = import_styled_components6.default.div`
    position: relative;
    display: block;
    picture {
        display: block;
        width: 100%;
        height: 100%;
    }
    img {
        display: block;
    }
    .${verified_badge_default.className} {
        g {
            fill: white;
        }
    }
    .${verified_badge_default.checkMarkClassName} {
        opacity: 0;
    }
`;
var GRID_COLORS = [import_colors3.giphyBlue, import_colors3.giphyGreen, import_colors3.giphyPurple, import_colors3.giphyRed, import_colors3.giphyYellow];
var getColor = () => GRID_COLORS[Math.round(Math.random() * (GRID_COLORS.length - 1))];
var hoverTimeoutDelay = 200;
var placeholder = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";
var canUseDOM = !!(typeof window !== "undefined" && window.document && window.document.createElement);
var noop = () => {
};
var RenderOnClient = ({ children }) => {
  const [render, setRender] = (0, import_react9.useState)(false);
  (0, import_react9.useEffect)(() => {
    setRender(true);
  }, []);
  return render ? /* @__PURE__ */ import_react9.default.createElement(import_react9.default.Fragment, null, children) : null;
};
var Gif = ({
  gif,
  width,
  percentWidth,
  height: forcedHeight,
  onGifRightClick = noop,
  className = "",
  onGifClick: onGifClick2 = noop,
  onGifKeyPress = noop,
  onGifSeen: onGifSeen2 = noop,
  onGifVisible = noop,
  user = {},
  backgroundColor,
  overlay,
  hideAttribution = false,
  noLink = false,
  borderRadius = 4,
  style,
  tabIndex
}) => {
  const [hasFiredSeen, setHasFiredSeen] = (0, import_react9.useState)(false);
  const [isHovered, setHovered] = (0, import_react9.useState)(false);
  const [shouldShowMedia, setShouldShowMedia] = (0, import_react9.useState)(!canUseDOM);
  const [loadedClassname, setLoadedClassName] = (0, import_react9.useState)("");
  const defaultBgColor = (0, import_react9.useRef)(getColor());
  const container = (0, import_react9.useRef)(null);
  const image = (0, import_react9.useRef)(null);
  const showGifObserver = (0, import_react9.useRef)();
  const fullGifObserver = (0, import_react9.useRef)();
  const hoverTimeout = (0, import_react9.useRef)();
  const sendOnSeen = (0, import_react9.useRef)(noop);
  const { attributes } = (0, import_react9.useContext)(PingbackContext);
  let Overlay = overlay;
  if (!Overlay && !hideAttribution) {
    Overlay = overlay_default;
  }
  const onMouseOver = (e) => {
    clearTimeout(hoverTimeout.current);
    e.persist();
    setHovered(true);
    hoverTimeout.current = window.setTimeout(() => {
      onGifHover(gif, user == null ? void 0 : user.id, e.target, attributes);
    }, hoverTimeoutDelay);
  };
  const onMouseLeave = () => {
    clearTimeout(hoverTimeout.current);
    setHovered(false);
  };
  const onClick = (e) => {
    onGifClick(gif, user == null ? void 0 : user.id, e.target, attributes);
    onGifClick2(gif, e);
  };
  const onKeyPress = (e) => {
    onGifKeyPress(gif, e);
  };
  sendOnSeen.current = (entry) => {
    setHasFiredSeen(true);
    import_js_util2.Logger.debug(`GIF ${gif.id} seen. ${gif.title}`);
    onGifSeen(gif, user == null ? void 0 : user.id, entry.boundingClientRect, attributes);
    onGifSeen2 == null ? void 0 : onGifSeen2(gif, entry.boundingClientRect);
    if (fullGifObserver.current) {
      fullGifObserver.current.disconnect();
    }
  };
  const watchGif = () => {
    if (!fullGifObserver.current) {
      fullGifObserver.current = new IntersectionObserver(
        ([entry]) => {
          if (entry.isIntersecting) {
            sendOnSeen.current(entry);
          }
        },
        { threshold: [0.99] }
      );
    }
    if (!hasFiredSeen && container.current && fullGifObserver.current) {
      fullGifObserver.current.observe(container.current);
    }
  };
  const onImageLoad = (e) => {
    watchGif();
    onGifVisible(gif, e);
    setLoadedClassName(Gif.imgLoadedClassName);
  };
  (0, import_react9.useEffect)(() => {
    var _a, _b;
    if ((_a = image.current) == null ? void 0 : _a.complete) {
      watchGif();
      onGifVisible(gif);
    }
    (_b = fullGifObserver.current) == null ? void 0 : _b.disconnect();
    setHasFiredSeen(false);
  }, [gif.id]);
  (0, import_react9.useEffect)(() => {
    showGifObserver.current = new IntersectionObserver(([entry]) => {
      const { isIntersecting } = entry;
      setShouldShowMedia(isIntersecting);
      if (!isIntersecting && fullGifObserver.current) {
        fullGifObserver.current.disconnect();
      }
    });
    showGifObserver.current.observe(container.current);
    return () => {
      if (showGifObserver.current)
        showGifObserver.current.disconnect();
      if (fullGifObserver.current)
        fullGifObserver.current.disconnect();
      if (hoverTimeout.current)
        clearTimeout(hoverTimeout.current);
    };
  }, []);
  const height = forcedHeight || (0, import_js_util2.getGifHeight)(gif, width);
  let percentHeight;
  if (percentWidth) {
    const ratio = Math.round(height / width * 100);
    percentHeight = `${ratio}%`;
  }
  const bestRendition = (0, import_js_util2.getBestRendition)(gif.images, width, height);
  const rendition = gif.images[bestRendition.renditionName];
  const background = backgroundColor || // <- specified background prop
  // sticker has black if no backgroundColor is specified
  (gif.is_sticker ? `url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADgAAAA4AQMAAACSSKldAAAABlBMVEUhIiIWFhYoSqvJAAAAGElEQVQY02MAAv7///8PWxqIPwDZw5UGABtgwz2xhFKxAAAAAElFTkSuQmCC') 0 0` : defaultBgColor.current);
  const overflow = borderRadius ? "hidden" : "unset";
  return /* @__PURE__ */ import_react9.default.createElement(
    Container4,
    {
      as: noLink ? "div" : "a",
      href: noLink ? void 0 : gif.url,
      "data-giphy-id": gif.id,
      "data-giphy-is-sticker": gif.is_sticker,
      style: __spreadValues({
        width: percentWidth || width,
        height: percentHeight || height,
        overflow,
        borderRadius
      }, style),
      className: [Gif.className, className].join(" "),
      onMouseOver,
      onMouseLeave,
      onClick,
      onContextMenu: (e) => onGifRightClick(gif, e),
      onKeyPress,
      tabIndex,
      ref: container
    },
    /* @__PURE__ */ import_react9.default.createElement("picture", null, /* @__PURE__ */ import_react9.default.createElement(
      "source",
      {
        type: "image/webp",
        srcSet: shouldShowMedia ? rendition.webp : placeholder,
        suppressHydrationWarning: true
      }
    ), /* @__PURE__ */ import_react9.default.createElement(
      "img",
      {
        ref: image,
        suppressHydrationWarning: true,
        className: [Gif.imgClassName, loadedClassname].join(" "),
        src: shouldShowMedia ? rendition.url : placeholder,
        style: { background },
        width: "100%",
        height: "100%",
        alt: (0, import_js_util2.getAltText)(gif),
        onLoad: shouldShowMedia ? onImageLoad : () => {
        }
      }
    )),
    Overlay && // only render the overlay on the client since it depends on shouldShowMedia
    /* @__PURE__ */ import_react9.default.createElement(RenderOnClient, null, shouldShowMedia && /* @__PURE__ */ import_react9.default.createElement(Overlay, { gif, isHovered, width, height }))
  );
};
Gif.className = "giphy-gif";
Gif.imgClassName = "giphy-gif-img";
Gif.imgLoadedClassName = "giphy-img-loaded";
var gif_default = Gif;

// src/components/loader.tsx
var import_colors4 = require("@giphy/colors");
var import_react10 = __toESM(require("react"));
var import_styled_components7 = __toESM(require("styled-components"));
var bouncer = import_styled_components7.keyframes`
     to {
    transform: scale(1.75) translateY(-20px);
  }
`;
var loaderHeight = 37;
var Container5 = import_styled_components7.default.div`
    display: flex;
    align-items: center;
    height: ${loaderHeight}px;
    padding-top: 15px;
    margin: 0 auto;
    text-align: center;
    justify-content: center;
    animation: pulse 0.8s ease-in-out 0s infinite alternate backwards;
`;
var Dot = import_styled_components7.default.div`
    display: inline-block;
    height: 10px;
    width: 10px;
    margin: ${loaderHeight}px 10px 10px 10px;
    position: relative;
    box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.3);
    animation: ${bouncer} cubic-bezier(0.455, 0.03, 0.515, 0.955) 0.75s infinite alternate;
    background: ${(props) => props.$color};
    animation-delay: ${(props) => props.$delay};
`;
var Loader = ({ className = "" }) => /* @__PURE__ */ import_react10.default.createElement(Container5, { className }, /* @__PURE__ */ import_react10.default.createElement(Dot, { $color: import_colors4.giphyGreen, $delay: "0" }), /* @__PURE__ */ import_react10.default.createElement(Dot, { $color: import_colors4.giphyBlue, $delay: ".1s" }), /* @__PURE__ */ import_react10.default.createElement(Dot, { $color: import_colors4.giphyPurple, $delay: ".2s" }), /* @__PURE__ */ import_react10.default.createElement(Dot, { $color: import_colors4.giphyRed, $delay: ".3s" }), /* @__PURE__ */ import_react10.default.createElement(Dot, { $color: import_colors4.giphyYellow, $delay: ".4s" }));
var loader_default = Loader;

// src/components/masonry-grid.tsx
var import_react11 = __toESM(require("react"));
function fillArray(length, columnOffsets = []) {
  return Array.apply(null, Array(length)).map((_, index) => columnOffsets[index] || 0);
}
var MasonryGrid = ({
  columns,
  gutter,
  useTransform = true,
  itemWidth,
  itemHeights,
  children,
  columnOffsets = []
}) => {
  const containerStyle = {};
  function getChildren() {
    let columnTarget;
    const columnHeights = fillArray(columns, columnOffsets);
    const result = import_react11.default.Children.map(children, (child, index) => {
      const style = {
        position: "absolute"
      };
      columnTarget = columnHeights.indexOf(Math.min.apply(Math, columnHeights));
      const top = `${columnHeights[columnTarget]}px`;
      const left = `${columnTarget * itemWidth + columnTarget * gutter}px`;
      if (useTransform) {
        style.transform = `translate3d(${left}, ${top}, 0)`;
      } else {
        style.top = top;
        style.left = left;
      }
      const height = itemHeights[index];
      if (height) {
        columnHeights[columnTarget] += height + gutter;
      }
      return import_react11.default.cloneElement(child, { style });
    });
    containerStyle.position = "relative";
    containerStyle.width = `${columns * itemWidth + (columns - 1) * gutter}px`;
    containerStyle.height = `${Math.max.apply(Math, columnHeights) - gutter}px`;
    return result;
  }
  return /* @__PURE__ */ import_react11.default.createElement("div", { style: containerStyle }, getChildren());
};
var masonry_grid_default = (0, import_react11.memo)(MasonryGrid);

// src/components/grid.tsx
var Loader2 = import_styled_components8.default.div`
    opacity: ${(props) => props.$isFirstLoad ? 0 : 1};
`;
var defaultProps = Object.freeze({ gutter: 6, user: {}, initialGifs: [] });
var initialState = Object.freeze({
  isFetching: false,
  isError: false,
  gifWidth: 0,
  gifs: [],
  isLoaderVisible: false,
  isDoneFetching: false
});
var _Grid = class _Grid extends import_react12.PureComponent {
  constructor() {
    super(...arguments);
    this.state = __spreadProps(__spreadValues({}, initialState), { gifs: this.props.initialGifs || [] });
    this.unmounted = false;
    this.paginator = (0, import_js_fetch_api.gifPaginator)(this.props.fetchGifs, this.state.gifs);
    this.onLoaderVisible = (isVisible) => {
      if (this.unmounted)
        return;
      this.setState({ isLoaderVisible: isVisible }, this.onFetch);
    };
    this.onFetch = (0, import_throttle_debounce.debounce)(_Grid.fetchDebounce, () => __async(this, null, function* () {
      if (this.unmounted)
        return;
      const { isFetching, isLoaderVisible, gifs } = this.state;
      const prefetchCount = gifs.length;
      if (!isFetching && isLoaderVisible) {
        this.setState({ isFetching: true, isError: false });
        let gifs2;
        try {
          gifs2 = yield this.paginator();
          if (this.unmounted)
            return;
        } catch (error) {
          if (this.unmounted)
            return;
          this.setState({ isFetching: false, isError: true });
          const { onGifsFetchError } = this.props;
          if (onGifsFetchError)
            onGifsFetchError(error);
        }
        if (gifs2) {
          if (prefetchCount === gifs2.length) {
            this.setState({ isDoneFetching: true });
          } else {
            this.setState({ gifs: gifs2, isFetching: false });
            const { onGifsFetched } = this.props;
            if (onGifsFetched)
              onGifsFetched(gifs2);
            this.onFetch();
          }
        }
      }
    }));
  }
  componentDidMount() {
    this.unmounted = false;
    this.onFetch();
  }
  componentWillUnmount() {
    this.unmounted = true;
  }
  render() {
    const {
      onGifVisible,
      onGifRightClick,
      className = _Grid.className,
      onGifSeen: onGifSeen2,
      onGifClick: onGifClick2,
      onGifKeyPress,
      user,
      overlay,
      hideAttribution,
      noLink,
      borderRadius,
      noResultsMessage,
      columns,
      width,
      gutter,
      useTransform,
      columnOffsets,
      backgroundColor,
      loaderConfig,
      tabIndex = 0,
      layoutType = "GRID",
      loader: LoaderVisual = loader_default
    } = this.props;
    const { gifWidth, gifs, isError, isDoneFetching } = this.state;
    const showLoader = !isDoneFetching;
    const isFirstLoad = gifs.length === 0;
    const itemHeights = gifs.map((gif) => (0, import_js_util3.getGifHeight)(gif, gifWidth));
    return /* @__PURE__ */ import_react12.default.createElement(pingback_context_manager_default, { attributes: { layout_type: layoutType } }, /* @__PURE__ */ import_react12.default.createElement("div", { className, style: { width } }, /* @__PURE__ */ import_react12.default.createElement(
      masonry_grid_default,
      {
        itemHeights,
        useTransform,
        itemWidth: gifWidth,
        columns,
        gutter,
        columnOffsets
      },
      gifs.map((gif) => /* @__PURE__ */ import_react12.default.createElement(
        gif_default,
        {
          gif,
          tabIndex,
          key: gif.id,
          width: gifWidth,
          onGifClick: onGifClick2,
          onGifKeyPress,
          onGifSeen: onGifSeen2,
          onGifVisible,
          onGifRightClick,
          user,
          overlay,
          backgroundColor,
          hideAttribution,
          noLink,
          borderRadius
        }
      ))
    ), !showLoader && gifs.length === 0 && noResultsMessage, isError ? /* @__PURE__ */ import_react12.default.createElement(fetch_error_default, { onClick: this.onFetch }) : showLoader && /* @__PURE__ */ import_react12.default.createElement(observer_default, { onVisibleChange: this.onLoaderVisible, config: loaderConfig }, /* @__PURE__ */ import_react12.default.createElement(Loader2, { $isFirstLoad: isFirstLoad }, /* @__PURE__ */ import_react12.default.createElement(LoaderVisual, { className: _Grid.loaderClassName })))));
  }
};
_Grid.className = "giphy-grid";
_Grid.loaderClassName = "loader";
_Grid.fetchDebounce = 250;
_Grid.defaultProps = defaultProps;
_Grid.getDerivedStateFromProps = ({ columns, gutter, width }, prevState) => {
  const gutterOffset = gutter * (columns - 1);
  const gifWidth = Math.floor((width - gutterOffset) / columns);
  if (prevState.gifWidth !== gifWidth) {
    return { gifWidth };
  }
  return null;
};
var Grid = _Grid;
var grid_default = Grid;
