// @giphy/react-components/dist/components/loader.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/loader.tsx
var loader_exports = {};
__export(loader_exports, {
  default: () => loader_default
});
module.exports = __toCommonJS(loader_exports);
var import_colors = require("@giphy/colors");
var import_react = __toESM(require("react"));
var import_styled_components = __toESM(require("styled-components"));
var bouncer = import_styled_components.keyframes`
     to {
    transform: scale(1.75) translateY(-20px);
  }
`;
var loaderHeight = 37;
var Container = import_styled_components.default.div`
    display: flex;
    align-items: center;
    height: ${loaderHeight}px;
    padding-top: 15px;
    margin: 0 auto;
    text-align: center;
    justify-content: center;
    animation: pulse 0.8s ease-in-out 0s infinite alternate backwards;
`;
var Dot = import_styled_components.default.div`
    display: inline-block;
    height: 10px;
    width: 10px;
    margin: ${loaderHeight}px 10px 10px 10px;
    position: relative;
    box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.3);
    animation: ${bouncer} cubic-bezier(0.455, 0.03, 0.515, 0.955) 0.75s infinite alternate;
    background: ${(props) => props.$color};
    animation-delay: ${(props) => props.$delay};
`;
var Loader = ({ className = "" }) => /* @__PURE__ */ import_react.default.createElement(Container, { className }, /* @__PURE__ */ import_react.default.createElement(Dot, { $color: import_colors.giphyGreen, $delay: "0" }), /* @__PURE__ */ import_react.default.createElement(Dot, { $color: import_colors.giphyBlue, $delay: ".1s" }), /* @__PURE__ */ import_react.default.createElement(Dot, { $color: import_colors.giphyPurple, $delay: ".2s" }), /* @__PURE__ */ import_react.default.createElement(Dot, { $color: import_colors.giphyRed, $delay: ".3s" }), /* @__PURE__ */ import_react.default.createElement(Dot, { $color: import_colors.giphyYellow, $delay: ".4s" }));
var loader_default = Loader;
