// @replyke/core/dist/cjs/store/api/appNotificationsApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.countUnreadNotifications = exports.markAllNotificationsAsRead = exports.markNotificationAsRead = exports.fetchAppNotifications = exports.useLazyCountUnreadNotificationsQuery = exports.useCountUnreadNotificationsQuery = exports.useMarkAllNotificationsAsReadMutation = exports.useMarkNotificationAsReadMutation = exports.useLazyFetchAppNotificationsQuery = exports.useFetchAppNotificationsQuery = exports.appNotificationsApi = void 0;
var baseApi_1 = require("./baseApi");
// Extended API with app notifications endpoints
exports.appNotificationsApi = baseApi_1.baseApi.injectEndpoints({
    endpoints: function (builder) { return ({
        // Fetch paginated app notifications
        fetchAppNotifications: builder.query({
            query: function (_a) {
                var projectId = _a.projectId, page = _a.page, limit = _a.limit;
                return ({
                    url: "/".concat(projectId, "/app-notifications"),
                    method: "GET",
                    params: {
                        page: page,
                        limit: limit,
                    },
                });
            },
            providesTags: function (result, error, _a) {
                var _b;
                var projectId = _a.projectId;
                return __spreadArray([
                    { type: "AppNotification", id: "".concat(projectId, "-LIST") }
                ], ((_b = result === null || result === void 0 ? void 0 : result.map(function (_a) {
                    var id = _a.id;
                    return ({
                        type: "AppNotification",
                        id: id,
                    });
                })) !== null && _b !== void 0 ? _b : []), true);
            },
        }),
        // Mark a notification as read
        markNotificationAsRead: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, notificationId = _a.notificationId;
                return ({
                    url: "/".concat(projectId, "/app-notifications/").concat(notificationId, "/mark-as-read"),
                    method: "PATCH",
                    body: {},
                    responseHandler: function (response) { return __awaiter(void 0, void 0, void 0, function () {
                        var contentType;
                        return __generator(this, function (_a) {
                            contentType = response.headers.get('content-type') || '';
                            if (contentType.includes('application/json')) {
                                return [2 /*return*/, response.json()];
                            }
                            // For text responses, just return void since we don't need the content
                            return [2 /*return*/, response.text().then(function () { return undefined; })];
                        });
                    }); },
                });
            },
            // Optimistically update the cache
            onQueryStarted: function (_a, _b) {
                return __awaiter(this, arguments, void 0, function (_c, _d) {
                    var patches, _e;
                    var projectId = _c.projectId, notificationId = _c.notificationId;
                    var dispatch = _d.dispatch, queryFulfilled = _d.queryFulfilled;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                patches = [];
                                dispatch(exports.appNotificationsApi.util.updateQueryData("fetchAppNotifications", 
                                // We need to find all queries for this projectId - this is a simplified approach
                                { projectId: projectId, page: 1, limit: 10 }, // This should be more dynamic in practice
                                function (draft) {
                                    var notification = draft.find(function (n) { return n.id === notificationId; });
                                    if (notification) {
                                        notification.isRead = true;
                                    }
                                }));
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _f.sent();
                                return [3 /*break*/, 4];
                            case 3:
                                _e = _f.sent();
                                // Revert optimistic update on failure
                                patches.forEach(function (patch) { return patch.undo(); });
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
            invalidatesTags: function (result, error, _a) {
                var projectId = _a.projectId, notificationId = _a.notificationId;
                return [
                    { type: "AppNotification", id: notificationId },
                    { type: "AppNotification", id: "".concat(projectId, "-LIST") },
                ];
            },
        }),
        // Count unread notifications
        countUnreadNotifications: builder.query({
            query: function (_a) {
                var projectId = _a.projectId;
                return ({
                    url: "/".concat(projectId, "/app-notifications/count"),
                    method: "GET",
                });
            },
            providesTags: function (result, error, _a) {
                var projectId = _a.projectId;
                return [
                    { type: "AppNotification", id: "".concat(projectId, "-COUNT") },
                ];
            },
        }),
        // Mark all notifications as read
        markAllNotificationsAsRead: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId;
                return ({
                    url: "/".concat(projectId, "/app-notifications/mark-all-as-read"),
                    method: "PATCH",
                    body: {},
                });
            },
            // Optimistically update the cache
            onQueryStarted: function (_a, _b) {
                return __awaiter(this, arguments, void 0, function (_c, _d) {
                    var patches, _e;
                    var projectId = _c.projectId;
                    var dispatch = _d.dispatch, queryFulfilled = _d.queryFulfilled;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                patches = [];
                                // Update all notifications to be marked as read
                                dispatch(exports.appNotificationsApi.util.updateQueryData("fetchAppNotifications", 
                                // We need to find all queries for this projectId - this is a simplified approach
                                { projectId: projectId, page: 1, limit: 10 }, // This should be more dynamic in practice
                                function (draft) {
                                    draft.forEach(function (notification) {
                                        notification.isRead = true;
                                    });
                                }));
                                // Update unread count to 0
                                dispatch(exports.appNotificationsApi.util.updateQueryData("countUnreadNotifications", { projectId: projectId }, function () { return 0; }));
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _f.sent();
                                return [3 /*break*/, 4];
                            case 3:
                                _e = _f.sent();
                                // Revert optimistic update on failure
                                patches.forEach(function (patch) { return patch.undo(); });
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
            invalidatesTags: function (result, error, _a) {
                var projectId = _a.projectId;
                return [
                    { type: "AppNotification", id: "".concat(projectId, "-LIST") },
                    { type: "AppNotification", id: "".concat(projectId, "-COUNT") },
                ];
            },
        }),
    }); },
});
// Export hooks for use in components
exports.useFetchAppNotificationsQuery = exports.appNotificationsApi.useFetchAppNotificationsQuery, exports.useLazyFetchAppNotificationsQuery = exports.appNotificationsApi.useLazyFetchAppNotificationsQuery, exports.useMarkNotificationAsReadMutation = exports.appNotificationsApi.useMarkNotificationAsReadMutation, exports.useMarkAllNotificationsAsReadMutation = exports.appNotificationsApi.useMarkAllNotificationsAsReadMutation, exports.useCountUnreadNotificationsQuery = exports.appNotificationsApi.useCountUnreadNotificationsQuery, exports.useLazyCountUnreadNotificationsQuery = exports.appNotificationsApi.useLazyCountUnreadNotificationsQuery;
// Export for manual cache management
exports.fetchAppNotifications = (_a = exports.appNotificationsApi.endpoints, _a.fetchAppNotifications), exports.markNotificationAsRead = _a.markNotificationAsRead, exports.markAllNotificationsAsRead = _a.markAllNotificationsAsRead, exports.countUnreadNotifications = _a.countUnreadNotifications;
