// @replyke/core/dist/cjs/store/api/listsApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeFromList = exports.addToList = exports.deleteList = exports.updateList = exports.createList = exports.fetchSubLists = exports.fetchRootList = exports.useRemoveFromListMutation = exports.useAddToListMutation = exports.useDeleteListMutation = exports.useUpdateListMutation = exports.useCreateListMutation = exports.useLazyFetchSubListsQuery = exports.useFetchSubListsQuery = exports.useLazyFetchRootListQuery = exports.useFetchRootListQuery = exports.listsApi = void 0;
var baseApi_1 = require("./baseApi");
// Extended API with lists endpoints
exports.listsApi = baseApi_1.baseApi.injectEndpoints({
    endpoints: function (builder) { return ({
        // Fetch root list
        fetchRootList: builder.query({
            query: function (_a) {
                var projectId = _a.projectId;
                return ({
                    url: "/".concat(projectId, "/lists/root"),
                    method: "GET",
                });
            },
            providesTags: function (result, error, _a) {
                var projectId = _a.projectId;
                return __spreadArray([
                    { type: "List", id: "".concat(projectId, "-ROOT") }
                ], (result ? [{ type: "List", id: result.id }] : []), true);
            },
        }),
        // Fetch sub-lists for a parent list
        fetchSubLists: builder.query({
            query: function (_a) {
                var projectId = _a.projectId, listId = _a.listId;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(listId, "/sub-lists"),
                    method: "GET",
                });
            },
            providesTags: function (result, error, _a) {
                var _b;
                var projectId = _a.projectId, listId = _a.listId;
                return __spreadArray([
                    { type: "List", id: "".concat(projectId, "-").concat(listId, "-SUBS") }
                ], ((_b = result === null || result === void 0 ? void 0 : result.map(function (_a) {
                    var id = _a.id;
                    return ({
                        type: "List",
                        id: id,
                    });
                })) !== null && _b !== void 0 ? _b : []), true);
            },
        }),
        // Create a new sub-list
        createList: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, parentListId = _a.parentListId, listName = _a.listName;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(parentListId, "/sub-lists"),
                    method: "POST",
                    body: { listName: listName },
                });
            },
            invalidatesTags: function (result, error, _a) {
                var projectId = _a.projectId, parentListId = _a.parentListId;
                return __spreadArray([
                    { type: "List", id: "".concat(projectId, "-").concat(parentListId, "-SUBS") }
                ], (result ? [{ type: "List", id: result.id }] : []), true);
            },
        }),
        // Update list properties
        updateList: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, listId = _a.listId, update = _a.update;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(listId),
                    method: "PATCH",
                    body: { update: update },
                });
            },
            // Optimistically update the cache
            onQueryStarted: function (_a, _b) {
                return __awaiter(this, arguments, void 0, function (_c, _d) {
                    var patches, _e;
                    var projectId = _c.projectId, listId = _c.listId, update = _c.update;
                    var dispatch = _d.dispatch, queryFulfilled = _d.queryFulfilled;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                patches = [];
                                // Update root list if it's the one being updated
                                patches.push(dispatch(exports.listsApi.util.updateQueryData("fetchRootList", { projectId: projectId }, function (draft) {
                                    if (draft && draft.id === listId) {
                                        if (update.name !== undefined) {
                                            draft.name = update.name;
                                        }
                                    }
                                })));
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _f.sent();
                                return [3 /*break*/, 4];
                            case 3:
                                _e = _f.sent();
                                // Revert optimistic update on failure
                                patches.forEach(function (patch) { return patch.undo(); });
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
            invalidatesTags: function (result, error, _a) {
                var listId = _a.listId;
                return [
                    { type: "List", id: listId },
                ];
            },
        }),
        // Delete a list
        deleteList: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, listId = _a.listId;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(listId),
                    method: "DELETE",
                });
            },
            invalidatesTags: function (result, error, _a) {
                var projectId = _a.projectId, listId = _a.listId;
                return [
                    { type: "List", id: listId },
                    // Invalidate all sub-lists queries since we don't know which parent this belonged to
                    { type: "List", id: "".concat(projectId, "-SUBS-ALL") },
                ];
            },
        }),
        // Add entity to list
        addToList: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, listId = _a.listId, entityId = _a.entityId;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(listId, "/add-entity"),
                    method: "PATCH",
                    body: { entityId: entityId },
                });
            },
            // Optimistically update the cache
            onQueryStarted: function (_a, _b) {
                return __awaiter(this, arguments, void 0, function (_c, _d) {
                    var patches, _e;
                    var projectId = _c.projectId, listId = _c.listId, entityId = _c.entityId;
                    var dispatch = _d.dispatch, queryFulfilled = _d.queryFulfilled;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                patches = [];
                                // Update root list if it's the target
                                patches.push(dispatch(exports.listsApi.util.updateQueryData("fetchRootList", { projectId: projectId }, function (draft) {
                                    if (draft && draft.id === listId) {
                                        if (!draft.entityIds.includes(entityId)) {
                                            draft.entityIds.push(entityId);
                                        }
                                    }
                                })));
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _f.sent();
                                return [3 /*break*/, 4];
                            case 3:
                                _e = _f.sent();
                                // Revert optimistic update on failure
                                patches.forEach(function (patch) { return patch.undo(); });
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
            invalidatesTags: function (result, error, _a) {
                var listId = _a.listId;
                return [
                    { type: "List", id: listId },
                ];
            },
        }),
        // Remove entity from list
        removeFromList: builder.mutation({
            query: function (_a) {
                var projectId = _a.projectId, listId = _a.listId, entityId = _a.entityId;
                return ({
                    url: "/".concat(projectId, "/lists/").concat(listId, "/remove-entity"),
                    method: "PATCH",
                    body: { entityId: entityId },
                });
            },
            // Optimistically update the cache
            onQueryStarted: function (_a, _b) {
                return __awaiter(this, arguments, void 0, function (_c, _d) {
                    var patches, _e;
                    var projectId = _c.projectId, listId = _c.listId, entityId = _c.entityId;
                    var dispatch = _d.dispatch, queryFulfilled = _d.queryFulfilled;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0:
                                patches = [];
                                // Update root list if it's the target
                                patches.push(dispatch(exports.listsApi.util.updateQueryData("fetchRootList", { projectId: projectId }, function (draft) {
                                    if (draft && draft.id === listId) {
                                        draft.entityIds = draft.entityIds.filter(function (id) { return id !== entityId; });
                                    }
                                })));
                                _f.label = 1;
                            case 1:
                                _f.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _f.sent();
                                return [3 /*break*/, 4];
                            case 3:
                                _e = _f.sent();
                                // Revert optimistic update on failure
                                patches.forEach(function (patch) { return patch.undo(); });
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
            invalidatesTags: function (result, error, _a) {
                var listId = _a.listId;
                return [
                    { type: "List", id: listId },
                ];
            },
        }),
    }); },
});
// Export hooks for use in components
exports.useFetchRootListQuery = exports.listsApi.useFetchRootListQuery, exports.useLazyFetchRootListQuery = exports.listsApi.useLazyFetchRootListQuery, exports.useFetchSubListsQuery = exports.listsApi.useFetchSubListsQuery, exports.useLazyFetchSubListsQuery = exports.listsApi.useLazyFetchSubListsQuery, exports.useCreateListMutation = exports.listsApi.useCreateListMutation, exports.useUpdateListMutation = exports.listsApi.useUpdateListMutation, exports.useDeleteListMutation = exports.listsApi.useDeleteListMutation, exports.useAddToListMutation = exports.listsApi.useAddToListMutation, exports.useRemoveFromListMutation = exports.listsApi.useRemoveFromListMutation;
// Export for manual cache management
exports.fetchRootList = (_a = exports.listsApi.endpoints, _a.fetchRootList), exports.fetchSubLists = _a.fetchSubLists, exports.createList = _a.createList, exports.updateList = _a.updateList, exports.deleteList = _a.deleteList, exports.addToList = _a.addToList, exports.removeFromList = _a.removeFromList;
