// @replyke/core/dist/cjs/hooks/entity-lists/useEntityListActions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useEntityListActions = useEntityListActions;
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var entityListsSlice_1 = require("../../store/slices/entityListsSlice");
var entityListsApi_1 = require("../../store/api/entityListsApi");
var handleError_1 = require("../../utils/handleError");
var useProject_1 = __importDefault(require("../projects/useProject"));
/**
 * Redux-powered hook that provides all entity list actions
 * Uses RTK Query for stable function references and proper caching
 */
function useEntityListActions() {
    var _this = this;
    var dispatch = (0, react_redux_1.useDispatch)();
    // Get project and user context
    var projectId = (0, useProject_1.default)().projectId;
    // RTK Query hooks
    var triggerFetchEntities = (0, entityListsApi_1.useLazyFetchEntitiesQuery)()[0];
    var createEntityMutation = (0, entityListsApi_1.useCreateEntityMutation)()[0];
    var deleteEntityMutation = (0, entityListsApi_1.useDeleteEntityMutation)()[0];
    // Fetch entities action
    var fetchEntities = (0, react_1.useCallback)(function (listId, options) { return __awaiter(_this, void 0, void 0, function () {
        var result, append, err_1;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No project ID available");
                    }
                    if (!options.sortBy) {
                        console.warn("sortBy is required for fetching entities");
                        return [2 /*return*/, null];
                    }
                    dispatch((0, entityListsSlice_1.setEntityListLoading)({ listId: listId, loading: true }));
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, triggerFetchEntities({
                            projectId: projectId,
                            page: options.page,
                            sortBy: options.sortBy,
                            sortDir: options.sortDir,
                            sortType: options.sortType,
                            timeFrame: options.timeFrame,
                            userId: options.userId,
                            sourceId: options.sourceId,
                            followedOnly: (_a = options.followedOnly) !== null && _a !== void 0 ? _a : false,
                            limit: options.limit,
                            locationFilters: options.locationFilters,
                            keywordsFilters: options.keywordsFilters,
                            metadataFilters: options.metadataFilters,
                            titleFilters: options.titleFilters,
                            contentFilters: options.contentFilters,
                            attachmentsFilters: options.attachmentsFilters,
                        }).unwrap()];
                case 2:
                    result = _b.sent();
                    if (result) {
                        append = options.page > 1;
                        dispatch((0, entityListsSlice_1.setEntityListEntities)({ listId: listId, entities: result, append: append }));
                        dispatch((0, entityListsSlice_1.setEntityListHasMore)({
                            listId: listId,
                            hasMore: result.length >= options.limit,
                        }));
                        return [2 /*return*/, result];
                    }
                    return [2 /*return*/, null];
                case 3:
                    err_1 = _b.sent();
                    console.error("[EntityListActionsRedux] Failed to fetch entities for listId: ".concat(listId), err_1);
                    (0, handleError_1.handleError)(err_1, "Failed to fetch entities:");
                    dispatch((0, entityListsSlice_1.setEntityListError)({ listId: listId, error: "Failed to fetch entities" }));
                    throw err_1;
                case 4:
                    dispatch((0, entityListsSlice_1.setEntityListLoading)({ listId: listId, loading: false }));
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId, triggerFetchEntities]);
    // Create entity action
    var createEntity = (0, react_1.useCallback)(function (listId, options) { return __awaiter(_this, void 0, void 0, function () {
        var newEntity, err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No project ID available");
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, createEntityMutation({
                            projectId: projectId,
                            foreignId: options.foreignId,
                            title: options.title,
                            content: options.content,
                            attachments: options.attachments,
                            keywords: options.keywords,
                            location: options.location,
                            metadata: options.metadata,
                            sourceId: options.sourceId,
                        }).unwrap()];
                case 2:
                    newEntity = _a.sent();
                    // Add to the list
                    dispatch((0, entityListsSlice_1.addEntity)({
                        listId: listId,
                        entity: newEntity,
                        insertPosition: options.insertPosition,
                    }));
                    return [2 /*return*/, newEntity];
                case 3:
                    err_2 = _a.sent();
                    (0, handleError_1.handleError)(err_2, "Failed to create entity");
                    throw err_2;
                case 4: return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId, createEntityMutation]);
    // Delete entity action
    var deleteEntity = (0, react_1.useCallback)(function (listId, options) { return __awaiter(_this, void 0, void 0, function () {
        var err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No project ID available");
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, deleteEntityMutation({
                            projectId: projectId,
                            entityId: options.entityId,
                        }).unwrap()];
                case 2:
                    _a.sent();
                    // Remove from the list
                    dispatch((0, entityListsSlice_1.removeEntity)({ listId: listId, entityId: options.entityId }));
                    return [3 /*break*/, 4];
                case 3:
                    err_3 = _a.sent();
                    (0, handleError_1.handleError)(err_3, "Failed to delete entity");
                    throw err_3;
                case 4: return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId, deleteEntityMutation]);
    return {
        fetchEntities: fetchEntities,
        createEntity: createEntity,
        deleteEntity: deleteEntity,
    };
}
exports.default = useEntityListActions;
