// @replyke/core/dist/esm/hooks/entities/useEntityData.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { useCallback, useEffect, useRef, useState } from "react";
import useEntityVotes from "./useEntityVotes";
import useFetchEntity from "./useFetchEntity";
import useFetchEntityByForeignId from "./useFetchEntityByForeignId";
import useFetchEntityByShortId from "./useFetchEntityByShortId";
import useUpdateEntity from "./useUpdateEntity";
import useDeleteEntity from "./useDeleteEntity";
import useIncrementEntityViews from "./useIncrementEntityViews";
import { handleError } from "../../utils/handleError";
import { useUser } from "../user";
function useEntityData(_a) {
    var _this = this;
    var entityId = _a.entityId, foreignId = _a.foreignId, shortId = _a.shortId, entityProp = _a.entity, createIfNotFound = _a.createIfNotFound;
    var user = useUser().user;
    var _b = useState(entityProp), entity = _b[0], setEntity = _b[1];
    var entityViewsIncremented = useRef(false);
    // Cache to store fetched entities keyed by unique identifier
    var entityCache = useRef({});
    var userUpvotedEntity = !!((entity === null || entity === void 0 ? void 0 : entity.upvotes) &&
        (user === null || user === void 0 ? void 0 : user.id) &&
        entity.upvotes.includes(user.id));
    var userDownvotedEntity = !!((entity === null || entity === void 0 ? void 0 : entity.downvotes) &&
        (user === null || user === void 0 ? void 0 : user.id) &&
        entity.downvotes.includes(user.id));
    var fetchEntity = useFetchEntity();
    var fetchEntityByForeignId = useFetchEntityByForeignId();
    var fetchEntityByShortId = useFetchEntityByShortId();
    var _c = useEntityVotes({
        entity: entity,
        setEntity: setEntity,
    }), upvoteEntity = _c.upvoteEntity, removeEntityUpvote = _c.removeEntityUpvote, downvoteEntity = _c.downvoteEntity, removeEntityDownvote = _c.removeEntityDownvote;
    var updateEntity = useUpdateEntity();
    var incrementEntityViews = useIncrementEntityViews();
    var deleteEntity = useDeleteEntity();
    var handleUpdateEntity = useCallback(function (_a) { return __awaiter(_this, [_a], void 0, function (_b) {
        var newEntity, err_1;
        var update = _b.update;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!entity)
                        return [2 /*return*/];
                    _c.label = 1;
                case 1:
                    _c.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, updateEntity({
                            entityId: entity.id,
                            update: update,
                        })];
                case 2:
                    newEntity = _c.sent();
                    if (newEntity)
                        setEntity(newEntity);
                    return [2 /*return*/, newEntity];
                case 3:
                    err_1 = _c.sent();
                    handleError(err_1, "Failed to update entity");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [entity, updateEntity]);
    var handleIncrementEntityViews = useCallback(function () { return __awaiter(_this, void 0, void 0, function () {
        var err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!entity || entityViewsIncremented.current)
                        return [2 /*return*/];
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, incrementEntityViews({ entityId: entity.id })];
                case 2:
                    _a.sent();
                    entityViewsIncremented.current = true;
                    return [3 /*break*/, 4];
                case 3:
                    err_2 = _a.sent();
                    handleError(err_2, "Failed to increment entity views");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [entity, incrementEntityViews, entityViewsIncremented]);
    var handleDeleteEntity = useCallback(function () { return __awaiter(_this, void 0, void 0, function () {
        var err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!entity)
                        return [2 /*return*/];
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, deleteEntity({ entityId: entity.id })];
                case 2:
                    _a.sent();
                    setEntity(undefined);
                    return [3 /*break*/, 4];
                case 3:
                    err_3 = _a.sent();
                    handleError(err_3, "Failed to delete entity");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [entity, deleteEntity]);
    useEffect(function () {
        var handleFetchEntity = function () { return __awaiter(_this, void 0, void 0, function () {
            var uniqueKey, fetchedEntity, err_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!foreignId && !entityId && !shortId)
                            return [2 /*return*/];
                        if (entity && entityId && entity.id === entityId)
                            return [2 /*return*/];
                        if (entity && foreignId && entity.foreignId === foreignId)
                            return [2 /*return*/];
                        if (entity && shortId && entity.shortId === shortId)
                            return [2 /*return*/];
                        uniqueKey = "".concat(entityId !== null && entityId !== void 0 ? entityId : "", "-").concat(foreignId !== null && foreignId !== void 0 ? foreignId : "", "-").concat(shortId !== null && shortId !== void 0 ? shortId : "");
                        // If we have a cached entity, update the state and exit.
                        if (entityCache.current[uniqueKey]) {
                            setEntity(entityCache.current[uniqueKey]);
                            return [2 /*return*/];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 8, , 9]);
                        fetchedEntity = null;
                        if (!entityId) return [3 /*break*/, 3];
                        return [4 /*yield*/, fetchEntity({
                                entityId: entityId,
                            })];
                    case 2:
                        fetchedEntity = _a.sent();
                        return [3 /*break*/, 7];
                    case 3:
                        if (!foreignId) return [3 /*break*/, 5];
                        return [4 /*yield*/, fetchEntityByForeignId({
                                foreignId: foreignId,
                                createIfNotFound: createIfNotFound,
                            })];
                    case 4:
                        fetchedEntity = _a.sent();
                        return [3 /*break*/, 7];
                    case 5:
                        if (!shortId) return [3 /*break*/, 7];
                        return [4 /*yield*/, fetchEntityByShortId({
                                shortId: shortId,
                            })];
                    case 6:
                        fetchedEntity = _a.sent();
                        _a.label = 7;
                    case 7:
                        if (fetchedEntity) {
                            // Store the fetched entity in cache.
                            entityCache.current[uniqueKey] = fetchedEntity;
                            setEntity(fetchedEntity);
                        }
                        else {
                            setEntity(null);
                        }
                        return [3 /*break*/, 9];
                    case 8:
                        err_4 = _a.sent();
                        handleError(err_4, "Failed to fetch entity");
                        return [3 /*break*/, 9];
                    case 9: return [2 /*return*/];
                }
            });
        }); };
        handleFetchEntity();
    }, [
        fetchEntity,
        fetchEntityByForeignId,
        fetchEntityByShortId,
        entityId,
        foreignId,
        shortId,
        entity,
        createIfNotFound,
    ]);
    useEffect(function () {
        if (entityProp)
            setEntity(entityProp);
    }, [entityProp]);
    return {
        entity: entity,
        setEntity: setEntity,
        userUpvotedEntity: userUpvotedEntity,
        userDownvotedEntity: userDownvotedEntity,
        upvoteEntity: upvoteEntity,
        removeEntityUpvote: removeEntityUpvote,
        downvoteEntity: downvoteEntity,
        removeEntityDownvote: removeEntityDownvote,
        updateEntity: handleUpdateEntity,
        incrementEntityViews: handleIncrementEntityViews,
        deleteEntity: handleDeleteEntity,
    };
}
export default useEntityData;
