// @replyke/comments-threaded-react-js/dist/esm/components/modals/CommentMenuModalOwner/CommentMenuModalOwner.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useCommentSection } from "@replyke/react-js";
import { Modal, resetButton, resetDiv, resetUl, } from "@replyke/ui-core-react-js";
import useModalManager from "../../../hooks/useModalManager";
function CommentMenuModalOwner() {
    const { optionsComment, isCommentOptionsModalOwnerOpen, closeCommentOptionsModalOwner, } = useModalManager();
    const { deleteComment } = useCommentSection();
    const handleDeleteComment = async () => {
        const userConfirmed = window.confirm("Are you sure you want to delete this comment?");
        if (userConfirmed && optionsComment) {
            closeCommentOptionsModalOwner?.();
            await deleteComment?.({ commentId: optionsComment.id });
        }
        else {
            closeCommentOptionsModalOwner?.();
        }
    };
    return (_jsx(Modal, { show: !!isCommentOptionsModalOwnerOpen, onClose: closeCommentOptionsModalOwner, children: _jsx("div", { onClick: (e) => e.stopPropagation(), style: {
                ...resetDiv,
                backgroundColor: "white",
                borderRadius: 8,
                width: "100%",
                maxWidth: 240,
                alignSelf: "center",
                paddingTop: 8,
                paddingBottom: 8,
            }, children: _jsxs("ul", { style: {
                    ...resetUl,
                    width: "100%",
                }, children: [_jsx("li", { style: {
                            display: "flex",
                            justifyContent: "center",
                            justifySelf: "center",
                        }, children: _jsx("button", { style: {
                                ...resetButton,
                                fontWeight: 600,
                                color: "#DC2626",
                                paddingLeft: 24,
                                paddingRight: 24,
                                paddingTop: 8,
                                paddingBottom: 8,
                            }, onClick: handleDeleteComment, children: "Remove" }) }), _jsx("div", { style: { height: 1, width: "100%", backgroundColor: "#e7e7e7" } }), _jsx("li", { style: {
                            display: "flex",
                            justifyContent: "center",
                            justifySelf: "center",
                        }, children: _jsx("button", { style: {
                                ...resetButton,
                                paddingLeft: 24,
                                paddingRight: 24,
                                paddingTop: 8,
                                paddingBottom: 8,
                            }, onClick: closeCommentOptionsModalOwner, children: "Cancel" }) })] }) }) }));
}
export default CommentMenuModalOwner;
