// @replyke/comments-thread-react-js/dist/esm/components/modals/CommentMenuModal/CommentMenuModal.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { Modal, resetDiv } from "@replyke/ui-core-react-js";
import useModalManager from "../../../hooks/useModalManager";
import MainContent from "./MainContent";
import ReportContent from "./ReportContent";
function CommentMenuModal() {
    const { isCommentOptionsModalOpen, closeCommentOptionsModal } = useModalManager();
    const [view, setView] = useState("main");
    return (_jsx(Modal, { show: !!isCommentOptionsModalOpen, onClose: () => {
            closeCommentOptionsModal?.();
            setView("main");
        }, children: _jsxs("div", { onClick: (e) => e.stopPropagation(), style: {
                ...resetDiv,
                backgroundColor: "white",
                borderRadius: 8,
                width: "100%",
                maxWidth: 520,
                alignSelf: "center",
                paddingTop: 8,
                paddingBottom: 8,
            }, children: [view === "main" && (_jsx(MainContent, { clickReport: () => setView("report") })), view === "report" && (_jsx(ReportContent, { resetView: () => setView("main") }))] }) }));
}
export default CommentMenuModal;
