// @replyke/comments-thread-react-js/dist/cjs/components/modals/CommentMenuModal/CommentMenuModal.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const useModalManager_1 = __importDefault(require("../../../hooks/useModalManager"));
const MainContent_1 = __importDefault(require("./MainContent"));
const ReportContent_1 = __importDefault(require("./ReportContent"));
function CommentMenuModal() {
    const { isCommentOptionsModalOpen, closeCommentOptionsModal } = (0, useModalManager_1.default)();
    const [view, setView] = (0, react_1.useState)("main");
    return ((0, jsx_runtime_1.jsx)(ui_core_react_js_1.Modal, { show: !!isCommentOptionsModalOpen, onClose: () => {
            closeCommentOptionsModal?.();
            setView("main");
        }, children: (0, jsx_runtime_1.jsxs)("div", { onClick: (e) => e.stopPropagation(), style: {
                ...ui_core_react_js_1.resetDiv,
                backgroundColor: "white",
                borderRadius: 8,
                width: "100%",
                maxWidth: 520,
                alignSelf: "center",
                paddingTop: 8,
                paddingBottom: 8,
            }, children: [view === "main" && ((0, jsx_runtime_1.jsx)(MainContent_1.default, { clickReport: () => setView("report") })), view === "report" && ((0, jsx_runtime_1.jsx)(ReportContent_1.default, { resetView: () => setView("main") }))] }) }));
}
exports.default = CommentMenuModal;
