// @replyke/comments-thread-react-js/dist/cjs/components/modals/CommentMenuModal/ReportContent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_js_1 = require("@replyke/react-js");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const useModalManager_1 = __importDefault(require("../../../hooks/useModalManager"));
function ReportContent({ resetView }) {
    const { user } = (0, react_js_1.useUser)();
    const { callbacks } = (0, react_js_1.useCommentSection)();
    const { optionsComment, closeCommentOptionsModal } = (0, useModalManager_1.default)();
    const { createCommentReport } = (0, react_js_1.useCreateReport)();
    const [submitting, setSubmitting] = (0, react_1.useState)(false);
    const [reason, setReason] = (0, react_1.useState)(null);
    const buttonActive = (0, react_1.useMemo)(() => !!reason && !!optionsComment, [reason, optionsComment]);
    const handleSubmitReport = async () => {
        try {
            if (!optionsComment)
                throw new Error("No comment to report selected");
            if (!reason)
                throw new Error("No reason to report selected");
            if (!user) {
                callbacks?.loginRequiredCallback?.();
                return;
            }
            setSubmitting(true);
            await createCommentReport({ targetId: optionsComment.id, reason });
            closeCommentOptionsModal?.();
            setReason(null);
            resetView();
        }
        catch (err) {
            (0, react_js_1.handleError)(err, "Submitting report failed");
        }
        finally {
            setSubmitting(false);
        }
    };
    return ((0, jsx_runtime_1.jsxs)("div", { style: { padding: 24, width: "100%" }, children: [(0, jsx_runtime_1.jsxs)("div", { style: { display: "flex", alignItems: "center", gap: "16px" }, children: [(0, jsx_runtime_1.jsx)(ui_core_react_js_1.FlagIcon, { size: 24, color: "#000" }), (0, jsx_runtime_1.jsx)("span", { style: { fontSize: "24px" }, children: "Submit a report" })] }), (0, jsx_runtime_1.jsx)("p", { style: { marginTop: "24px" }, children: "Thank you for looking out for our community. Let us know what is happening, and we'll look into it." }), (0, jsx_runtime_1.jsx)("div", { style: {
                    display: "flex",
                    flexWrap: "wrap",
                    gap: "6px",
                    marginTop: "24px",
                }, children: Object.entries(react_js_1.reportReasons).map(([key, value], index) => ((0, jsx_runtime_1.jsx)("button", { onClick: () => setReason(key), style: {
                        padding: "4px 8px",
                        borderRadius: "6px",
                        backgroundColor: key === reason ? "#000" : "#e5e7eb",
                        cursor: "pointer",
                        color: key === reason ? "#fff" : "#9d9d9e",
                        fontSize: 14,
                    }, children: value }, index))) }), (0, jsx_runtime_1.jsx)("button", { onClick: handleSubmitReport, disabled: !buttonActive, style: {
                    width: "100%",
                    marginTop: 16,
                    display: "inline-flex",
                    alignItems: "center",
                    justifyContent: "center",
                    padding: "0.5rem 1rem",
                    fontSize: "0.875rem",
                    fontWeight: "500",
                    lineHeight: "1.25",
                    borderRadius: "0.375rem",
                    backgroundColor: buttonActive ? "#000" : "#e5e7eb",
                    color: buttonActive ? "#ffffff" : "#9d9d9e",
                    cursor: buttonActive ? "pointer" : "not-allowed",
                    transition: "background-color 0.2s ease, border-color 0.2s ease, color 0.2s ease",
                }, children: submitting ? "Submitting.." : "Submit Report" })] }));
}
exports.default = ReportContent;
