// @replyke/core/dist/cjs/store/slices/authThunks.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeAuthThunk = exports.changePasswordThunk = exports.verifyExternalUserThunk = exports.requestNewAccessTokenThunk = exports.signOutThunk = exports.signInWithEmailAndPasswordThunk = exports.signUpWithEmailAndPasswordThunk = void 0;
var toolkit_1 = require("@reduxjs/toolkit");
var axios_1 = __importDefault(require("../../config/axios"));
var isReactNative_1 = require("../../utils/isReactNative");
var handleError_1 = require("../../utils/handleError");
var authSlice_1 = require("./authSlice");
var userSlice_1 = require("./userSlice");
// Auth service functions - calling existing API patterns directly
var authService = {
    signUpWithEmailAndPassword: function (projectId, data) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            var _a, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/sign-up"), {
                            email: data.email,
                            password: data.password,
                            name: (_a = data.name) === null || _a === void 0 ? void 0 : _a.trim(),
                            username: (_b = data.username) === null || _b === void 0 ? void 0 : _b.trim(),
                            avatar: data.avatar,
                            bio: (_c = data.bio) === null || _c === void 0 ? void 0 : _c.trim(),
                            location: data.location,
                            birthdate: data.birthdate,
                            metadata: data.metadata,
                            secureMetadata: data.secureMetadata,
                        }, { withCredentials: !(0, isReactNative_1.isReactNative)() })];
                    case 1:
                        response = _d.sent();
                        return [2 /*return*/, response.data];
                }
            });
        });
    },
    signInWithEmailAndPassword: function (projectId, data) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/sign-in"), data, {
                            withCredentials: !(0, isReactNative_1.isReactNative)(),
                        })];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response.data];
                }
            });
        });
    },
    signOut: function (projectId, refreshToken) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/sign-out"), { refreshToken: refreshToken }, { withCredentials: !(0, isReactNative_1.isReactNative)() })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    },
    requestNewAccessToken: function (projectId, refreshToken) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/request-new-access-token"), { refreshToken: refreshToken }, { withCredentials: !(0, isReactNative_1.isReactNative)() })];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response.data];
                }
            });
        });
    },
    verifyExternalUser: function (projectId, userJwt) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/verify-external-user"), { userJwt: userJwt })];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response.data];
                }
            });
        });
    },
    changePassword: function (projectId, data) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.post("/".concat(projectId, "/auth/change-password"), data, {
                            withCredentials: !(0, isReactNative_1.isReactNative)(),
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    },
};
// Async Thunks
exports.signUpWithEmailAndPasswordThunk = (0, toolkit_1.createAsyncThunk)("auth/signUpWithEmailAndPassword", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var result, error_1;
    var dispatch = _b.dispatch, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _c.trys.push([0, 2, 3, 4]);
                dispatch((0, authSlice_1.setAuthenticating)(true));
                return [4 /*yield*/, authService.signUpWithEmailAndPassword(data.projectId, data)];
            case 1:
                result = _c.sent();
                // Update auth state
                dispatch((0, authSlice_1.setTokens)({
                    accessToken: result.accessToken,
                    refreshToken: result.refreshToken,
                }));
                dispatch((0, authSlice_1.setUser)(result.user));
                dispatch((0, userSlice_1.setUser)(result.user)); // Sync user to user slice
                return [2 /*return*/, result];
            case 2:
                error_1 = _c.sent();
                (0, handleError_1.handleError)(error_1, "Failed to register user with email and password:");
                return [2 /*return*/, rejectWithValue(error_1 instanceof Error ? error_1.message : "Unknown error")];
            case 3:
                dispatch((0, authSlice_1.setAuthenticating)(false));
                return [7 /*endfinally*/];
            case 4: return [2 /*return*/];
        }
    });
}); });
exports.signInWithEmailAndPasswordThunk = (0, toolkit_1.createAsyncThunk)("auth/signInWithEmailAndPassword", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var result, error_2;
    var dispatch = _b.dispatch, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _c.trys.push([0, 2, 3, 4]);
                dispatch((0, authSlice_1.setAuthenticating)(true));
                return [4 /*yield*/, authService.signInWithEmailAndPassword(data.projectId, data)];
            case 1:
                result = _c.sent();
                // Update auth state
                dispatch((0, authSlice_1.setTokens)({
                    accessToken: result.accessToken,
                    refreshToken: result.refreshToken,
                }));
                dispatch((0, authSlice_1.setUser)(result.user));
                dispatch((0, userSlice_1.setUser)(result.user)); // Sync user to user slice
                return [2 /*return*/, result];
            case 2:
                error_2 = _c.sent();
                (0, handleError_1.handleError)(error_2, "Failed to log user in:");
                return [2 /*return*/, rejectWithValue(error_2 instanceof Error ? error_2.message : "Unknown error")];
            case 3:
                dispatch((0, authSlice_1.setAuthenticating)(false));
                return [7 /*endfinally*/];
            case 4: return [2 /*return*/];
        }
    });
}); });
exports.signOutThunk = (0, toolkit_1.createAsyncThunk)("auth/signOut", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var state, refreshToken, error_3;
    var dispatch = _b.dispatch, getState = _b.getState, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                state = getState();
                refreshToken = state.auth.refreshToken;
                // If React Native and no refresh token, throw error (matches original logic)
                if ((0, isReactNative_1.isReactNative)() && !refreshToken) {
                    throw new Error("No refresh token");
                }
                _c.label = 1;
            case 1:
                _c.trys.push([1, 3, 4, 5]);
                dispatch((0, authSlice_1.setAuthenticating)(true));
                return [4 /*yield*/, authService.signOut(data.projectId, refreshToken)];
            case 2:
                _c.sent();
                // Clear auth state
                dispatch((0, authSlice_1.resetAuth)());
                dispatch((0, userSlice_1.clearUser)()); // Clear user from user slice
                return [2 /*return*/];
            case 3:
                error_3 = _c.sent();
                (0, handleError_1.handleError)(error_3, "Failed to log user out:");
                return [2 /*return*/, rejectWithValue(error_3 instanceof Error ? error_3.message : "Unknown error")];
            case 4:
                dispatch((0, authSlice_1.setAuthenticating)(false));
                return [7 /*endfinally*/];
            case 5: return [2 /*return*/];
        }
    });
}); });
exports.requestNewAccessTokenThunk = (0, toolkit_1.createAsyncThunk)("auth/requestNewAccessToken", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var state, refreshToken, result, error_4;
    var dispatch = _b.dispatch, getState = _b.getState, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                state = getState();
                refreshToken = state.auth.refreshToken;
                // If React Native and no refresh token, return early
                if ((0, isReactNative_1.isReactNative)() && !refreshToken) {
                    return [2 /*return*/];
                }
                _c.label = 1;
            case 1:
                _c.trys.push([1, 3, , 4]);
                return [4 /*yield*/, authService.requestNewAccessToken(data.projectId, refreshToken)];
            case 2:
                result = _c.sent();
                // Update auth state
                dispatch((0, authSlice_1.setTokens)({ accessToken: result.accessToken }));
                dispatch((0, authSlice_1.setUser)(result.user));
                dispatch((0, userSlice_1.setUser)(result.user)); // Sync user to user slice
                return [2 /*return*/, result.accessToken];
            case 3:
                error_4 = _c.sent();
                (0, handleError_1.handleError)(error_4, "Request new access token error:");
                return [2 /*return*/, rejectWithValue(error_4 instanceof Error ? error_4.message : "Unknown error")];
            case 4: return [2 /*return*/];
        }
    });
}); });
exports.verifyExternalUserThunk = (0, toolkit_1.createAsyncThunk)("auth/verifyExternalUser", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var result, error_5;
    var dispatch = _b.dispatch, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _c.trys.push([0, 2, , 3]);
                return [4 /*yield*/, authService.verifyExternalUser(data.projectId, data.userJwt)];
            case 1:
                result = _c.sent();
                // Update auth state
                dispatch((0, authSlice_1.setTokens)({
                    accessToken: result.accessToken,
                    refreshToken: result.refreshToken,
                }));
                dispatch((0, authSlice_1.setUser)(result.user));
                dispatch((0, userSlice_1.setUser)(result.user)); // Sync user to user slice
                return [2 /*return*/, result];
            case 2:
                error_5 = _c.sent();
                (0, handleError_1.handleError)(error_5, "Verify external user error:");
                return [2 /*return*/, rejectWithValue(error_5 instanceof Error ? error_5.message : "Unknown error")];
            case 3: return [2 /*return*/];
        }
    });
}); });
exports.changePasswordThunk = (0, toolkit_1.createAsyncThunk)("auth/changePassword", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var state, error_6;
    var dispatch = _b.dispatch, getState = _b.getState, rejectWithValue = _b.rejectWithValue;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                state = getState();
                if (!state.auth.user) {
                    throw new Error("No user is authenticated");
                }
                _c.label = 1;
            case 1:
                _c.trys.push([1, 3, 4, 5]);
                dispatch((0, authSlice_1.setAuthenticating)(true));
                return [4 /*yield*/, authService.changePassword(data.projectId, data)];
            case 2:
                _c.sent();
                return [2 /*return*/];
            case 3:
                error_6 = _c.sent();
                (0, handleError_1.handleError)(error_6, "Failed to change password:");
                return [2 /*return*/, rejectWithValue(error_6 instanceof Error ? error_6.message : "Unknown error")];
            case 4:
                dispatch((0, authSlice_1.setAuthenticating)(false));
                return [7 /*endfinally*/];
            case 5: return [2 /*return*/];
        }
    });
}); });
// Initialize auth - handles the startup flow from useAuthData
exports.initializeAuthThunk = (0, toolkit_1.createAsyncThunk)("auth/initialize", function (data_1, _a) { return __awaiter(void 0, [data_1, _a], void 0, function (data, _b) {
    var error_7;
    var dispatch = _b.dispatch;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _c.trys.push([0, 3, , 4]);
                dispatch((0, authSlice_1.setLoadingInitial)(true));
                if (!data.signedToken) return [3 /*break*/, 2];
                return [4 /*yield*/, dispatch((0, exports.verifyExternalUserThunk)({
                        projectId: data.projectId,
                        userJwt: data.signedToken,
                    }))];
            case 1:
                _c.sent();
                _c.label = 2;
            case 2:
                // Step 2: Try to refresh access token (matches original setTimeout logic)
                setTimeout(function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, dispatch((0, exports.requestNewAccessTokenThunk)({ projectId: data.projectId }))];
                            case 1:
                                _a.sent();
                                dispatch((0, authSlice_1.setLoadingInitial)(false));
                                dispatch((0, authSlice_1.setInitialized)(true));
                                return [2 /*return*/];
                        }
                    });
                }); }, 0);
                return [3 /*break*/, 4];
            case 3:
                error_7 = _c.sent();
                dispatch((0, authSlice_1.setLoadingInitial)(false));
                dispatch((0, authSlice_1.setInitialized)(true));
                (0, handleError_1.handleError)(error_7, "Auth initialization failed:");
                return [3 /*break*/, 4];
            case 4: return [2 /*return*/];
        }
    });
}); });
