// @replyke/core/dist/cjs/hooks/auth/useAuth.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = useAuth;
var react_redux_1 = require("react-redux");
var react_1 = require("react");
var authSlice_1 = require("../../store/slices/authSlice");
var authThunks_1 = require("../../store/slices/authThunks");
var useProject_1 = __importDefault(require("../projects/useProject"));
function useAuth() {
    var _this = this;
    var dispatch = (0, react_redux_1.useDispatch)();
    var projectId = (0, useProject_1.default)().projectId;
    // Selectors
    var loadingInitial = (0, react_redux_1.useSelector)(function (state) { return (0, authSlice_1.selectLoadingInitial)(state); });
    var accessToken = (0, react_redux_1.useSelector)(function (state) { return (0, authSlice_1.selectAccessToken)(state); });
    var refreshToken = (0, react_redux_1.useSelector)(function (state) { return (0, authSlice_1.selectRefreshToken)(state); });
    // Actions
    var handleSignUpWithEmailAndPassword = (0, react_1.useCallback)(function (props) { return __awaiter(_this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No projectId available.");
                    }
                    return [4 /*yield*/, dispatch((0, authThunks_1.signUpWithEmailAndPasswordThunk)(__assign({ projectId: projectId }, props)))];
                case 1:
                    result = _a.sent();
                    if (authThunks_1.signUpWithEmailAndPasswordThunk.rejected.match(result)) {
                        throw new Error(result.payload);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId]);
    var handleSignInWithEmailAndPassword = (0, react_1.useCallback)(function (props) { return __awaiter(_this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No projectId available.");
                    }
                    return [4 /*yield*/, dispatch((0, authThunks_1.signInWithEmailAndPasswordThunk)(__assign({ projectId: projectId }, props)))];
                case 1:
                    result = _a.sent();
                    if (authThunks_1.signInWithEmailAndPasswordThunk.rejected.match(result)) {
                        throw new Error(result.payload);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId]);
    var handleSignOut = (0, react_1.useCallback)(function () { return __awaiter(_this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No projectId available.");
                    }
                    return [4 /*yield*/, dispatch((0, authThunks_1.signOutThunk)({ projectId: projectId }))];
                case 1:
                    result = _a.sent();
                    if (authThunks_1.signOutThunk.rejected.match(result)) {
                        throw new Error(result.payload);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId]);
    var handleChangePassword = (0, react_1.useCallback)(function (props) { return __awaiter(_this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        throw new Error("No projectId available.");
                    }
                    return [4 /*yield*/, dispatch((0, authThunks_1.changePasswordThunk)(__assign({ projectId: projectId }, props)))];
                case 1:
                    result = _a.sent();
                    if (authThunks_1.changePasswordThunk.rejected.match(result)) {
                        throw new Error(result.payload);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId]);
    var handleRequestNewAccessToken = (0, react_1.useCallback)(function () { return __awaiter(_this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId)
                        return [2 /*return*/];
                    return [4 /*yield*/, dispatch((0, authThunks_1.requestNewAccessTokenThunk)({ projectId: projectId }))];
                case 1:
                    result = _a.sent();
                    if (authThunks_1.requestNewAccessTokenThunk.fulfilled.match(result)) {
                        return [2 /*return*/, result.payload];
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [dispatch, projectId]);
    var handleSetRefreshToken = (0, react_1.useCallback)(function (token) {
        // Handle both direct value and function setter patterns
        if (typeof token === 'function') {
            var currentToken = refreshToken;
            var newToken = token(currentToken);
            dispatch((0, authSlice_1.setRefreshToken)(newToken));
        }
        else {
            dispatch((0, authSlice_1.setRefreshToken)(token));
        }
    }, [dispatch, refreshToken]);
    return {
        loadingInitial: loadingInitial,
        accessToken: accessToken,
        refreshToken: refreshToken,
        setRefreshToken: handleSetRefreshToken,
        signUpWithEmailAndPassword: handleSignUpWithEmailAndPassword,
        signInWithEmailAndPassword: handleSignInWithEmailAndPassword,
        signOut: handleSignOut,
        changePassword: handleChangePassword,
        requestNewAccessToken: handleRequestNewAccessToken,
    };
}
