// @replyke/core/dist/cjs/hooks/comments/useCommentVotes.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var handleError_1 = require("../../utils/handleError");
var user_1 = require("../user");
// Import the communication hooks
var useUpvoteComment_1 = __importDefault(require("./useUpvoteComment"));
var useRemoveCommentUpvote_1 = __importDefault(require("./useRemoveCommentUpvote"));
var useDownvoteComment_1 = __importDefault(require("./useDownvoteComment"));
var useRemoveCommentDownvote_1 = __importDefault(require("./useRemoveCommentDownvote"));
function useCommentVotes(props) {
    var _this = this;
    var user = (0, user_1.useUser)().user;
    var comment = props.comment, setComment = props.setComment;
    // Save the previous state in case we need to revert
    var _a = (0, react_1.useState)(comment), previousComment = _a[0], setPreviousComment = _a[1];
    var _b = (0, react_1.useState)(false), isUpdating = _b[0], setIsUpdating = _b[1];
    // Helper to update state and remember the previous value
    var updateCommentState = function (newCommentState) {
        setPreviousComment(comment);
        setComment(newCommentState);
    };
    var revertToPreviousState = (0, react_1.useCallback)(function () {
        setComment(previousComment);
    }, [previousComment, setComment]);
    // Get our request functions from the extracted hooks
    var upvoteCommentRequest = (0, useUpvoteComment_1.default)();
    var removeUpvoteRequest = (0, useRemoveCommentUpvote_1.default)();
    var downvoteCommentRequest = (0, useDownvoteComment_1.default)();
    var removeDownvoteRequest = (0, useRemoveCommentDownvote_1.default)();
    var handleUpvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedComment, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: add upvote and remove any downvote from the user
                    updateCommentState(__assign(__assign({}, comment), { upvotes: __spreadArray(__spreadArray([], comment.upvotes, true), [user.id], false), downvotes: comment.downvotes.filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, upvoteCommentRequest({
                            commentId: comment.id,
                        })];
                case 2:
                    updatedComment = _a.sent();
                    setComment(updatedComment);
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_1, "Failed to update comment vote: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleRemoveUpvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedComment, err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: remove the user's upvote
                    updateCommentState(__assign(__assign({}, comment), { upvotes: comment.upvotes.filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, removeUpvoteRequest({
                            commentId: comment.id,
                        })];
                case 2:
                    updatedComment = _a.sent();
                    setComment(updatedComment);
                    return [3 /*break*/, 5];
                case 3:
                    err_2 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_2, "Failed to remove upvote from comment: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleDownvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedComment, err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: add downvote and remove any upvote from the user
                    updateCommentState(__assign(__assign({}, comment), { downvotes: __spreadArray(__spreadArray([], comment.downvotes, true), [user.id], false), upvotes: comment.upvotes.filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, downvoteCommentRequest({
                            commentId: comment.id,
                        })];
                case 2:
                    updatedComment = _a.sent();
                    setComment(updatedComment);
                    return [3 /*break*/, 5];
                case 3:
                    err_3 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_3, "Failed to update comment vote: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var handleRemoveDownvote = function () { return __awaiter(_this, void 0, void 0, function () {
        var updatedComment, err_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user)
                        throw new Error("No authenticated user");
                    if (isUpdating)
                        return [2 /*return*/];
                    // Optimistic update: remove the user's downvote
                    updateCommentState(__assign(__assign({}, comment), { downvotes: comment.downvotes.filter(function (id) { return id !== user.id; }) }));
                    setIsUpdating(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, removeDownvoteRequest({
                            commentId: comment.id,
                        })];
                case 2:
                    updatedComment = _a.sent();
                    setComment(updatedComment);
                    return [3 /*break*/, 5];
                case 3:
                    err_4 = _a.sent();
                    revertToPreviousState();
                    (0, handleError_1.handleError)(err_4, "Failed to remove downvote from comment: ");
                    return [3 /*break*/, 5];
                case 4:
                    setIsUpdating(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    return {
        upvoteComment: handleUpvote,
        removeCommentUpvote: handleRemoveUpvote,
        downvoteComment: handleDownvote,
        removeCommentDownvote: handleRemoveDownvote,
    };
}
exports.default = useCommentVotes;
