// @replyke/core/dist/esm/hooks/user/useUser.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { useEffect, useMemo, useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { setProjectContext, selectUser, selectUserLoading, selectUserUpdating, selectCurrentProjectId, selectUserError, } from "../../store/slices/userSlice";
import { useUserActions } from "./useUserActions";
import useProject from "../projects/useProject";
import { selectUser as selectAuthUser } from "../../store/slices/authSlice";
/**
 * Redux-powered hook that provides comprehensive user management
 * This replaces useUserData and provides the same interface with Redux state management
 */
function useUser(_) {
    var _this = this;
    if (_ === void 0) { _ = {}; }
    var dispatch = useDispatch();
    // Get external context
    var projectId = useProject().projectId;
    // Get Redux state
    var user = useSelector(function (state) { return selectUser(state); });
    var authUser = useSelector(function (state) { return selectAuthUser(state); }); // Fallback to auth user
    var loading = useSelector(function (state) { return selectUserLoading(state); });
    var updating = useSelector(function (state) { return selectUserUpdating(state); });
    var error = useSelector(function (state) { return selectUserError(state); });
    var currentProjectId = useSelector(function (state) { return selectCurrentProjectId(state); });
    // Get actions
    var _a = useUserActions(), setUser = _a.setUser, updateUserAction = _a.updateUser, clearError = _a.clearError;
    // Update Redux state when project changes
    useEffect(function () {
        if (projectId && projectId !== currentProjectId) {
            dispatch(setProjectContext(projectId));
        }
    }, [dispatch, projectId, currentProjectId]);
    // Sync auth user to user slice when auth user changes and we don't have a user yet
    // IMPORTANT: Only sync if authUser has valid data (has an id) to prevent empty objects from being set
    useEffect(function () {
        if (authUser && authUser.id && !user) {
            setUser(authUser);
        }
    }, [authUser, user, setUser]);
    // Current user operations with projectId included automatically
    var handleUpdateUser = useCallback(function (update) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!user) {
                        throw new Error("No user available to update");
                    }
                    if (!projectId) {
                        throw new Error("No projectId available");
                    }
                    return [4 /*yield*/, updateUserAction(projectId, user.id, update, user)];
                case 1: 
                // Pass current user for optimistic update reversion if needed
                return [2 /*return*/, _a.sent()];
            }
        });
    }); }, [updateUserAction, user, projectId]);
    // Return focused interface for current user management
    return useMemo(function () { return ({
        user: user || authUser, // Fallback to auth user if user slice is empty
        loading: loading,
        updating: updating,
        error: error,
        updateUser: handleUpdateUser,
        clearError: clearError,
    }); }, [
        user,
        authUser,
        loading,
        updating,
        error,
        handleUpdateUser,
        clearError,
    ]);
}
export default useUser;
