// @replyke/comments-threaded-react-js/dist/cjs/components/CommentsFeed/CommentThread/ActionMenu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_js_1 = require("@replyke/react-js");
const useModalManager_1 = __importDefault(require("../../../hooks/useModalManager"));
function ActionMenu({ comment }) {
    const { user } = (0, react_js_1.useUser)();
    const { openCommentOptionsModal, openCommentOptionsModalOwner } = (0, useModalManager_1.default)();
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: "relative" }, children: (0, jsx_runtime_1.jsx)("button", { onClick: () => user && user.id === comment.userId
                ? openCommentOptionsModalOwner?.(comment)
                : openCommentOptionsModal?.(comment), style: {
                padding: "2px",
                color: "#9CA3AF",
                transition: "colors 150ms ease-in-out",
                borderRadius: "4px",
                background: "transparent",
                border: "none",
                cursor: "pointer",
            }, onMouseEnter: (e) => {
                e.currentTarget.style.color = "#4B5563";
                e.currentTarget.style.backgroundColor = "#F3F4F6";
            }, onMouseLeave: (e) => {
                e.currentTarget.style.color = "#9CA3AF";
                e.currentTarget.style.backgroundColor = "transparent";
            }, title: "More options", children: (0, jsx_runtime_1.jsx)("svg", { style: { width: "12px", height: "12px" }, fill: "currentColor", viewBox: "0 0 20 20", children: (0, jsx_runtime_1.jsx)("path", { d: "M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z" }) }) }) }));
}
exports.default = ActionMenu;
