// @replyke/comments-thread-react-js/dist/cjs/components/NewCommentForm.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewCommentForm = NewCommentForm;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_js_1 = require("@replyke/react-js");
function NewCommentForm() {
    const { user } = (0, react_js_1.useUser)();
    const [content, setContent] = (0, react_1.useState)("");
    const [isSubmitting, setIsSubmitting] = (0, react_1.useState)(false);
    const { createComment, callbacks } = (0, react_js_1.useCommentSection)();
    const hasContent = content.trim().length > 0;
    const handleSubmit = async (e) => {
        e.preventDefault();
        if (!hasContent || isSubmitting)
            return;
        if (!user) {
            callbacks?.loginRequiredCallback();
            return;
        }
        setIsSubmitting(true);
        try {
            await createComment?.({ content: content.trim(), mentions: [] });
            setContent("");
        }
        catch (error) {
            console.error("Error creating comment:", error);
        }
        finally {
            setIsSubmitting(false);
        }
    };
    return ((0, jsx_runtime_1.jsx)("form", { onSubmit: handleSubmit, style: { position: "relative" }, children: (0, jsx_runtime_1.jsxs)("div", { style: {
                display: "flex",
                alignItems: "flex-end",
                backgroundColor: "#FFFFFF",
                borderRadius: "16px",
                border: `1px solid ${hasContent ? "#BFDBFE" : "#E5E7EB"}`,
                boxShadow: hasContent
                    ? "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)"
                    : "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
                transition: "all 300ms ease-in-out",
                padding: "8px",
            }, onMouseEnter: (e) => {
                if (!hasContent) {
                    e.currentTarget.style.boxShadow =
                        "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)";
                }
            }, onMouseLeave: (e) => {
                if (!hasContent) {
                    e.currentTarget.style.boxShadow =
                        "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)";
                }
            }, children: [(0, jsx_runtime_1.jsx)("textarea", { value: content, onChange: (e) => setContent(e.target.value), placeholder: "Add your reply...", style: {
                        flex: 1,
                        padding: "8px",
                        backgroundColor: "transparent",
                        color: "#111827",
                        fontSize: "12px",
                        lineHeight: "1.625",
                        outline: "none",
                        resize: "none",
                        border: "none",
                    }, rows: 2 }), (0, jsx_runtime_1.jsx)("button", { type: "submit", disabled: !hasContent || isSubmitting, style: {
                        flexShrink: 0,
                        padding: "8px",
                        borderRadius: "50%",
                        backgroundColor: hasContent && !isSubmitting ? "#2563EB" : "#E5E7EB",
                        color: hasContent && !isSubmitting ? "#FFFFFF" : "#9CA3AF",
                        boxShadow: "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
                        transition: "all 200ms ease-in-out",
                        border: "none",
                        cursor: !hasContent || isSubmitting ? "not-allowed" : "pointer",
                        outline: "none",
                    }, onMouseEnter: (e) => {
                        if (hasContent && !isSubmitting) {
                            e.currentTarget.style.backgroundColor = "#1D4ED8";
                            e.currentTarget.style.boxShadow =
                                "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)";
                        }
                    }, onMouseLeave: (e) => {
                        if (hasContent && !isSubmitting) {
                            e.currentTarget.style.backgroundColor = "#2563EB";
                            e.currentTarget.style.boxShadow =
                                "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)";
                        }
                    }, onFocus: (e) => {
                        e.currentTarget.style.outline = "2px solid #3B82F6";
                        e.currentTarget.style.outlineOffset = "2px";
                    }, onBlur: (e) => {
                        e.currentTarget.style.outline = "none";
                        e.currentTarget.style.outlineOffset = "0";
                    }, children: (0, jsx_runtime_1.jsx)("svg", { style: {
                            height: "12px",
                            width: "12px",
                            transition: "transform 200ms ease-in-out",
                            transform: hasContent ? "scale(1)" : "scale(1)",
                        }, onMouseEnter: (e) => {
                            if (hasContent) {
                                e.currentTarget.style.transform = "scale(1.1)";
                            }
                        }, onMouseLeave: (e) => {
                            if (hasContent) {
                                e.currentTarget.style.transform = "scale(1)";
                            }
                        }, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", children: (0, jsx_runtime_1.jsx)("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M5 10l7-7m0 0l7 7m-7-7v18" }) }) })] }) }));
}
