// @replyke/comments-social-core/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.socialBaseStyle = exports.useSocialStyleConfig = exports.useSocialStyle = exports.SocialStyleConfigProvider = exports.mergeSocialStyleData = exports.isSocialStyleConfig = void 0;
// Social section - helpers
var isSocialStyleConfig_1 = require("./helpers/isSocialStyleConfig");
Object.defineProperty(exports, "isSocialStyleConfig", { enumerable: true, get: function () { return isSocialStyleConfig_1.isSocialStyleConfig; } });
var mergeSocialStyleData_1 = require("./helpers/mergeSocialStyleData");
Object.defineProperty(exports, "mergeSocialStyleData", { enumerable: true, get: function () { return mergeSocialStyleData_1.mergeSocialStyleData; } });
// Social section -context
var social_style_config_context_1 = require("./context/social-style-config-context");
Object.defineProperty(exports, "SocialStyleConfigProvider", { enumerable: true, get: function () { return social_style_config_context_1.SocialStyleConfigProvider; } });
// Social section - hooks
var useSocialStyle_1 = require("./hooks/useSocialStyle");
Object.defineProperty(exports, "useSocialStyle", { enumerable: true, get: function () { return __importDefault(useSocialStyle_1).default; } });
var useSocialStyleConfig_1 = require("./hooks/useSocialStyleConfig");
Object.defineProperty(exports, "useSocialStyleConfig", { enumerable: true, get: function () { return __importDefault(useSocialStyleConfig_1).default; } });
// Social section - other
var social_base_style_1 = require("./social-base-style");
Object.defineProperty(exports, "socialBaseStyle", { enumerable: true, get: function () { return social_base_style_1.socialBaseStyle; } });
