// @replyke/comments-social-core/dist/cjs/social-base-style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.socialBaseStyle = void 0;
exports.socialBaseStyle = {
    type: "social",
    commentFeedProps: {
        backgroundColor: "#fff",
        commentsGap: 8,
    },
    commentProps: {
        horizontalItemsGap: 12,
        verticalItemsGap: 8,
        authorAvatarSize: 32,
        authorFontSize: 13,
        authorFontWeight: 700,
        authorFontColor: "#000",
        fromNowFontSize: 12,
        fromNowFontColor: "#737373",
        commentBodyFontSize: 14,
        commentBodyFontColor: "#000",
        actionsItemGap: 8,
        replyButtonFontSize: 12,
        replyButtonFontWeight: 600,
        replyButtonFontColor: "#737373",
        heartIconSize: 14,
        heartIconFullColor: "#DC2626",
        heartIconEmptyColor: "#8E8E8E",
        heartIconPaddingBottom: 4,
        likesCountFontSize: 12,
        likesCountFontWeight: 400,
        likesCountFontColor: "#8E8E8E",
        viewRepliesPaddingTop: 8,
        repliesGap: 12,
        repliesPaddingTop: 8,
        viewMoreRepliesText: "View more replies ($count)",
        hideRepliesText: "Hide replies",
        justNowText: "Just now",
    },
    newCommentFormProps: {
        backgroundColor: "#fff",
        withAvatar: true,
        itemsGap: 8,
        verticalPadding: 8,
        paddingLeft: 8,
        paddingRight: 16,
        authorAvatarSize: 32,
        placeholderText: "Add a comment...",
        textareaTextSize: 14,
        textareaBackgroundColor: "#fff",
        textareaTextColor: "#000",
        postButtonText: "Post",
        postButtonFontSize: 14,
        postButtonFontWeight: 600,
        postButtonFontColor: "#0A99F6",
    },
};
