// @replyke/comments-threaded-react-js/dist/cjs/components/CommentsFeed/CommentThread/SingleComment/IndentationThreadingLines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const comments_threaded_core_1 = require("@replyke/comments-threaded-core");
function IndentationThreadingLines({ isLastReply }) {
    const { styleConfig } = (0, comments_threaded_core_1.useThreadedStyleConfig)();
    const { threadingLineColor } = styleConfig.commentProps;
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [!isLastReply && ((0, jsx_runtime_1.jsx)("div", { style: {
                    position: "absolute",
                    width: "1px",
                    backgroundColor: threadingLineColor,
                    zIndex: 0,
                    left: "-12px",
                    top: "0px",
                    height: "100%",
                } })), (0, jsx_runtime_1.jsx)("div", { style: {
                    position: "absolute",
                    top: "-8px",
                    width: "1px",
                    height: "18px",
                    backgroundColor: threadingLineColor,
                    zIndex: 0,
                    left: "-12px",
                } }), (0, jsx_runtime_1.jsx)("div", { style: {
                    position: "absolute",
                    top: "10px",
                    width: "12px",
                    height: "16px",
                    borderLeft: `1.5px solid ${threadingLineColor}`,
                    borderBottom: `1.5px solid ${threadingLineColor}`,
                    borderTop: "1.5px solid transparent",
                    borderRight: "1.5px solid transparent",
                    borderBottomLeftRadius: "12px",
                    zIndex: 0,
                    left: "-12px",
                } })] }));
}
exports.default = IndentationThreadingLines;
