// @replyke/comments-threaded-react-js/dist/cjs/components/CommentsFeed/CommentThread/SingleComment/ReplyButtonAndInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const comments_threaded_core_1 = require("@replyke/comments-threaded-core");
function ReplyButtonAndInfo({ hasReplies, replyCount, setShowReplyForm, }) {
    const { styleConfig } = (0, comments_threaded_core_1.useThreadedStyleConfig)();
    const { replyButtonFontSize, replyButtonFontWeight, replyButtonFontColor, actionsItemGap, fromNowFontColor, } = styleConfig.commentProps;
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            display: "flex",
            alignItems: "center",
            gap: `${actionsItemGap}px`,
            fontSize: `${replyButtonFontSize}px`,
        }, children: [(0, jsx_runtime_1.jsx)("button", { onClick: () => setShowReplyForm((prev) => !prev), style: {
                    color: replyButtonFontColor,
                    fontWeight: replyButtonFontWeight,
                    padding: "4px 8px",
                    borderRadius: "4px",
                    marginLeft: "-8px",
                    transition: "all 150ms ease-in-out",
                    background: "transparent",
                    border: "none",
                    cursor: "pointer",
                }, onMouseEnter: (e) => {
                    e.currentTarget.style.color = "#2563EB";
                    e.currentTarget.style.backgroundColor = "#EFF6FF";
                }, onMouseLeave: (e) => {
                    e.currentTarget.style.color = "#6B7280";
                    e.currentTarget.style.backgroundColor = "transparent";
                }, children: "Reply" }), hasReplies && ((0, jsx_runtime_1.jsxs)("span", { style: { color: fromNowFontColor }, children: [replyCount, " ", replyCount === 1 ? "reply" : "replies"] }))] }));
}
exports.default = ReplyButtonAndInfo;
