// @replyke/comments-threaded-react-js/dist/cjs/components/CommentsFeed/CommentThread/SingleComment/SingleComment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_js_1 = require("@replyke/react-js");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const comments_threaded_core_1 = require("@replyke/comments-threaded-core");
const VoteButtons_1 = __importDefault(require("./VoteButtons"));
const ActionMenu_1 = __importDefault(require("../ActionMenu"));
const NewReplyForm_1 = __importDefault(require("../NewReplyForm"));
const ToggleRepliesVisibilty_1 = __importDefault(require("./ToggleRepliesVisibilty"));
const IndentationThreadingLines_1 = __importDefault(require("./IndentationThreadingLines"));
const ReplyButtonAndInfo_1 = __importDefault(require("./ReplyButtonAndInfo"));
function SingleComment({ comment: commentFromSection, depth, hasReplies, isCollapsed, replyCount, isLastReply = false, onToggleCollapse, }) {
    const { user } = (0, react_js_1.useUser)();
    const { callbacks, highlightedComment } = (0, react_js_1.useCommentSection)();
    const [comment, setComment] = (0, react_1.useState)(commentFromSection);
    const [showReplyForm, setShowReplyForm] = (0, react_1.useState)(false);
    const { styleConfig } = (0, comments_threaded_core_1.useThreadedStyleConfig)();
    const { authorAvatarSize, authorFontSize, authorFontWeight, authorFontColor, fromNowFontSize, fromNowFontColor, commentBodyFontSize, commentBodyFontColor, horizontalItemsGap, verticalItemsGap, threadingLineColor, } = styleConfig.commentProps;
    const maxDepth = 6; // Limit visual nesting depth
    const actualDepth = Math.min(depth, maxDepth);
    // Calculate progressive indentation using inline styles for reliability
    const indentationPx = actualDepth * 24; // 24px per level (same as ml-10 = 2.5rem = 40px)
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            position: "relative",
            // marginBottom: "8px",
            marginLeft: `${indentationPx}px`,
            backgroundColor: highlightedComment?.comment.id === comment.id
                ? "#dbeafe"
                : "transparent",
        }, children: [actualDepth > 0 && ((0, jsx_runtime_1.jsx)(IndentationThreadingLines_1.default, { isLastReply: isLastReply })), (0, jsx_runtime_1.jsx)("div", { style: {
                    padding: `${verticalItemsGap}px 0`,
                    borderRadius: "6px",
                    transition: "colors 150ms ease-in-out",
                }, children: (0, jsx_runtime_1.jsxs)("div", { style: { display: "flex" }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                flexShrink: 0,
                                marginRight: `${horizontalItemsGap}px`,
                                position: "relative",
                                marginTop: `${verticalItemsGap / 2}px`,
                            }, children: [(0, jsx_runtime_1.jsx)("div", { style: { position: "relative", zIndex: 10 }, children: (0, jsx_runtime_1.jsx)(ui_core_react_js_1.UserAvatar, { user: comment.user, borderRadius: authorAvatarSize, size: authorAvatarSize }) }), hasReplies && !isCollapsed && ((0, jsx_runtime_1.jsx)("div", { style: {
                                        position: "absolute",
                                        left: "50%",
                                        top: "20px",
                                        width: "1px",
                                        backgroundColor: threadingLineColor,
                                        zIndex: 0,
                                        height: "calc(100% + 10px)",
                                    } }))] }), (0, jsx_runtime_1.jsxs)("div", { style: { flex: 1, minWidth: 0 }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                        display: "flex",
                                        alignItems: "center",
                                        justifyContent: "space-between",
                                        marginBottom: `${verticalItemsGap / 2}px`,
                                    }, children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                gap: `${verticalItemsGap}px`,
                                                fontSize: `${fromNowFontSize}px`,
                                                color: fromNowFontColor,
                                            }, children: [(0, jsx_runtime_1.jsx)("span", { style: {
                                                        fontWeight: authorFontWeight,
                                                        fontSize: authorFontSize,
                                                        color: authorFontColor,
                                                    }, children: (0, react_js_1.getUserName)(comment.user) }), (0, jsx_runtime_1.jsx)("span", { children: "\u2022" }), (0, jsx_runtime_1.jsx)("span", { children: new Date(comment.createdAt).toLocaleDateString() }), isCollapsed && hasReplies && ((0, jsx_runtime_1.jsxs)("span", { className: "text-blue-600 text-xs", children: ["(", replyCount, " ", replyCount === 1 ? "reply" : "replies", ")"] })), hasReplies && ((0, jsx_runtime_1.jsx)(ToggleRepliesVisibilty_1.default, { isCollapsed: isCollapsed, onToggleCollapse: onToggleCollapse }))] }), (0, jsx_runtime_1.jsx)(ActionMenu_1.default, { comment: comment })] }), !isCollapsed && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [comment.content && ((0, jsx_runtime_1.jsx)("p", { style: {
                                                fontSize: `${commentBodyFontSize}px`,
                                                color: commentBodyFontColor,
                                                marginBottom: `${horizontalItemsGap}px`,
                                                lineHeight: "1.625",
                                            }, children: (0, ui_core_react_js_1.parseContentWithMentions)(comment.content, comment.mentions, user?.id, callbacks?.currentUserClickCallback, callbacks?.otherUserClickCallback) })), comment.gif && ((0, jsx_runtime_1.jsx)("img", { src: comment.gif.gifUrl, alt: comment.gif.altText, style: {
                                                width: comment.gif.aspectRatio > 1
                                                    ? 200
                                                    : 200 * comment.gif.aspectRatio,
                                                height: comment.gif.aspectRatio < 1
                                                    ? 200
                                                    : 200 / comment.gif.aspectRatio,
                                                borderRadius: "0.25rem",
                                                overflow: "hidden",
                                                objectFit: "cover",
                                                marginBottom: `${horizontalItemsGap}px`,
                                            } })), (0, jsx_runtime_1.jsxs)("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                justifyContent: "space-between",
                                            }, children: [(0, jsx_runtime_1.jsx)(ReplyButtonAndInfo_1.default, { hasReplies: hasReplies, replyCount: replyCount, setShowReplyForm: setShowReplyForm }), (0, jsx_runtime_1.jsx)("div", { style: { flexShrink: 0 }, children: (0, jsx_runtime_1.jsx)(VoteButtons_1.default, { comment: comment, setComment: setComment, size: "small" }) })] }), showReplyForm && ((0, jsx_runtime_1.jsx)(NewReplyForm_1.default, { comment: comment, setShowReplyForm: setShowReplyForm }))] }))] })] }) })] }));
}
exports.default = SingleComment;
