// @replyke/comments-threaded-react-js/dist/cjs/components/CommentsFeed/CommentThread/SingleComment/ToggleRepliesVisibilty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
function ToggleRepliesVisibilty({ isCollapsed, onToggleCollapse, }) {
    return ((0, jsx_runtime_1.jsx)("button", { onClick: onToggleCollapse, style: {
            marginLeft: "4px",
            width: "16px",
            height: "16px",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            color: "#6B7280",
            backgroundColor: "#F3F4F6",
            borderRadius: "2px",
            transition: "all 150ms ease-in-out",
            fontSize: "14px",
            fontWeight: "bold",
            border: isCollapsed ? "1px solid #D1D5DB" : "none",
            cursor: "pointer",
        }, onMouseEnter: (e) => {
            e.currentTarget.style.color = "#374151";
            e.currentTarget.style.backgroundColor = "#E5E7EB";
        }, onMouseLeave: (e) => {
            e.currentTarget.style.color = "#6B7280";
            e.currentTarget.style.backgroundColor = "#F3F4F6";
        }, title: isCollapsed ? "Expand thread" : "Collapse thread", children: isCollapsed ? "+" : "−" }));
}
exports.default = ToggleRepliesVisibilty;
