// @replyke/comments-thread-react-js/dist/esm/components/CommentsFeed/CommentThread/ActionMenu.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useUser } from "@replyke/react-js";
import { useState, useRef, useEffect } from "react";
function ActionMenu({ comment, onDelete, onReport }) {
    const { user } = useUser();
    const [isOpen, setIsOpen] = useState(false);
    const menuRef = useRef(null);
    const isAuthor = comment.user?.id === user?.id;
    useEffect(() => {
        function handleClickOutside(event) {
            if (menuRef.current && !menuRef.current.contains(event.target)) {
                setIsOpen(false);
            }
        }
        document.addEventListener("mousedown", handleClickOutside);
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, []);
    const handleAction = (action) => {
        if (action) {
            action();
        }
        setIsOpen(false);
    };
    return (_jsxs("div", { style: { position: "relative" }, ref: menuRef, children: [_jsx("button", { onClick: () => setIsOpen(!isOpen), style: {
                    padding: "2px",
                    color: "#9CA3AF",
                    transition: "colors 150ms ease-in-out",
                    borderRadius: "4px",
                    background: "transparent",
                    border: "none",
                    cursor: "pointer",
                }, onMouseEnter: (e) => {
                    e.currentTarget.style.color = "#4B5563";
                    e.currentTarget.style.backgroundColor = "#F3F4F6";
                }, onMouseLeave: (e) => {
                    e.currentTarget.style.color = "#9CA3AF";
                    e.currentTarget.style.backgroundColor = "transparent";
                }, title: "More options", children: _jsx("svg", { style: { width: "12px", height: "12px" }, fill: "currentColor", viewBox: "0 0 20 20", children: _jsx("path", { d: "M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z" }) }) }), isOpen && (_jsx("div", { style: {
                    position: "absolute",
                    right: 0,
                    top: "20px",
                    marginTop: "4px",
                    width: "96px",
                    backgroundColor: "#FFFFFF",
                    borderRadius: "6px",
                    boxShadow: "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)",
                    border: "1px solid #E5E7EB",
                    paddingTop: "4px",
                    paddingBottom: "4px",
                    zIndex: 10,
                }, children: isAuthor ? (_jsxs("button", { onClick: () => onDelete && handleAction(onDelete), style: {
                        width: "100%",
                        padding: "6px 8px",
                        textAlign: "left",
                        fontSize: "12px",
                        color: "#DC2626",
                        transition: "colors 150ms ease-in-out",
                        display: "flex",
                        alignItems: "center",
                        gap: "6px",
                        background: "transparent",
                        border: "none",
                        cursor: "pointer",
                    }, onMouseEnter: (e) => {
                        e.currentTarget.style.backgroundColor = "#FEF2F2";
                    }, onMouseLeave: (e) => {
                        e.currentTarget.style.backgroundColor = "transparent";
                    }, children: [_jsx("svg", { style: { width: "10px", height: "10px" }, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", children: _jsx("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16" }) }), "Delete"] })) : (_jsxs("button", { onClick: () => onReport && handleAction(onReport), style: {
                        width: "100%",
                        padding: "6px 8px",
                        textAlign: "left",
                        fontSize: "12px",
                        color: "#EA580C",
                        transition: "colors 150ms ease-in-out",
                        display: "flex",
                        alignItems: "center",
                        gap: "6px",
                        background: "transparent",
                        border: "none",
                        cursor: "pointer",
                    }, onMouseEnter: (e) => {
                        e.currentTarget.style.backgroundColor = "#FFF7ED";
                    }, onMouseLeave: (e) => {
                        e.currentTarget.style.backgroundColor = "transparent";
                    }, children: [_jsx("svg", { style: { width: "10px", height: "10px" }, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", children: _jsx("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.732-.833-2.464 0L4.35 16.5c-.77.833.192 2.5 1.732 2.5z" }) }), "Report"] })) }))] }));
}
export default ActionMenu;
