// @replyke/comments-thread-react-js/dist/esm/components/CommentsFeed/CommentThread/SingleComment.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { useCommentSection } from "@replyke/react-js";
import { UserAvatar } from "@replyke/ui-core-react-js";
import VoteButtons from "./VoteButtons";
import ActionMenu from "./ActionMenu";
import NewReplyForm from "./NewReplyForm";
function SingleComment({ comment: commentFromSection, depth, hasReplies, isCollapsed, replyCount, isLastReply = false, onToggleCollapse, }) {
    const [comment, setComment] = useState(commentFromSection);
    const { deleteComment } = useCommentSection();
    const [showReplyForm, setShowReplyForm] = useState(false);
    const maxDepth = 6; // Limit visual nesting depth
    const actualDepth = Math.min(depth, maxDepth);
    // Calculate progressive indentation using inline styles for reliability
    const indentationPx = actualDepth * 24; // 40px per level (same as ml-10 = 2.5rem = 40px)
    return (_jsxs("div", { style: {
            position: "relative",
            marginBottom: "8px",
            marginLeft: `${indentationPx}px`,
        }, children: [actualDepth > 0 && (_jsxs(_Fragment, { children: [!isLastReply && (_jsx("div", { style: {
                            position: "absolute",
                            width: "1px",
                            backgroundColor: "#D1D5DB",
                            zIndex: 0,
                            left: "-12px",
                            top: "0px",
                            height: "100%",
                        } })), _jsx("div", { style: {
                            position: "absolute",
                            top: "-8px",
                            width: "1px",
                            height: "18px",
                            backgroundColor: "#D1D5DB",
                            zIndex: 0,
                            left: "-12px",
                        } }), _jsx("div", { style: {
                            position: "absolute",
                            top: "10px",
                            width: "12px",
                            height: "16px",
                            borderLeft: "1.5px solid #D1D5DB",
                            borderBottom: "1.5px solid #D1D5DB",
                            borderTop: "1.5px solid transparent",
                            borderRight: "1.5px solid transparent",
                            borderBottomLeftRadius: "12px",
                            zIndex: 0,
                            left: "-12px",
                        } })] })), _jsx("div", { style: {
                    padding: "8px 0",
                    borderRadius: "6px",
                    transition: "colors 150ms ease-in-out",
                }, children: _jsxs("div", { style: { display: "flex" }, children: [_jsxs("div", { style: {
                                flexShrink: 0,
                                marginRight: "12px",
                                position: "relative",
                                marginTop: "4px",
                            }, children: [_jsx("div", { style: { position: "relative", zIndex: 10 }, children: _jsx(UserAvatar, { user: comment.user, borderRadius: 24, size: 24 }) }), hasReplies && !isCollapsed && (_jsx("div", { style: {
                                        position: "absolute",
                                        left: "50%",
                                        top: "20px",
                                        width: "1px",
                                        backgroundColor: "#D1D5DB",
                                        zIndex: 0,
                                        height: "calc(100% + 10px)",
                                    } }))] }), _jsxs("div", { style: { flex: 1, minWidth: 0 }, children: [_jsxs("div", { style: {
                                        display: "flex",
                                        alignItems: "center",
                                        justifyContent: "space-between",
                                        marginBottom: "4px",
                                    }, children: [_jsxs("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                gap: "8px",
                                                fontSize: "12px",
                                                color: "#6B7280",
                                            }, children: [_jsx("span", { style: { fontWeight: "500", color: "#374151" }, children: comment.user?.name || "Anonymous" }), _jsx("span", { children: "\u2022" }), _jsx("span", { children: new Date(comment.createdAt).toLocaleDateString() }), hasReplies && (_jsx("button", { onClick: onToggleCollapse, style: {
                                                        marginLeft: "4px",
                                                        width: "16px",
                                                        height: "16px",
                                                        display: "flex",
                                                        alignItems: "center",
                                                        justifyContent: "center",
                                                        color: "#6B7280",
                                                        backgroundColor: "#F3F4F6",
                                                        borderRadius: "2px",
                                                        transition: "all 150ms ease-in-out",
                                                        fontSize: "14px",
                                                        fontWeight: "bold",
                                                        border: isCollapsed ? "1px solid #D1D5DB" : "none",
                                                        cursor: "pointer",
                                                    }, onMouseEnter: (e) => {
                                                        e.currentTarget.style.color = "#374151";
                                                        e.currentTarget.style.backgroundColor = "#E5E7EB";
                                                    }, onMouseLeave: (e) => {
                                                        e.currentTarget.style.color = "#6B7280";
                                                        e.currentTarget.style.backgroundColor = "#F3F4F6";
                                                    }, title: isCollapsed ? "Expand thread" : "Collapse thread", children: isCollapsed ? "+" : "−" }))] }), _jsx(ActionMenu, { comment: comment, onDelete: () => {
                                                deleteComment?.({ commentId: comment.id });
                                            } })] }), !isCollapsed && (_jsxs(_Fragment, { children: [_jsx("p", { style: {
                                                fontSize: "12px",
                                                color: "#1F2937",
                                                marginBottom: "12px",
                                                lineHeight: "1.625",
                                            }, children: comment.content }), _jsxs("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                justifyContent: "space-between",
                                            }, children: [_jsxs("div", { style: {
                                                        display: "flex",
                                                        alignItems: "center",
                                                        gap: "16px",
                                                        fontSize: "12px",
                                                    }, children: [_jsx("button", { onClick: () => setShowReplyForm((prev) => !prev), style: {
                                                                color: "#6B7280",
                                                                fontWeight: "500",
                                                                padding: "4px 8px",
                                                                borderRadius: "4px",
                                                                marginLeft: "-8px",
                                                                transition: "all 150ms ease-in-out",
                                                                background: "transparent",
                                                                border: "none",
                                                                cursor: "pointer",
                                                            }, onMouseEnter: (e) => {
                                                                e.currentTarget.style.color = "#2563EB";
                                                                e.currentTarget.style.backgroundColor = "#EFF6FF";
                                                            }, onMouseLeave: (e) => {
                                                                e.currentTarget.style.color = "#6B7280";
                                                                e.currentTarget.style.backgroundColor = "transparent";
                                                            }, children: "Reply" }), hasReplies && (_jsxs("span", { style: { color: "#9CA3AF" }, children: [replyCount, " ", replyCount === 1 ? "reply" : "replies"] }))] }), _jsx("div", { style: { flexShrink: 0 }, children: _jsx(VoteButtons, { comment: comment, setComment: setComment, size: "small" }) })] }), showReplyForm && (_jsx(NewReplyForm, { comment: comment, setShowReplyForm: setShowReplyForm }))] }))] })] }) })] }));
}
export default SingleComment;
