// @replyke/comments-thread-react-js/dist/esm/components/CommentsFeed/CommentThread/VoteButtons.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useCommentSection, useUser, useCommentVotes, } from "@replyke/react-js";
function VoteButtons({ comment, setComment, size = "small", }) {
    const { user } = useUser();
    const { callbacks } = useCommentSection();
    const { upvoteComment, downvoteComment, removeCommentUpvote, removeCommentDownvote, } = useCommentVotes({
        comment,
        setComment,
    });
    const upvotes = comment.upvotes?.length || 0;
    const downvotes = comment.downvotes?.length || 0;
    const netScore = upvotes - downvotes;
    // Check if user has voted on this comment
    const userUpvotedComment = !!(user && comment.upvotes.includes(user.id));
    const userDownvotedComment = !!(user && comment.downvotes.includes(user.id));
    const userVote = userUpvotedComment
        ? "up"
        : userDownvotedComment
            ? "down"
            : null;
    const handleVote = (voteType) => {
        if (voteType === "up") {
            if (userUpvotedComment) {
                removeCommentUpvote?.();
            }
            else {
                upvoteComment?.();
            }
        }
        else {
            if (userDownvotedComment) {
                removeCommentDownvote?.();
            }
            else {
                downvoteComment?.();
            }
        }
    };
    const iconSize = size === "small"
        ? { width: "12px", height: "12px" }
        : { width: "16px", height: "16px" };
    const textSize = size === "small" ? "12px" : "14px";
    const padding = size === "small" ? "4px 8px" : "6px 12px";
    return (_jsxs("div", { style: {
            display: "inline-flex",
            alignItems: "center",
            backgroundColor: "#F3F4F6",
            borderRadius: "9999px",
            padding: padding,
            gap: "4px",
        }, children: [_jsx("button", { onClick: () => {
                    if (!user) {
                        callbacks?.loginRequiredCallback ||
                            (() => alert("Please login first."))();
                        return;
                    }
                    handleVote("up");
                }, style: {
                    padding: "4px",
                    borderRadius: "50%",
                    transition: "colors 150ms ease-in-out",
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "center",
                    backgroundColor: userVote === "up" ? "#F97316" : "transparent",
                    color: userVote === "up" ? "#FFFFFF" : "#6B7280",
                    border: "none",
                    cursor: "pointer",
                }, onMouseEnter: (e) => {
                    if (userVote !== "up") {
                        e.currentTarget.style.color = "#F97316";
                        e.currentTarget.style.backgroundColor = "#FFF7ED";
                    }
                }, onMouseLeave: (e) => {
                    if (userVote !== "up") {
                        e.currentTarget.style.color = "#6B7280";
                        e.currentTarget.style.backgroundColor = "transparent";
                    }
                }, title: "Upvote", children: _jsx("svg", { style: iconSize, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2, children: _jsx("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "m4.5 15.75 7.5-7.5 7.5 7.5" }) }) }), _jsx("span", { style: {
                    fontWeight: "500",
                    fontSize: textSize,
                    minWidth: "20px",
                    textAlign: "center",
                    color: netScore > 0 ? "#EA580C" : netScore < 0 ? "#2563EB" : "#374151",
                }, children: netScore }), _jsx("button", { onClick: () => {
                    if (!user) {
                        callbacks?.loginRequiredCallback ||
                            (() => alert("Please login first."))();
                        return;
                    }
                    handleVote("down");
                }, style: {
                    padding: "4px",
                    borderRadius: "50%",
                    transition: "colors 150ms ease-in-out",
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "center",
                    backgroundColor: userVote === "down" ? "#3B82F6" : "transparent",
                    color: userVote === "down" ? "#FFFFFF" : "#6B7280",
                    border: "none",
                    cursor: "pointer",
                }, onMouseEnter: (e) => {
                    if (userVote !== "down") {
                        e.currentTarget.style.color = "#3B82F6";
                        e.currentTarget.style.backgroundColor = "#EFF6FF";
                    }
                }, onMouseLeave: (e) => {
                    if (userVote !== "down") {
                        e.currentTarget.style.color = "#6B7280";
                        e.currentTarget.style.backgroundColor = "transparent";
                    }
                }, title: "Downvote", children: _jsx("svg", { style: iconSize, fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2, children: _jsx("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "m19.5 8.25-7.5 7.5-7.5-7.5" }) }) })] }));
}
export default VoteButtons;
