// @replyke/core/dist/cjs/hooks/app-notifications/useAppNotifications.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var appNotificationsSlice_1 = require("../../store/slices/appNotificationsSlice");
var useAppNotificationsActions_1 = require("./useAppNotificationsActions");
var useProject_1 = __importDefault(require("../projects/useProject"));
var user_1 = require("../user");
/**
 * Redux-powered hook that provides the exact same interface as useAppNotificationsData()
 * This is a drop-in replacement for the Context-based hook
 */
function useAppNotifications(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.limit, limit = _c === void 0 ? 10 : _c, notificationTemplates = _b.notificationTemplates;
    var dispatch = (0, react_redux_1.useDispatch)();
    // Get external context
    var projectId = (0, useProject_1.default)().projectId;
    var user = (0, user_1.useUser)().user;
    // Get Redux state
    var appNotifications = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectAppNotifications)(state);
    });
    var unreadAppNotificationsCount = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectUnreadCount)(state);
    });
    var loading = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectAppNotificationsLoading)(state);
    });
    var hasMore = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectAppNotificationsHasMore)(state);
    });
    var currentPage = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectAppNotificationsPage)(state);
    });
    var currentProjectId = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectCurrentProjectId)(state);
    });
    // Get actions
    var _d = (0, useAppNotificationsActions_1.useAppNotificationsActions)(), loadMore = _d.loadMore, markNotificationAsRead = _d.markNotificationAsRead, markAllNotificationsAsRead = _d.markAllNotificationsAsRead, resetAppNotifications = _d.resetAppNotifications, fetchMoreNotifications = _d.fetchMoreNotifications, updateUnreadCount = _d.updateUnreadCount;
    // Update Redux state when props change
    (0, react_1.useEffect)(function () {
        if (projectId && projectId !== currentProjectId) {
            dispatch((0, appNotificationsSlice_1.setProjectContext)(projectId));
        }
    }, [dispatch, projectId, currentProjectId]);
    (0, react_1.useEffect)(function () {
        dispatch((0, appNotificationsSlice_1.setLimit)(limit));
    }, [dispatch, limit]);
    // Prevent infinite re-renders by comparing current vs new templates
    var currentTemplates = (0, react_redux_1.useSelector)(function (state) {
        return (0, appNotificationsSlice_1.selectNotificationTemplates)(state);
    });
    var templatesChanged = (0, react_1.useMemo)(function () {
        // If no templates provided, skip comparison
        if (!notificationTemplates)
            return false;
        // Deep comparison using JSON stringify
        return (JSON.stringify(currentTemplates) !== JSON.stringify(notificationTemplates));
    }, [currentTemplates, notificationTemplates]);
    (0, react_1.useEffect)(function () {
        if (notificationTemplates && templatesChanged) {
            dispatch((0, appNotificationsSlice_1.setNotificationTemplates)(notificationTemplates));
        }
    }, [dispatch, notificationTemplates, templatesChanged]);
    // Fetch unread count on mount and when dependencies change
    (0, react_1.useEffect)(function () {
        if (projectId && user) {
            updateUnreadCount();
        }
    }, [updateUnreadCount, projectId, user]);
    // Reset and fetch initial notifications when dependencies change
    (0, react_1.useEffect)(function () {
        if (projectId && user) {
            resetAppNotifications();
        }
    }, [resetAppNotifications, projectId, user]);
    // Handle page changes (load more notifications)
    (0, react_1.useEffect)(function () {
        if (currentPage > 1 && projectId && user) {
            fetchMoreNotifications(currentPage);
        }
    }, [currentPage, fetchMoreNotifications, projectId, user]);
    // Return the same interface as the original hook
    return (0, react_1.useMemo)(function () { return ({
        appNotifications: appNotifications,
        unreadAppNotificationsCount: unreadAppNotificationsCount,
        loading: loading,
        hasMore: hasMore,
        loadMore: loadMore,
        markNotificationAsRead: markNotificationAsRead,
        markAllNotificationsAsRead: markAllNotificationsAsRead,
        resetAppNotifications: resetAppNotifications,
    }); }, [
        appNotifications,
        unreadAppNotificationsCount,
        loading,
        hasMore,
        loadMore,
        markNotificationAsRead,
        markAllNotificationsAsRead,
        resetAppNotifications,
    ]);
}
exports.default = useAppNotifications;
