// @replyke/comments-threaded-react-js/dist/cjs/components/modals/CommentMenuModal/MainContent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const ui_core_react_js_1 = require("@replyke/ui-core-react-js");
const useModalManager_1 = __importDefault(require("../../../hooks/useModalManager"));
function MainContent({ clickReport }) {
    const { closeCommentOptionsModal } = (0, useModalManager_1.default)();
    return ((0, jsx_runtime_1.jsxs)("ul", { style: {
            ...ui_core_react_js_1.resetUl,
            width: "100%",
        }, children: [(0, jsx_runtime_1.jsx)("li", { style: {
                    display: "flex",
                    justifyContent: "center",
                    justifySelf: "center",
                }, children: (0, jsx_runtime_1.jsx)("button", { style: {
                        ...ui_core_react_js_1.resetButton,
                        fontWeight: 600,
                        color: "#DC2626",
                        paddingLeft: 24,
                        paddingRight: 24,
                        paddingTop: 8,
                        paddingBottom: 8,
                    }, onClick: clickReport, children: "Report" }) }), (0, jsx_runtime_1.jsx)("div", { style: { height: 1, width: "100%", backgroundColor: "#e7e7e7" } }), (0, jsx_runtime_1.jsx)("li", { style: {
                    display: "flex",
                    justifyContent: "center",
                    justifySelf: "center",
                }, children: (0, jsx_runtime_1.jsx)("button", { style: {
                        ...ui_core_react_js_1.resetButton,
                        paddingLeft: 24,
                        paddingRight: 24,
                        paddingTop: 8,
                        paddingBottom: 8,
                    }, onClick: closeCommentOptionsModal, children: "Cancel" }) })] }));
}
exports.default = MainContent;
