// @giphy/react-components/dist/esm/chunk-S5ZL7V5V.js
import {
  pingback_context_manager_default
} from "./chunk-LFPSSP6P.js";
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-BG6MEPCE.js";

// src/components/search-bar/context.tsx
import { giphyBlack, giphyCharcoal, giphyWhite } from "@giphy/colors";
import { GiphyFetch, request } from "@giphy/js-fetch-api";
import { Logger } from "@giphy/js-util";
import React, { createContext, useCallback, useEffect, useMemo, useState } from "react";
import styled, { css } from "styled-components";
var CssVars = {
  searchbarHeight: `--searchbar-height`,
  bgColor: `--searchbar-bg-color`,
  bgColor2: `--searchbar-bg-color-2`,
  fgColor: `--searchbar-fg-color`,
  cancelButtonDisplay: `--searchbar-cancel-button-display`
};
var Container = styled.div`
    ${CssVars.searchbarHeight}: ${(props) => props.$searchbarHeight || 42}px;
    @media (${(props) => props.$mobileMediaQuery}) {
        ${CssVars.searchbarHeight}: ${(props) => props.$mobileSearchbarHeight || 35}px;
    }
    ${CssVars.bgColor}: ${giphyWhite};
    ${CssVars.bgColor2}: ${giphyWhite};
    ${CssVars.fgColor}: ${giphyBlack};
    ${(props) => props.$darkMode && css`
            ${CssVars.fgColor}: ${giphyWhite};
            ${CssVars.bgColor}: ${giphyBlack};
            ${CssVars.bgColor2}: ${giphyCharcoal};
        `}
    ${CssVars.cancelButtonDisplay}: ${(props) => props.$hideCancelButton ? "none" : "block"};
`;
var SearchContext = createContext({});
var _SearchContext = createContext({});
var emptyChannels = [];
var emptyGifsResult = {
  data: [],
  pagination: { total_count: 0, count: 0, offset: 0 },
  meta: { status: 200, msg: "OK", response_id: "" }
};
var SearchContextManager = ({
  children,
  options = {},
  apiKey,
  initialTerm = "",
  initialChannel,
  shouldDefaultToTrending = true,
  shouldFetchChannels = true,
  theme
}) => {
  var _a, _b;
  const gf = useMemo(() => new GiphyFetch(apiKey), [apiKey]);
  const [currentChannels, setChannels] = useState([]);
  const [term, _setSearch] = useState(initialTerm);
  const [_inputValOverride, setSearch] = useState(initialTerm);
  const [isFetching, setIsFetching] = useState(false);
  const [activeChannel, _setActiveChannel] = useState(initialChannel);
  const setActiveChannel = useCallback((activeChannel2) => {
    _setSearch("");
    _setActiveChannel(activeChannel2);
  }, []);
  const [trendingSearches, setTrendingSearches] = useState([]);
  const [isFocused, setIsFocused] = useState(false);
  const searchKey = [term, options.type, ((_a = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _a.username) || ""].filter((val) => !!val).join(" / ");
  const fetchGifs = useCallback(
    (offset) => __async(void 0, null, function* () {
      var _a2;
      setIsFetching(true);
      let result = emptyGifsResult;
      if (term) {
        result = yield gf.search(term, __spreadProps(__spreadValues({}, options), {
          offset,
          channel: (_a2 = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _a2.username
        }));
      } else if (shouldDefaultToTrending) {
        result = yield gf.trending(__spreadProps(__spreadValues({}, options), { offset }));
      }
      setIsFetching(false);
      return result;
    }),
    [(_b = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _b.username, gf, options, term, shouldDefaultToTrending]
  );
  const fetchAnimatedText = useCallback(
    (offset) => __async(void 0, null, function* () {
      const limit = options.limit || 50;
      const result = yield gf.animate(term, { offset, limit });
      if (!result.pagination) {
        result.pagination = { count: limit, total_count: limit, offset };
      }
      return result;
    }),
    [gf, options.limit, term]
  );
  const fetchChannelSearch = useCallback(
    (offset) => __async(void 0, null, function* () {
      const search = term.indexOf("@") === 0 ? term.split(" ")[0] : term;
      const result = yield gf.channels(search, { offset });
      return result.data;
    }),
    [gf, term]
  );
  useEffect(() => {
    const fetchTrendingSearches = () => __async(void 0, null, function* () {
      let searches = [];
      try {
        const { data = [] } = yield request(`trending/searches?api_key=${apiKey}`);
        searches = data;
      } catch (error) {
        Logger.warn(`Trending searches request failed: ${error}`);
      }
      setTrendingSearches(searches);
    });
    fetchTrendingSearches();
  }, [apiKey]);
  useEffect(() => {
    const foundChannel = currentChannels.find(({ slug }) => term.indexOf(`@${slug} `) === 0);
    if (foundChannel) {
      _setActiveChannel(foundChannel);
    }
  }, [term, currentChannels, _setActiveChannel]);
  useEffect(() => {
    if (shouldFetchChannels) {
      const fetchChannels = () => __async(void 0, null, function* () {
        const channels = yield fetchChannelSearch(0);
        setChannels(channels || emptyChannels);
      });
      if (!activeChannel && term.replace("@", "")) {
        fetchChannels();
      }
    }
  }, [term, activeChannel]);
  return /* @__PURE__ */ React.createElement(
    SearchContext.Provider,
    {
      value: {
        activeChannel,
        setChannels,
        currentChannels,
        setActiveChannel,
        fetchChannelSearch,
        term,
        trendingSearches,
        setSearch,
        fetchGifs,
        fetchAnimatedText,
        searchKey,
        isFetching,
        isFocused
      }
    },
    /* @__PURE__ */ React.createElement(_SearchContext.Provider, { value: { setIsFocused, _setSearch, _inputValOverride } }, /* @__PURE__ */ React.createElement(
      Container,
      {
        $darkMode: theme == null ? void 0 : theme.darkMode,
        $searchbarHeight: theme == null ? void 0 : theme.searchbarHeight,
        $hideCancelButton: theme == null ? void 0 : theme.hideCancelButton,
        $mobileSearchbarHeight: theme == null ? void 0 : theme.mobileSearchbarHeight,
        $mobileMediaQuery: (theme == null ? void 0 : theme.mobileMediaQuery) || "max-width: 480px"
      },
      /* @__PURE__ */ React.createElement(pingback_context_manager_default, { attributes: { layout_type: "SEARCH" } }, children)
    ))
  );
};
var context_default = SearchContextManager;

export {
  CssVars,
  SearchContext,
  _SearchContext,
  context_default
};
