// @giphy/react-components/dist/esm/chunk-VTXVWGQH.js
import {
  CssVars,
  SearchContext
} from "./chunk-S5ZL7V5V.js";
import {
  avatar_default
} from "./chunk-Y2GBD252.js";
import {
  verified_badge_default
} from "./chunk-QKMOCXMK.js";

// src/components/search-bar/search-bar-channel.tsx
import { giphyDarkCharcoal, giphyWhiteSmoke } from "@giphy/colors";
import React, { useContext } from "react";
import styled, { keyframes } from "styled-components";
var channelMargin = 6;
var innerHeight = `calc(var(${CssVars.searchbarHeight}) - ${channelMargin * 2}px)`;
var animateAvatar = keyframes`
to {
    width: ${innerHeight};
}
`;
var Avatar = styled(avatar_default)`
    height: ${innerHeight};
    margin: 0;
    width: 0;
    animation: ${animateAvatar} 100ms ease-in-out forwards;
`;
var Username = styled.div`
    background: var(${CssVars.bgColor2});
    display: flex;
    align-items: center;
    padding-left: ${channelMargin}px;
`;
var UsernamePill = styled.div`
    background: ${giphyWhiteSmoke};
    display: flex;
    padding: 0 4px;
    color: ${giphyDarkCharcoal};
    font-family: interface, Helvetica Neue, helvetica, sans-serif;
    font-weight: 600;
    font-size: 12px;
    align-items: center;
    height: ${innerHeight};
    @media (max-width: 480px) {
        display: none;
    }
`;
var SearchBarChannel = ({ className = "" }) => {
  const { activeChannel } = useContext(SearchContext);
  return activeChannel ? /* @__PURE__ */ React.createElement(Username, { className }, /* @__PURE__ */ React.createElement(Avatar, { user: activeChannel.user }), /* @__PURE__ */ React.createElement(UsernamePill, { key: activeChannel.id }, /* @__PURE__ */ React.createElement("div", null, "@", activeChannel.user.username), activeChannel.user.is_verified && /* @__PURE__ */ React.createElement(verified_badge_default, { size: 14 }))) : null;
};
var search_bar_channel_default = SearchBarChannel;

export {
  search_bar_channel_default
};
