// @giphy/react-components/dist/esm/chunk-VTHTAQVK.js
// src/components/loader.tsx
import { giphyBlue, giphyGreen, giphyPurple, giphyRed, giphyYellow } from "@giphy/colors";
import React from "react";
import styled, { keyframes } from "styled-components";
var bouncer = keyframes`
     to {
    transform: scale(1.75) translateY(-20px);
  }
`;
var loaderHeight = 37;
var Container = styled.div`
    display: flex;
    align-items: center;
    height: ${loaderHeight}px;
    padding-top: 15px;
    margin: 0 auto;
    text-align: center;
    justify-content: center;
    animation: pulse 0.8s ease-in-out 0s infinite alternate backwards;
`;
var Dot = styled.div`
    display: inline-block;
    height: 10px;
    width: 10px;
    margin: ${loaderHeight}px 10px 10px 10px;
    position: relative;
    box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.3);
    animation: ${bouncer} cubic-bezier(0.455, 0.03, 0.515, 0.955) 0.75s infinite alternate;
    background: ${(props) => props.$color};
    animation-delay: ${(props) => props.$delay};
`;
var Loader = ({ className = "" }) => /* @__PURE__ */ React.createElement(Container, { className }, /* @__PURE__ */ React.createElement(Dot, { $color: giphyGreen, $delay: "0" }), /* @__PURE__ */ React.createElement(Dot, { $color: giphyBlue, $delay: ".1s" }), /* @__PURE__ */ React.createElement(Dot, { $color: giphyPurple, $delay: ".2s" }), /* @__PURE__ */ React.createElement(Dot, { $color: giphyRed, $delay: ".3s" }), /* @__PURE__ */ React.createElement(Dot, { $color: giphyYellow, $delay: ".4s" }));
var loader_default = Loader;

export {
  loader_default
};
