// @giphy/react-components/dist/esm/chunk-TDFQDVQT.js
import {
  fetch_error_default
} from "./chunk-76BUZ77K.js";
import {
  gif_default
} from "./chunk-Y3VZKJV7.js";
import {
  __async,
  __objRest,
  __spreadProps,
  __spreadValues
} from "./chunk-BG6MEPCE.js";

// src/components/emoji-variations-list.tsx
import { giphyCharcoal, giphyDarkGrey } from "@giphy/colors";
import { getGifWidth } from "@giphy/js-util";
import React, { useEffect } from "react";
import useAsyncFn from "react-use/lib/useAsyncFn";
import useLatest from "react-use/lib/useLatest";
import styled from "styled-components";
var DEFAULT_BG_COLOR = giphyDarkGrey;
var DEFAULT_DIVIDER_COLOR = giphyCharcoal;
var DIVIDER_RELATIVE_HEIGHT = 0.75;
var Root = styled.div`
    align-items: center;
    background-color: ${({ $backgroundColor }) => $backgroundColor};
    border-radius: 16px;
    display: flex;
    max-width: fit-content;
    overflow: hidden;
    padding: 4px 5px 5px 6px;
`;
var VariationsViewport = styled.div`
    overflow-x: auto;
    overflow-y: hidden;
    position: relative;
    -webkit-overflow-scrolling: touch;
`;
var VariationsContainer = styled.div`
    display: inline-flex;
    justify-content: space-between;
    overflow: hidden;
    white-space: nowrap;
    width: ${({ $width }) => `${$width}px`};
`;
var Divider = styled.div`
    background-color: ${({ $color: color }) => color};
    box-sizing: border-box;
    height: ${({ $gifHeight }) => `${Math.round($gifHeight * DIVIDER_RELATIVE_HEIGHT)}px`};
    margin: ${({ $gutter }) => `0 ${$gutter}px`};
    width: 2px;
`;
var Emoji = styled(gif_default)`
    display: inline-block;
    flex-shrink: 0;
    list-style: none;
    /* make sure gifs are fully visible with a scrollbar */
    margin-bottom: 1px;
    position: relative;

    .${gif_default.imgClassName} {
        position: absolute;
        top: 0;
        left: 0;
    }
`;
function EmojiVariationsList(props) {
  var _b;
  const _a = props, {
    backgroundColor = DEFAULT_BG_COLOR,
    className,
    dividerColor = DEFAULT_DIVIDER_COLOR,
    fetchVariations,
    gif,
    gifHeight,
    GifProps = {},
    gifWidth: gifWidthProp,
    gutter = 6,
    hideAttribution,
    loader: LoaderComponent,
    noLink,
    onGifClick,
    onGifKeyPress,
    onGifRightClick,
    onGifSeen,
    onGifVisible,
    onVariationsFetched,
    overlay,
    tabIndex = 0,
    user
  } = _a, other = __objRest(_a, [
    "backgroundColor",
    "className",
    "dividerColor",
    "fetchVariations",
    "gif",
    "gifHeight",
    "GifProps",
    "gifWidth",
    "gutter",
    "hideAttribution",
    "loader",
    "noLink",
    "onGifClick",
    "onGifKeyPress",
    "onGifRightClick",
    "onGifSeen",
    "onGifVisible",
    "onVariationsFetched",
    "overlay",
    "tabIndex",
    "user"
  ]);
  const variationCount = (_b = gif.variation_count) != null ? _b : 0;
  const gifWidth = gifWidthProp || getGifWidth(gif, gifHeight);
  const variationsContainerWidth = variationCount * gifWidth + Math.max(0, variationCount - 1) * gutter;
  const fetchVariationsRef = useLatest(fetchVariations);
  const onVariationsFetchedRef = useLatest(onVariationsFetched);
  const [variationsState, syncVariations] = useAsyncFn(
    () => __async(this, null, function* () {
      var _a2;
      const result = yield fetchVariationsRef.current(gif.id);
      const variations = (_a2 = result == null ? void 0 : result.data) != null ? _a2 : [];
      if (onVariationsFetchedRef.current) {
        onVariationsFetchedRef.current(variations);
      }
      return variations;
    }),
    [gif],
    { loading: true, value: [] }
  );
  useEffect(() => {
    syncVariations();
  }, [syncVariations]);
  const emojiCommonProps = __spreadValues({
    backgroundColor: "transparent",
    height: gifHeight,
    hideAttribution,
    noLink,
    onGifClick,
    onGifKeyPress,
    onGifRightClick,
    onGifSeen,
    onGifVisible,
    overlay,
    tabIndex,
    user,
    width: gifWidth
  }, GifProps);
  const getVariationsContent = () => {
    var _a2;
    if (variationsState.loading) {
      return LoaderComponent ? /* @__PURE__ */ React.createElement(LoaderComponent, { className: EmojiVariationsList.loaderClassName }) : null;
    }
    if (variationsState.error) {
      return /* @__PURE__ */ React.createElement(fetch_error_default, { onClick: syncVariations });
    }
    return /* @__PURE__ */ React.createElement(React.Fragment, null, ((_a2 = variationsState.value) != null ? _a2 : []).map((gif2) => /* @__PURE__ */ React.createElement(Emoji, __spreadProps(__spreadValues({}, emojiCommonProps), { key: gif2.id, gif: gif2 }))));
  };
  return /* @__PURE__ */ React.createElement(
    Root,
    __spreadValues({
      $backgroundColor: backgroundColor,
      className: [EmojiVariationsList.className, className].join(" ")
    }, other),
    /* @__PURE__ */ React.createElement(Emoji, __spreadProps(__spreadValues({}, emojiCommonProps), { gif })),
    variationCount ? /* @__PURE__ */ React.createElement(
      Divider,
      {
        className: EmojiVariationsList.dividerClassName,
        $color: dividerColor,
        $gifHeight: gifHeight,
        $gutter: gutter
      }
    ) : null,
    /* @__PURE__ */ React.createElement(VariationsViewport, { className: EmojiVariationsList.variationsViewportClassName }, /* @__PURE__ */ React.createElement(
      VariationsContainer,
      {
        className: EmojiVariationsList.variationsContainerClassName,
        $width: variationsContainerWidth
      },
      getVariationsContent()
    ))
  );
}
EmojiVariationsList.className = "giphy-emoji-variations-list";
EmojiVariationsList.dividerClassName = "giphy-emoji-variations-divider";
EmojiVariationsList.loaderClassName = "giphy-emoji-variations-loader";
EmojiVariationsList.variationsViewportClassName = "giphy-emoji-variations-viewport";
EmojiVariationsList.variationsContainerClassName = "giphy-emoji-variations-container";

export {
  EmojiVariationsList
};
