// @giphy/react-components/dist/esm/chunk-JLX5DGZ4.js
import {
  search_button_default
} from "./chunk-GJFSGHO5.js";
import {
  cancel_icon_default
} from "./chunk-EKRZRQR4.js";
import {
  search_bar_channel_default
} from "./chunk-VTXVWGQH.js";
import {
  CssVars,
  SearchContext,
  _SearchContext
} from "./chunk-S5ZL7V5V.js";

// src/components/search-bar/index.tsx
import { giphyIndigo, giphyLightGrey } from "@giphy/colors";
import React, { useContext, useEffect, useRef, useState } from "react";
import useDebounce from "react-use/lib/useDebounce";
import styled, { css } from "styled-components";
function usePrevious(value) {
  const ref = useRef(value);
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}
var SEARCH_DEBOUNCE = 500;
var Container = styled.div`
    display: flex;
    background: white;
    align-items: center;
    border-radius: 4px;
    height: var(${CssVars.searchbarHeight});
    background: var(${CssVars.bgColor2});
`;
var Input = styled.input`
    background: inherit;
    box-sizing: border-box;
    border: 0;
    appearance: none;
    font-weight: normal;
    font-family: interface, Helvetica Neue, helvetica, sans-serif;
    outline: 0;
    font-size: 15px;
    padding: 0 10px;
    border-radius: 0;
    text-overflow: ellipsis;
    color: var(${CssVars.fgColor});
    &::placeholder {
        color: ${giphyLightGrey};
    }
    min-width: 150px;
    flex: 1;
    ${(props) => props.$isUsernameSearch && css`
            color: ${giphyIndigo};
        `}
`;
var SearchBar = ({
  className,
  placeholder = `Search GIPHY`,
  clear = false,
  autoFocus,
  searchDebounce = SEARCH_DEBOUNCE,
  onEnter
}) => {
  const { activeChannel, setActiveChannel, term, setChannels } = useContext(SearchContext);
  const { setIsFocused, _inputValOverride, _setSearch } = useContext(_SearchContext);
  const [val, setVal] = useState(term);
  useDebounce(() => _setSearch(val), searchDebounce, [val]);
  const inputRef = useRef(null);
  const previousActiveChannel = usePrevious(activeChannel);
  const [isCleared, setCleared] = useState(clear);
  useEffect(() => {
    var _a;
    if (autoFocus) {
      (_a = inputRef.current) == null ? void 0 : _a.focus();
    }
  }, [autoFocus]);
  useEffect(() => {
    var _a;
    if (activeChannel && !previousActiveChannel) {
      (_a = inputRef.current) == null ? void 0 : _a.focus();
      if (val === " ") {
        setVal("");
      } else {
        setVal(val.replace(/@?\w*\s?/, ""));
      }
    }
  }, [val, activeChannel, previousActiveChannel]);
  useEffect(() => {
    setCleared(clear);
  }, [clear]);
  useEffect(() => {
    setVal(_inputValOverride);
  }, [_inputValOverride, setVal]);
  const onKeyUp = (e) => {
    const key = e.keyCode || e.key;
    switch (key) {
      case 27:
      case `Escape`:
        setActiveChannel(void 0);
        setChannels([]);
        break;
      case 13:
      case `Enter`:
        onEnter == null ? void 0 : onEnter(val);
        break;
      default:
        break;
    }
  };
  const onKeyDown = (e) => {
    const key = e.keyCode || e.key;
    switch (key) {
      case 8:
      case "Backspace":
        if (val === "") {
          setActiveChannel(void 0);
          setChannels([]);
        }
        break;
      default:
        break;
    }
  };
  return /* @__PURE__ */ React.createElement(Container, { className: [SearchBar.className, className].join(" ") }, /* @__PURE__ */ React.createElement(search_bar_channel_default, null), /* @__PURE__ */ React.createElement(
    Input,
    {
      $isUsernameSearch: term.indexOf("@") === 0,
      onChange: ({ target: { value } }) => {
        if (!isCleared || value !== "") {
          setCleared(false);
          setVal(value);
        }
      },
      onFocus: () => {
        setIsFocused(true);
      },
      onBlur: () => {
        setIsFocused(false);
      },
      value: isCleared ? "" : val,
      placeholder: activeChannel ? `Search ${activeChannel.display_name}` : placeholder,
      autoCapitalize: "off",
      autoCorrect: "off",
      autoComplete: "off",
      ref: inputRef,
      onKeyUp,
      onKeyDown
    }
  ), /* @__PURE__ */ React.createElement(cancel_icon_default, { setCleared: () => setVal("") }), /* @__PURE__ */ React.createElement(search_button_default, { onClick: () => onEnter == null ? void 0 : onEnter(val) }));
};
SearchBar.className = "giphy-search-bar";
var search_bar_default = SearchBar;

export {
  search_bar_default
};
