// @giphy/react-components/dist/esm/chunk-X27LNJ5O.js
import {
  observer_default
} from "./chunk-HMEPFNB4.js";
import {
  fetch_error_default
} from "./chunk-76BUZ77K.js";
import {
  gif_default
} from "./chunk-Y3VZKJV7.js";
import {
  pingback_context_manager_default
} from "./chunk-LFPSSP6P.js";
import {
  loader_default
} from "./chunk-VTHTAQVK.js";
import {
  masonry_grid_default
} from "./chunk-EP635Z2G.js";
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-BG6MEPCE.js";

// src/components/grid.tsx
import { gifPaginator } from "@giphy/js-fetch-api";
import { getGifHeight } from "@giphy/js-util";
import React, { PureComponent } from "react";
import styled from "styled-components";
import { debounce } from "throttle-debounce";
var Loader = styled.div`
    opacity: ${(props) => props.$isFirstLoad ? 0 : 1};
`;
var defaultProps = Object.freeze({ gutter: 6, user: {}, initialGifs: [] });
var initialState = Object.freeze({
  isFetching: false,
  isError: false,
  gifWidth: 0,
  gifs: [],
  isLoaderVisible: false,
  isDoneFetching: false
});
var _Grid = class _Grid extends PureComponent {
  constructor() {
    super(...arguments);
    this.state = __spreadProps(__spreadValues({}, initialState), { gifs: this.props.initialGifs || [] });
    this.unmounted = false;
    this.paginator = gifPaginator(this.props.fetchGifs, this.state.gifs);
    this.onLoaderVisible = (isVisible) => {
      if (this.unmounted)
        return;
      this.setState({ isLoaderVisible: isVisible }, this.onFetch);
    };
    this.onFetch = debounce(_Grid.fetchDebounce, () => __async(this, null, function* () {
      if (this.unmounted)
        return;
      const { isFetching, isLoaderVisible, gifs } = this.state;
      const prefetchCount = gifs.length;
      if (!isFetching && isLoaderVisible) {
        this.setState({ isFetching: true, isError: false });
        let gifs2;
        try {
          gifs2 = yield this.paginator();
          if (this.unmounted)
            return;
        } catch (error) {
          if (this.unmounted)
            return;
          this.setState({ isFetching: false, isError: true });
          const { onGifsFetchError } = this.props;
          if (onGifsFetchError)
            onGifsFetchError(error);
        }
        if (gifs2) {
          if (prefetchCount === gifs2.length) {
            this.setState({ isDoneFetching: true });
          } else {
            this.setState({ gifs: gifs2, isFetching: false });
            const { onGifsFetched } = this.props;
            if (onGifsFetched)
              onGifsFetched(gifs2);
            this.onFetch();
          }
        }
      }
    }));
  }
  componentDidMount() {
    this.unmounted = false;
    this.onFetch();
  }
  componentWillUnmount() {
    this.unmounted = true;
  }
  render() {
    const {
      onGifVisible,
      onGifRightClick,
      className = _Grid.className,
      onGifSeen,
      onGifClick,
      onGifKeyPress,
      user,
      overlay,
      hideAttribution,
      noLink,
      borderRadius,
      noResultsMessage,
      columns,
      width,
      gutter,
      useTransform,
      columnOffsets,
      backgroundColor,
      loaderConfig,
      tabIndex = 0,
      layoutType = "GRID",
      loader: LoaderVisual = loader_default
    } = this.props;
    const { gifWidth, gifs, isError, isDoneFetching } = this.state;
    const showLoader = !isDoneFetching;
    const isFirstLoad = gifs.length === 0;
    const itemHeights = gifs.map((gif) => getGifHeight(gif, gifWidth));
    return /* @__PURE__ */ React.createElement(pingback_context_manager_default, { attributes: { layout_type: layoutType } }, /* @__PURE__ */ React.createElement("div", { className, style: { width } }, /* @__PURE__ */ React.createElement(
      masonry_grid_default,
      {
        itemHeights,
        useTransform,
        itemWidth: gifWidth,
        columns,
        gutter,
        columnOffsets
      },
      gifs.map((gif) => /* @__PURE__ */ React.createElement(
        gif_default,
        {
          gif,
          tabIndex,
          key: gif.id,
          width: gifWidth,
          onGifClick,
          onGifKeyPress,
          onGifSeen,
          onGifVisible,
          onGifRightClick,
          user,
          overlay,
          backgroundColor,
          hideAttribution,
          noLink,
          borderRadius
        }
      ))
    ), !showLoader && gifs.length === 0 && noResultsMessage, isError ? /* @__PURE__ */ React.createElement(fetch_error_default, { onClick: this.onFetch }) : showLoader && /* @__PURE__ */ React.createElement(observer_default, { onVisibleChange: this.onLoaderVisible, config: loaderConfig }, /* @__PURE__ */ React.createElement(Loader, { $isFirstLoad: isFirstLoad }, /* @__PURE__ */ React.createElement(LoaderVisual, { className: _Grid.loaderClassName })))));
  }
};
_Grid.className = "giphy-grid";
_Grid.loaderClassName = "loader";
_Grid.fetchDebounce = 250;
_Grid.defaultProps = defaultProps;
_Grid.getDerivedStateFromProps = ({ columns, gutter, width }, prevState) => {
  const gutterOffset = gutter * (columns - 1);
  const gifWidth = Math.floor((width - gutterOffset) / columns);
  if (prevState.gifWidth !== gifWidth) {
    return { gifWidth };
  }
  return null;
};
var Grid = _Grid;
var grid_default = Grid;

export {
  grid_default
};
