// @giphy/react-components/dist/esm/chunk-BNMLYTE7.js
import {
  VolumeOffIcon,
  VolumeOnIcon,
  video_default
} from "./chunk-XQXMCEQB.js";
import {
  __spreadProps,
  __spreadValues
} from "./chunk-BG6MEPCE.js";

// src/components/video/video-overlay.tsx
import React, { useEffect, useRef, useState } from "react";
import styled from "styled-components";
var VideoContainer = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    height: 100%;
    width: 100%;
`;
var VideoStyled = styled(video_default)`
    height: 100%;
    display: inline-block;
    object-fit: fill;
    pointer-events: none;
    background: rgb(0, 0, 0, 0);
`;
var Button = styled.div`
    position: absolute;
    top: 6px;
    right: 6px;
    cursor: pointer;
    opacity: ${(props) => props.$isHovered ? 1 : 0.8};
    transition: opacity ease-out 800ms;
`;
var speakerClassName = "giphy-video-overlay-button";
var VolumeButton = ({ muted, toggleMute, mutedByBrowser, isHovered }) => /* @__PURE__ */ React.createElement(
  Button,
  {
    className: speakerClassName,
    onClick: (e) => {
      e.preventDefault();
      e.stopPropagation();
      toggleMute();
    },
    $isHovered: isHovered
  },
  muted || mutedByBrowser || !isHovered ? /* @__PURE__ */ React.createElement(VolumeOffIcon, null) : /* @__PURE__ */ React.createElement(VolumeOnIcon, null)
);
var VideoOverlay = ({
  gif,
  isHovered,
  hideMuteButton,
  width,
  height,
  className,
  muted: userPrefMuted = false,
  onUserMuted
}) => {
  const [muted, setMuted] = useState(userPrefMuted);
  const [mutedByBrowser, setMutedByBrowser] = useState(false);
  const lastMutedState = useRef(muted);
  const toggleMute = () => {
    if (mutedByBrowser) {
      setMutedByBrowser(false);
      setMuted(false);
    } else {
      onUserMuted == null ? void 0 : onUserMuted(!muted);
      setMuted(!muted);
    }
  };
  useEffect(() => {
    setMuted(userPrefMuted);
  }, [userPrefMuted, setMuted]);
  useEffect(() => {
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "hidden") {
        lastMutedState.current = muted;
        setMuted(true);
      } else {
        setMuted(lastMutedState.current);
      }
    });
  }, [muted, setMuted]);
  const props = { toggleMute, muted, mutedByBrowser };
  return /* @__PURE__ */ React.createElement(VideoContainer, { className }, isHovered && /* @__PURE__ */ React.createElement(
    VideoStyled,
    {
      gif,
      key: gif.id,
      loop: true,
      controls: true,
      hideAttribution: true,
      hideMute: true,
      persistentControls: true,
      muted,
      width,
      height,
      onMuted: setMutedByBrowser
    }
  ), !hideMuteButton && /* @__PURE__ */ React.createElement(VolumeButton, __spreadProps(__spreadValues({}, props), { isHovered })));
};
VideoOverlay.imgClassName = speakerClassName;
var video_overlay_default = VideoOverlay;

export {
  video_overlay_default
};
