// @giphy/react-components/dist/esm/chunk-7NP7G6TJ.js
// src/components/video/progress-bar.tsx
import { giphyWhite } from "@giphy/colors";
import React from "react";
import useRaf from "react-use/lib/useRaf";
import styled from "styled-components";
var Bar = styled.div`
    background: ${giphyWhite};
    height: ${(props) => props.$barHeight}px;
    position: absolute;
    width: 5px;
    bottom: 0;
    left: 0;
    opacity: 0.95;
`;
var ProgressBar = ({ videoEl }) => {
  useRaf(2147483647, 100);
  const time = (videoEl == null ? void 0 : videoEl.currentTime) || 0;
  const duration = (videoEl == null ? void 0 : videoEl.duration) || 0;
  const val = time / duration;
  let percentage = Math.round(val * 100);
  let barHeight = 5;
  if ((videoEl == null ? void 0 : videoEl.height) < 200) {
    barHeight = 3;
  } else if ((videoEl == null ? void 0 : videoEl.height) < 300) {
    barHeight = 4;
  }
  percentage = duration < 10 && percentage > 98 ? 100 : percentage;
  return /* @__PURE__ */ React.createElement(Bar, { style: { width: `${percentage}%` }, $barHeight: barHeight, className: "hide-in-percy" });
};
var progress_bar_default = ProgressBar;

export {
  progress_bar_default
};
