// @giphy/react-components/dist/esm/chunk-Q3JJULZD.js
import {
  attribution_default
} from "./chunk-7V5XIXM5.js";

// src/components/attribution/overlay.tsx
import React, { useRef } from "react";
import styled from "styled-components";
var Background = styled.div`
    background: linear-gradient(rgba(0, 0, 0, 0), rgba(18, 18, 18, 0.6));
    cursor: default;
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    height: 75px;
    pointer-events: none;
`;
var Attribution = styled(attribution_default)`
    position: absolute;
    bottom: 10px;
    left: 10px;
    right: 10px;
`;
var Container = styled.div`
    transition: opacity 150ms ease-in;
`;
var AttributionOverlay = ({ gif, isHovered, onClick }) => {
  const hasHovered = useRef(isHovered);
  if (isHovered) {
    hasHovered.current = true;
  }
  return gif.user && hasHovered.current ? /* @__PURE__ */ React.createElement(Container, { style: { opacity: isHovered ? 1 : 0 } }, /* @__PURE__ */ React.createElement(Background, null), /* @__PURE__ */ React.createElement(Attribution, { gif, onClick })) : null;
};
var overlay_default = AttributionOverlay;

export {
  overlay_default
};
