// @replyke/core/dist/esm/hooks/comments/useProfileComments.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useCallback, useEffect, useRef, useState } from "react";
import useFetchManyComments from "./useFetchManyComments";
import { handleError } from "../../utils/handleError";
function useProfileComments(props) {
    var _this = this;
    var userId = props.userId, _a = props.limit, limit = _a === void 0 ? 10 : _a, _b = props.defaultSortBy, defaultSortBy = _b === void 0 ? "new" : _b, includeEntity = props.includeEntity;
    var fetchManyComments = useFetchManyComments();
    var loading = useRef(true);
    var _c = useState(true), loadingState = _c[0], setLoadingState = _c[1]; // required to trigger rerenders
    var hasMore = useRef(true);
    var _d = useState(true), hasMoreState = _d[0], setHasMoreState = _d[1]; // required to trigger rerenders
    var _e = useState(defaultSortBy), sortBy = _e[0], setSortBy = _e[1];
    var _f = useState(1), page = _f[0], setPage = _f[1];
    var _g = useState([]), comments = _g[0], setComments = _g[1];
    var resetComments = useCallback(function () { return __awaiter(_this, void 0, void 0, function () {
        var newComments, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!userId) {
                        // console.warn(
                        //   "The 'fetch comments' operation was invoked without a valid user and has been aborted."
                        // );
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    loading.current = true;
                    setLoadingState(true);
                    hasMore.current = true;
                    setHasMoreState(true);
                    setPage(1);
                    return [4 /*yield*/, fetchManyComments({
                            userId: userId,
                            page: 1,
                            sortBy: sortBy,
                            limit: 10,
                            includeEntity: includeEntity,
                        })];
                case 2:
                    newComments = _a.sent();
                    if (newComments) {
                        setComments(newComments);
                        if (newComments.length < limit) {
                            hasMore.current = false;
                            setHasMoreState(false);
                        }
                    }
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    handleError(err_1, "Failed to reset profile comments:");
                    return [3 /*break*/, 5];
                case 4:
                    loading.current = false;
                    setLoadingState(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, [fetchManyComments, limit, sortBy, userId, includeEntity]);
    var loadMore = function () {
        if (loading.current || !hasMore.current)
            return;
        setPage(function (prevPage) {
            return prevPage + 1;
        });
    };
    useEffect(function () {
        resetComments();
    }, [resetComments]);
    // useEffect to get a new batch of entities
    useEffect(function () {
        var loadMoreComments = function () { return __awaiter(_this, void 0, void 0, function () {
            var newComments_1, err_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        loading.current = true;
                        setLoadingState(true);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, fetchManyComments({
                                userId: userId,
                                page: page,
                                sortBy: sortBy,
                                limit: limit,
                                includeEntity: includeEntity,
                            })];
                    case 2:
                        newComments_1 = _a.sent();
                        if (newComments_1) {
                            setComments(function (prevComments) { return __spreadArray(__spreadArray([], prevComments, true), newComments_1, true); });
                            if (newComments_1.length < limit) {
                                hasMore.current = false;
                                setHasMoreState(false);
                            }
                        }
                        return [3 /*break*/, 5];
                    case 3:
                        err_2 = _a.sent();
                        handleError(err_2, "Loading more comments failed:");
                        return [3 /*break*/, 5];
                    case 4:
                        loading.current = false;
                        setLoadingState(false);
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        // We only load more if th page changed
        if (page > 1 && hasMore.current && !loading.current) {
            loadMoreComments();
        }
    }, [page]);
    return {
        comments: comments,
        loading: loadingState,
        hasMore: hasMoreState,
        sortBy: sortBy,
        setSortBy: setSortBy,
        loadMore: loadMore,
    };
}
export default useProfileComments;
