// @replyke/core/dist/cjs/hooks/lists/useListsActions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useListsActions = useListsActions;
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var listsSlice_1 = require("../../store/slices/listsSlice");
var listsApi_1 = require("../../store/api/listsApi");
var handleError_1 = require("../../utils/handleError");
/**
 * Redux-powered hook that provides all lists actions
 * This replaces the individual hooks and provides a centralized way to manage lists
 */
function useListsActions() {
    var _this = this;
    var dispatch = (0, react_redux_1.useDispatch)();
    // RTK Query hooks
    var fetchRootListQuery = (0, listsApi_1.useLazyFetchRootListQuery)()[0];
    var fetchSubListsQuery = (0, listsApi_1.useLazyFetchSubListsQuery)()[0];
    var createListMutation = (0, listsApi_1.useCreateListMutation)()[0];
    var updateListMutation = (0, listsApi_1.useUpdateListMutation)()[0];
    var deleteListMutation = (0, listsApi_1.useDeleteListMutation)()[0];
    var addToListMutation = (0, listsApi_1.useAddToListMutation)()[0];
    var removeFromListMutation = (0, listsApi_1.useRemoveFromListMutation)()[0];
    // Navigation actions
    var openListAction = (0, react_1.useCallback)(function (list) {
        dispatch((0, listsSlice_1.openList)(list));
    }, [dispatch]);
    var goBackAction = (0, react_1.useCallback)(function () {
        dispatch((0, listsSlice_1.goBack)());
    }, [dispatch]);
    var goToRootAction = (0, react_1.useCallback)(function () {
        dispatch((0, listsSlice_1.goToRoot)());
    }, [dispatch]);
    // Fetch root list
    var fetchRootList = (0, react_1.useCallback)(function (projectId) { return __awaiter(_this, void 0, void 0, function () {
        var result, rootList, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId) {
                        console.warn("Can't fetch root list without projectId.");
                        return [2 /*return*/];
                    }
                    dispatch((0, listsSlice_1.setLoading)(true));
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, fetchRootListQuery({ projectId: projectId }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        rootList = __assign(__assign({}, result), { parentId: null });
                        dispatch((0, listsSlice_1.setCurrentList)(rootList));
                    }
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    (0, handleError_1.handleError)(err_1, "Failed fetching root list");
                    dispatch((0, listsSlice_1.handleError)());
                    return [3 /*break*/, 5];
                case 4:
                    dispatch((0, listsSlice_1.setLoading)(false));
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, [dispatch, fetchRootListQuery]);
    // Fetch sub-lists
    var fetchSubLists = (0, react_1.useCallback)(function (projectId, listId) { return __awaiter(_this, void 0, void 0, function () {
        var result, err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !listId) {
                        console.warn("Can't fetch sub-lists without projectId and listId.");
                        return [2 /*return*/];
                    }
                    dispatch((0, listsSlice_1.setLoading)(true));
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, fetchSubListsQuery({ projectId: projectId, listId: listId }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        dispatch((0, listsSlice_1.setSubLists)({ lists: result, parentListId: listId }));
                    }
                    return [3 /*break*/, 5];
                case 3:
                    err_2 = _a.sent();
                    (0, handleError_1.handleError)(err_2, "Failed fetching sub-lists");
                    dispatch((0, listsSlice_1.handleError)());
                    return [3 /*break*/, 5];
                case 4:
                    dispatch((0, listsSlice_1.setLoading)(false));
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, [dispatch, fetchSubListsQuery]);
    // Create list
    var createList = (0, react_1.useCallback)(function (projectId, parentListId, listName) { return __awaiter(_this, void 0, void 0, function () {
        var result, err_3;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !parentListId || !listName) {
                        console.error("Missing required parameters for creating list.");
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, createListMutation({
                            projectId: projectId,
                            parentListId: parentListId,
                            listName: listName,
                        }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        dispatch((0, listsSlice_1.addNewListAndNavigate)(result));
                    }
                    return [3 /*break*/, 4];
                case 3:
                    err_3 = _a.sent();
                    (0, handleError_1.handleError)(err_3, "Failed to create list");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [createListMutation, dispatch]);
    // Update list
    var updateList = (0, react_1.useCallback)(function (projectId, listId, update) { return __awaiter(_this, void 0, void 0, function () {
        var result, err_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !listId) {
                        console.error("Missing required parameters for updating list.");
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, updateListMutation({
                            projectId: projectId,
                            listId: listId,
                            update: update,
                        }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        // Check if it's the current list or a sub-list
                        dispatch((0, listsSlice_1.updateListInSubLists)(result));
                    }
                    return [3 /*break*/, 4];
                case 3:
                    err_4 = _a.sent();
                    (0, handleError_1.handleError)(err_4, "Failed to update list");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [updateListMutation, dispatch]);
    // Delete list
    var deleteList = (0, react_1.useCallback)(function (projectId, list) { return __awaiter(_this, void 0, void 0, function () {
        var err_5;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !list) {
                        console.error("Missing required parameters for deleting list.");
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, deleteListMutation({
                            projectId: projectId,
                            listId: list.id,
                        }).unwrap()];
                case 2:
                    _a.sent();
                    dispatch((0, listsSlice_1.handleListDeletion)({ listId: list.id, parentId: list.parentId }));
                    return [3 /*break*/, 4];
                case 3:
                    err_5 = _a.sent();
                    (0, handleError_1.handleError)(err_5, "Failed to delete list");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [deleteListMutation, dispatch]);
    // Add entity to list
    var addToList = (0, react_1.useCallback)(function (projectId, listId, entityId) { return __awaiter(_this, void 0, void 0, function () {
        var result, err_6;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !listId || !entityId) {
                        console.error("Missing required parameters for adding to list.");
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, addToListMutation({
                            projectId: projectId,
                            listId: listId,
                            entityId: entityId,
                        }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        dispatch((0, listsSlice_1.updateCurrentList)(result));
                    }
                    return [3 /*break*/, 4];
                case 3:
                    err_6 = _a.sent();
                    (0, handleError_1.handleError)(err_6, "Failed to add entity to list");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [addToListMutation, dispatch]);
    // Remove entity from list
    var removeFromList = (0, react_1.useCallback)(function (projectId, listId, entityId) { return __awaiter(_this, void 0, void 0, function () {
        var result, err_7;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!projectId || !listId || !entityId) {
                        console.error("Missing required parameters for removing from list.");
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, removeFromListMutation({
                            projectId: projectId,
                            listId: listId,
                            entityId: entityId,
                        }).unwrap()];
                case 2:
                    result = _a.sent();
                    if (result) {
                        dispatch((0, listsSlice_1.updateCurrentList)(result));
                    }
                    return [3 /*break*/, 4];
                case 3:
                    err_7 = _a.sent();
                    (0, handleError_1.handleError)(err_7, "Failed to remove entity from list");
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    }); }, [removeFromListMutation, dispatch]);
    // Reset lists
    var resetListsAction = (0, react_1.useCallback)(function () {
        dispatch((0, listsSlice_1.resetLists)());
    }, [dispatch]);
    return (0, react_1.useMemo)(function () { return ({
        // Navigation
        openList: openListAction,
        goBack: goBackAction,
        goToRoot: goToRootAction,
        // Data fetching
        fetchRootList: fetchRootList,
        fetchSubLists: fetchSubLists,
        // CRUD operations
        createList: createList,
        updateList: updateList,
        deleteList: deleteList,
        addToList: addToList,
        removeFromList: removeFromList,
        // Utility
        resetLists: resetListsAction,
    }); }, [
        openListAction,
        goBackAction,
        goToRootAction,
        fetchRootList,
        fetchSubLists,
        createList,
        updateList,
        deleteList,
        addToList,
        removeFromList,
        resetListsAction,
    ]);
}
exports.default = useListsActions;
