// @giphy/react-components/dist/components/search-bar/context.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/components/search-bar/context.tsx
var context_exports = {};
__export(context_exports, {
  CssVars: () => CssVars,
  SearchContext: () => SearchContext,
  _SearchContext: () => _SearchContext,
  default: () => context_default
});
module.exports = __toCommonJS(context_exports);
var import_colors = require("@giphy/colors");
var import_js_fetch_api = require("@giphy/js-fetch-api");
var import_js_util = require("@giphy/js-util");
var import_react2 = __toESM(require("react"));
var import_styled_components = __toESM(require("styled-components"));

// src/components/pingback-context-manager.tsx
var import_js_analytics = require("@giphy/js-analytics");
var import_react = __toESM(require("react"));
var PingbackContext = (0, import_react.createContext)({});
var PingbackContextManager = ({ attributes, children }) => {
  const { attributes: parentAttributes = {} } = (0, import_react.useContext)(PingbackContext);
  return /* @__PURE__ */ import_react.default.createElement(PingbackContext.Provider, { value: { attributes: (0, import_js_analytics.mergeAttributes)(parentAttributes, attributes, "layout_type") } }, children);
};
var pingback_context_manager_default = PingbackContextManager;

// src/components/search-bar/context.tsx
var CssVars = {
  searchbarHeight: `--searchbar-height`,
  bgColor: `--searchbar-bg-color`,
  bgColor2: `--searchbar-bg-color-2`,
  fgColor: `--searchbar-fg-color`,
  cancelButtonDisplay: `--searchbar-cancel-button-display`
};
var Container = import_styled_components.default.div`
    ${CssVars.searchbarHeight}: ${(props) => props.$searchbarHeight || 42}px;
    @media (${(props) => props.$mobileMediaQuery}) {
        ${CssVars.searchbarHeight}: ${(props) => props.$mobileSearchbarHeight || 35}px;
    }
    ${CssVars.bgColor}: ${import_colors.giphyWhite};
    ${CssVars.bgColor2}: ${import_colors.giphyWhite};
    ${CssVars.fgColor}: ${import_colors.giphyBlack};
    ${(props) => props.$darkMode && import_styled_components.css`
            ${CssVars.fgColor}: ${import_colors.giphyWhite};
            ${CssVars.bgColor}: ${import_colors.giphyBlack};
            ${CssVars.bgColor2}: ${import_colors.giphyCharcoal};
        `}
    ${CssVars.cancelButtonDisplay}: ${(props) => props.$hideCancelButton ? "none" : "block"};
`;
var SearchContext = (0, import_react2.createContext)({});
var _SearchContext = (0, import_react2.createContext)({});
var emptyChannels = [];
var emptyGifsResult = {
  data: [],
  pagination: { total_count: 0, count: 0, offset: 0 },
  meta: { status: 200, msg: "OK", response_id: "" }
};
var SearchContextManager = ({
  children,
  options = {},
  apiKey,
  initialTerm = "",
  initialChannel,
  shouldDefaultToTrending = true,
  shouldFetchChannels = true,
  theme
}) => {
  var _a, _b;
  const gf = (0, import_react2.useMemo)(() => new import_js_fetch_api.GiphyFetch(apiKey), [apiKey]);
  const [currentChannels, setChannels] = (0, import_react2.useState)([]);
  const [term, _setSearch] = (0, import_react2.useState)(initialTerm);
  const [_inputValOverride, setSearch] = (0, import_react2.useState)(initialTerm);
  const [isFetching, setIsFetching] = (0, import_react2.useState)(false);
  const [activeChannel, _setActiveChannel] = (0, import_react2.useState)(initialChannel);
  const setActiveChannel = (0, import_react2.useCallback)((activeChannel2) => {
    _setSearch("");
    _setActiveChannel(activeChannel2);
  }, []);
  const [trendingSearches, setTrendingSearches] = (0, import_react2.useState)([]);
  const [isFocused, setIsFocused] = (0, import_react2.useState)(false);
  const searchKey = [term, options.type, ((_a = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _a.username) || ""].filter((val) => !!val).join(" / ");
  const fetchGifs = (0, import_react2.useCallback)(
    (offset) => __async(void 0, null, function* () {
      var _a2;
      setIsFetching(true);
      let result = emptyGifsResult;
      if (term) {
        result = yield gf.search(term, __spreadProps(__spreadValues({}, options), {
          offset,
          channel: (_a2 = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _a2.username
        }));
      } else if (shouldDefaultToTrending) {
        result = yield gf.trending(__spreadProps(__spreadValues({}, options), { offset }));
      }
      setIsFetching(false);
      return result;
    }),
    [(_b = activeChannel == null ? void 0 : activeChannel.user) == null ? void 0 : _b.username, gf, options, term, shouldDefaultToTrending]
  );
  const fetchAnimatedText = (0, import_react2.useCallback)(
    (offset) => __async(void 0, null, function* () {
      const limit = options.limit || 50;
      const result = yield gf.animate(term, { offset, limit });
      if (!result.pagination) {
        result.pagination = { count: limit, total_count: limit, offset };
      }
      return result;
    }),
    [gf, options.limit, term]
  );
  const fetchChannelSearch = (0, import_react2.useCallback)(
    (offset) => __async(void 0, null, function* () {
      const search = term.indexOf("@") === 0 ? term.split(" ")[0] : term;
      const result = yield gf.channels(search, { offset });
      return result.data;
    }),
    [gf, term]
  );
  (0, import_react2.useEffect)(() => {
    const fetchTrendingSearches = () => __async(void 0, null, function* () {
      let searches = [];
      try {
        const { data = [] } = yield (0, import_js_fetch_api.request)(`trending/searches?api_key=${apiKey}`);
        searches = data;
      } catch (error) {
        import_js_util.Logger.warn(`Trending searches request failed: ${error}`);
      }
      setTrendingSearches(searches);
    });
    fetchTrendingSearches();
  }, [apiKey]);
  (0, import_react2.useEffect)(() => {
    const foundChannel = currentChannels.find(({ slug }) => term.indexOf(`@${slug} `) === 0);
    if (foundChannel) {
      _setActiveChannel(foundChannel);
    }
  }, [term, currentChannels, _setActiveChannel]);
  (0, import_react2.useEffect)(() => {
    if (shouldFetchChannels) {
      const fetchChannels = () => __async(void 0, null, function* () {
        const channels = yield fetchChannelSearch(0);
        setChannels(channels || emptyChannels);
      });
      if (!activeChannel && term.replace("@", "")) {
        fetchChannels();
      }
    }
  }, [term, activeChannel]);
  return /* @__PURE__ */ import_react2.default.createElement(
    SearchContext.Provider,
    {
      value: {
        activeChannel,
        setChannels,
        currentChannels,
        setActiveChannel,
        fetchChannelSearch,
        term,
        trendingSearches,
        setSearch,
        fetchGifs,
        fetchAnimatedText,
        searchKey,
        isFetching,
        isFocused
      }
    },
    /* @__PURE__ */ import_react2.default.createElement(_SearchContext.Provider, { value: { setIsFocused, _setSearch, _inputValOverride } }, /* @__PURE__ */ import_react2.default.createElement(
      Container,
      {
        $darkMode: theme == null ? void 0 : theme.darkMode,
        $searchbarHeight: theme == null ? void 0 : theme.searchbarHeight,
        $hideCancelButton: theme == null ? void 0 : theme.hideCancelButton,
        $mobileSearchbarHeight: theme == null ? void 0 : theme.mobileSearchbarHeight,
        $mobileMediaQuery: (theme == null ? void 0 : theme.mobileMediaQuery) || "max-width: 480px"
      },
      /* @__PURE__ */ import_react2.default.createElement(pingback_context_manager_default, { attributes: { layout_type: "SEARCH" } }, children)
    ))
  );
};
var context_default = SearchContextManager;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CssVars,
  SearchContext,
  _SearchContext
});
