// @giphy/react-components/dist/components/search-bar/index.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/search-bar/index.tsx
var search_bar_exports = {};
__export(search_bar_exports, {
  default: () => search_bar_default
});
module.exports = __toCommonJS(search_bar_exports);
var import_colors4 = require("@giphy/colors");
var import_react8 = __toESM(require("react"));
var import_useDebounce = __toESM(require("react-use/lib/useDebounce"));
var import_styled_components6 = __toESM(require("styled-components"));

// src/components/search-bar/cancel-icon.tsx
var import_react2 = __toESM(require("react"));
var import_styled_components2 = __toESM(require("styled-components"));

// src/components/search-bar/context.tsx
var import_colors = require("@giphy/colors");
var import_js_fetch_api = require("@giphy/js-fetch-api");
var import_js_util = require("@giphy/js-util");
var import_react = __toESM(require("react"));
var import_styled_components = __toESM(require("styled-components"));
var CssVars = {
  searchbarHeight: `--searchbar-height`,
  bgColor: `--searchbar-bg-color`,
  bgColor2: `--searchbar-bg-color-2`,
  fgColor: `--searchbar-fg-color`,
  cancelButtonDisplay: `--searchbar-cancel-button-display`
};
var Container = import_styled_components.default.div`
    ${CssVars.searchbarHeight}: ${(props) => props.$searchbarHeight || 42}px;
    @media (${(props) => props.$mobileMediaQuery}) {
        ${CssVars.searchbarHeight}: ${(props) => props.$mobileSearchbarHeight || 35}px;
    }
    ${CssVars.bgColor}: ${import_colors.giphyWhite};
    ${CssVars.bgColor2}: ${import_colors.giphyWhite};
    ${CssVars.fgColor}: ${import_colors.giphyBlack};
    ${(props) => props.$darkMode && import_styled_components.css`
            ${CssVars.fgColor}: ${import_colors.giphyWhite};
            ${CssVars.bgColor}: ${import_colors.giphyBlack};
            ${CssVars.bgColor2}: ${import_colors.giphyCharcoal};
        `}
    ${CssVars.cancelButtonDisplay}: ${(props) => props.$hideCancelButton ? "none" : "block"};
`;
var SearchContext = (0, import_react.createContext)({});
var _SearchContext = (0, import_react.createContext)({});

// src/components/search-bar/cancel-icon.tsx
var Icon = import_styled_components2.default.svg`
    position: relative;
    right: 10px;
    margin-left: 5px;
    cursor: pointer;
    display: var(${CssVars.cancelButtonDisplay});
`;
var CancelIcon = ({ width = 17, height = 17, setCleared }) => {
  const { term, setActiveChannel, activeChannel } = (0, import_react2.useContext)(SearchContext);
  return term || activeChannel ? /* @__PURE__ */ import_react2.default.createElement(
    Icon,
    {
      className: CancelIcon.className,
      width,
      height,
      viewBox: "0 0 17 17",
      version: "1.1",
      xmlns: "http://www.w3.org/2000/svg",
      onClick: (e) => {
        e.stopPropagation();
        e.preventDefault();
        setCleared(true);
        setActiveChannel(void 0);
      }
    },
    /* @__PURE__ */ import_react2.default.createElement("g", { stroke: "none", strokeWidth: "1", fill: "none", fillRule: "evenodd", opacity: "0.582730817" }, /* @__PURE__ */ import_react2.default.createElement("g", { transform: "translate(-300.000000, -150.000000)", fill: "#8E8E93", fillRule: "nonzero" }, /* @__PURE__ */ import_react2.default.createElement("g", { transform: "translate(0.000000, 135.000000)" }, /* @__PURE__ */ import_react2.default.createElement("g", null, /* @__PURE__ */ import_react2.default.createElement("g", null, /* @__PURE__ */ import_react2.default.createElement("g", { transform: "translate(11.000000, 13.000000)" }, /* @__PURE__ */ import_react2.default.createElement("g", { transform: "translate(289.000000, 2.000000)" }, /* @__PURE__ */ import_react2.default.createElement(
      "path",
      {
        d: "M8.5,0 C3.805875,0 0,3.805875 0,8.5 C0,13.194125 3.805875,17 8.5,17 C13.194125,17 17,13.194125 17,8.5 C17,3.805875 13.194125,0 8.5,0 Z M9.50158333,8.5 C9.50158333,8.5 11.7250417,10.7234583 11.834125,10.8325417 C12.1110833,11.1095 12.1110833,11.557875 11.834125,11.834125 C11.5571667,12.1110833 11.1087917,12.1110833 10.8325417,11.834125 C10.7234583,11.72575 8.5,9.50158333 8.5,9.50158333 C8.5,9.50158333 6.27654167,11.7250417 6.16745833,11.834125 C5.8905,12.1110833 5.442125,12.1110833 5.165875,11.834125 C4.88891667,11.5571667 4.88891667,11.1087917 5.165875,10.8325417 C5.27425,10.7234583 7.49841667,8.5 7.49841667,8.5 C7.49841667,8.5 5.27495833,6.27654167 5.165875,6.16745833 C4.88891667,5.8905 4.88891667,5.442125 5.165875,5.165875 C5.44283333,4.88891667 5.89120833,4.88891667 6.16745833,5.165875 C6.27654167,5.27425 8.5,7.49841667 8.5,7.49841667 C8.5,7.49841667 10.7234583,5.27495833 10.8325417,5.165875 C11.1095,4.88891667 11.557875,4.88891667 11.834125,5.165875 C12.1110833,5.44283333 12.1110833,5.89120833 11.834125,6.16745833 C11.72575,6.27654167 9.50158333,8.5 9.50158333,8.5 Z",
        opacity: "0.75"
      }
    ))))))))
  ) : null;
};
CancelIcon.className = "giphy-search-bar-cancel";
var cancel_icon_default = CancelIcon;

// src/components/search-bar/search-bar-channel.tsx
var import_colors3 = require("@giphy/colors");
var import_react5 = __toESM(require("react"));
var import_styled_components4 = __toESM(require("styled-components"));

// src/components/attribution/avatar.tsx
var import_react3 = __toESM(require("react"));
var import_styled_components3 = __toESM(require("styled-components"));
var getSmallAvatar = (avatar) => {
  var _a, _b;
  if (!avatar)
    return "";
  const ext = (_b = (_a = avatar == null ? void 0 : avatar.split(".")) == null ? void 0 : _a.pop()) == null ? void 0 : _b.toLowerCase();
  return avatar.replace(`.${ext}`, `/80h.${ext}`);
};
var Img = import_styled_components3.default.img`
    object-fit: cover;
    width: 32px;
    height: 32px;
    margin-right: 8px;
`;
var Avatar = ({ user, className = "" }) => {
  const defaultAvatarId = (0, import_react3.useRef)(Math.floor(Math.random() * 5) + 1);
  const url = user.avatar_url ? getSmallAvatar(user.avatar_url) : `https://media.giphy.com/avatars/default${defaultAvatarId.current}.gif`;
  return /* @__PURE__ */ import_react3.default.createElement(Img, { src: url, className });
};
var avatar_default = Avatar;

// src/components/attribution/verified-badge.tsx
var import_colors2 = require("@giphy/colors");
var import_react4 = __toESM(require("react"));
var VerifiedBadge = ({ className = "", size = 17, fill = "#15CDFF" }) => /* @__PURE__ */ import_react4.default.createElement("svg", { className: [VerifiedBadge.className, className].join(" "), height: size, width: "19px", viewBox: "0 0 19 17" }, /* @__PURE__ */ import_react4.default.createElement(
  "path",
  {
    className: VerifiedBadge.checkMarkClassName,
    d: "M9.32727273,9.44126709 L9.32727273,3.03016561 L6.55027155,3.03016561 L6.55027155,10.8150746 L6.55027155,12.188882 L12.1042739,12.188882 L12.1042739,9.44126709 L9.32727273,9.44126709 Z",
    fill: import_colors2.giphyBlack,
    transform: "translate(9.327273, 7.609524) scale(-1, 1) rotate(-45.000000) translate(-9.327273, -7.609524) "
  }
), /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(-532.000000, -466.000000)", fill }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(141.000000, 235.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(264.000000, 0.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(10.000000, 224.000000)" }, /* @__PURE__ */ import_react4.default.createElement("g", { transform: "translate(114.000000, 2.500000)" }, /* @__PURE__ */ import_react4.default.createElement("path", { d: "M15.112432,4.80769231 L16.8814194,6.87556817 L19.4157673,7.90116318 L19.6184416,10.6028916 L21.0594951,12.9065042 L19.6184416,15.2101168 L19.4157673,17.9118452 L16.8814194,18.9374402 L15.112432,21.0053161 L12.4528245,20.3611511 L9.79321699,21.0053161 L8.02422954,18.9374402 L5.48988167,17.9118452 L5.28720734,15.2101168 L3.84615385,12.9065042 L5.28720734,10.6028916 L5.48988167,7.90116318 L8.02422954,6.87556817 L9.79321699,4.80769231 L12.4528245,5.4518573 L15.112432,4.80769231 Z M17.8163503,10.8991009 L15.9282384,9.01098901 L11.5681538,13.3696923 L9.68115218,11.4818515 L7.81302031,13.3499833 L9.7011322,15.2380952 L11.5892441,17.1262071 L17.8163503,10.8991009 Z" })))))));
VerifiedBadge.className = "giphy-verified-badge";
VerifiedBadge.checkMarkClassName = "giphy-verified-checkmark";
var verified_badge_default = VerifiedBadge;

// src/components/search-bar/search-bar-channel.tsx
var channelMargin = 6;
var innerHeight = `calc(var(${CssVars.searchbarHeight}) - ${channelMargin * 2}px)`;
var animateAvatar = import_styled_components4.keyframes`
to {
    width: ${innerHeight};
}
`;
var Avatar2 = (0, import_styled_components4.default)(avatar_default)`
    height: ${innerHeight};
    margin: 0;
    width: 0;
    animation: ${animateAvatar} 100ms ease-in-out forwards;
`;
var Username = import_styled_components4.default.div`
    background: var(${CssVars.bgColor2});
    display: flex;
    align-items: center;
    padding-left: ${channelMargin}px;
`;
var UsernamePill = import_styled_components4.default.div`
    background: ${import_colors3.giphyWhiteSmoke};
    display: flex;
    padding: 0 4px;
    color: ${import_colors3.giphyDarkCharcoal};
    font-family: interface, Helvetica Neue, helvetica, sans-serif;
    font-weight: 600;
    font-size: 12px;
    align-items: center;
    height: ${innerHeight};
    @media (max-width: 480px) {
        display: none;
    }
`;
var SearchBarChannel = ({ className = "" }) => {
  const { activeChannel } = (0, import_react5.useContext)(SearchContext);
  return activeChannel ? /* @__PURE__ */ import_react5.default.createElement(Username, { className }, /* @__PURE__ */ import_react5.default.createElement(Avatar2, { user: activeChannel.user }), /* @__PURE__ */ import_react5.default.createElement(UsernamePill, { key: activeChannel.id }, /* @__PURE__ */ import_react5.default.createElement("div", null, "@", activeChannel.user.username), activeChannel.user.is_verified && /* @__PURE__ */ import_react5.default.createElement(verified_badge_default, { size: 14 }))) : null;
};
var search_bar_channel_default = SearchBarChannel;

// src/components/search-bar/search-button.tsx
var import_react7 = __toESM(require("react"));
var import_useThrottle = __toESM(require("react-use/lib/useThrottle"));
var import_styled_components5 = __toESM(require("styled-components"));

// src/components/search-bar/search-icon.tsx
var import_react6 = __toESM(require("react"));
var SearchIcon = ({ className = "" }) => /* @__PURE__ */ import_react6.default.createElement("svg", { viewBox: "0 0 30 30", version: "1.1", className }, /* @__PURE__ */ import_react6.default.createElement("defs", null, /* @__PURE__ */ import_react6.default.createElement(
  "path",
  {
    d: "M11.5482521,20.4090671 L4.24727698,28.2009189 C3.68084207,28.8054377 2.73159653,28.8363108 2.12707771,28.2698759 C1.5225589,27.703441 1.4916858,26.7541954 2.0581207,26.1496766 L9.40599838,18.3077689 C7.95982241,16.4371424 7.0978836,14.0789715 7.0978836,11.5181818 C7.0978836,5.44914339 11.9392549,0.518181818 17.9252787,0.518181818 C23.9113026,0.518181818 28.7526738,5.44914339 28.7526738,11.5181818 C28.7526738,17.5872202 23.9113026,22.5181818 17.9252787,22.5181818 C15.539851,22.5181818 13.3361963,21.7351359 11.5482521,20.4090671 Z M17.9252787,19.5181818 C22.242011,19.5181818 25.7526738,15.9425536 25.7526738,11.5181818 C25.7526738,7.09381 22.242011,3.51818182 17.9252787,3.51818182 C13.6085464,3.51818182 10.0978836,7.09381 10.0978836,11.5181818 C10.0978836,15.9425536 13.6085464,19.5181818 17.9252787,19.5181818 Z",
    id: "giphy-search-icon-path-1"
  }
)), /* @__PURE__ */ import_react6.default.createElement("g", { id: "search", stroke: "none", strokeWidth: "1", fill: "none", fillRule: "evenodd" }, /* @__PURE__ */ import_react6.default.createElement("g", { id: "icons/search" }, /* @__PURE__ */ import_react6.default.createElement("mask", { id: "giphy-search-icon-mask-2", fill: "white" }, /* @__PURE__ */ import_react6.default.createElement("use", { xlinkHref: "#giphy-search-icon-path-1" })), /* @__PURE__ */ import_react6.default.createElement("use", { id: "Mask", fill: "#FFFFFF", fillRule: "nonzero", xlinkHref: "#giphy-search-icon-path-1" }), /* @__PURE__ */ import_react6.default.createElement("g", { mask: "url(#giphy-search-icon-mask-2)" }, /* @__PURE__ */ import_react6.default.createElement("g", { transform: "translate(0.250000, 0.250000)" }, /* @__PURE__ */ import_react6.default.createElement("g", null))))));
var search_icon_default = SearchIcon;

// src/components/search-bar/search-button.tsx
var time = "2s";
var purp = "#E646B6";
var pink = "#FF6666";
var strongEasing = "cubic-bezier(0.920, 0.240, 0.185, 0.730)";
var searchFx = import_styled_components5.keyframes`
    0% {
      transform: rotate(34deg) translate(-10px, 80px);
    };
    
    100% {
      transform: rotate(34deg) translate(-10px, -20px);
    }
`;
var plus = import_styled_components5.keyframes`
    0% { 
      transform: translate(0px, 0px);
      opacity: 0;
    }
    50% {
      opacity: 1;
    }
    100% {
      transform: translate(10px, -17px);
      opacity: 0;
    }
`;
var gradientFade = import_styled_components5.keyframes`
    0% {
      opacity: 0;
      transform: translateX(-400%);
    }
    50% {
      opacity: 1;
    }
    100% {
      opacity: 0;
      transform: translateX(0);
    }
`;
var Container2 = import_styled_components5.default.div`
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;
    cursor: pointer;
    @media screen and (-ms-high-contrast: active), screen and (-ms-high-contrast: none) {
        display: none;
    }
    height: var(${CssVars.searchbarHeight});
    width: var(${CssVars.searchbarHeight});
`;
var GradientBox = import_styled_components5.default.div`
    position: absolute;
    height: 100%;
    width: 100%;
    background: linear-gradient(45deg, ${purp} 0%, ${pink} 100%);
    border-radius: 0 4px 4px 0;
    overflow: hidden;
    &:before {
        animation: ${gradientFade} ${time} linear 0s infinite;
        background-image: linear-gradient(45deg, ${purp} 0%, ${pink} 50%, ${purp} 100%);
        background-size: 400%;
        background-position: 0% 100%;
        content: '';
        position: absolute;
        left: 0;
        top: 0;
        right: 0;
        bottom: 0;
        width: 400%;
    }
`;
var Fx = import_styled_components5.default.div`
    width: 100%;
    height: 100%;
    position: absolute;
    overflow: hidden;
    &::after {
        content: '+';
        color: white;
        font-family: 'SS Standard';
        font-size: 8px;
        position: absolute;
        top: 65%;
        left: 66%;
        animation: ${plus} 1s ${strongEasing} 0s 1 forwards;
    }
`;
var Scanner = import_styled_components5.default.div`
    position: absolute;
    width: 200%;
    height: 20px;
    background: rgba(255, 255, 255, 0.5);
    transform: rotate(34deg) translate(-10px, -20px);
    animation: ${searchFx} 1s ${strongEasing} 0s 1;
    filter: blur(1px);
`;
var SearchIcon2 = (0, import_styled_components5.default)(search_icon_default)`
    z-index: 1;
    display: flex;
    width: 50%;
    height: 50%;
`;
var SearchButton = ({ onClick }) => {
  const { isFetching } = (0, import_react7.useContext)(SearchContext);
  const throttledFetch = (0, import_useThrottle.default)(isFetching, 1e3);
  return /* @__PURE__ */ import_react7.default.createElement(Container2, { onClick: () => onClick == null ? void 0 : onClick() }, /* @__PURE__ */ import_react7.default.createElement(GradientBox, { suppressHydrationWarning: true }), /* @__PURE__ */ import_react7.default.createElement(SearchIcon2, null), throttledFetch && /* @__PURE__ */ import_react7.default.createElement(Fx, { suppressHydrationWarning: true }, /* @__PURE__ */ import_react7.default.createElement(Scanner, { suppressHydrationWarning: true })));
};
var search_button_default = SearchButton;

// src/components/search-bar/index.tsx
function usePrevious(value) {
  const ref = (0, import_react8.useRef)(value);
  (0, import_react8.useEffect)(() => {
    ref.current = value;
  });
  return ref.current;
}
var SEARCH_DEBOUNCE = 500;
var Container3 = import_styled_components6.default.div`
    display: flex;
    background: white;
    align-items: center;
    border-radius: 4px;
    height: var(${CssVars.searchbarHeight});
    background: var(${CssVars.bgColor2});
`;
var Input = import_styled_components6.default.input`
    background: inherit;
    box-sizing: border-box;
    border: 0;
    appearance: none;
    font-weight: normal;
    font-family: interface, Helvetica Neue, helvetica, sans-serif;
    outline: 0;
    font-size: 15px;
    padding: 0 10px;
    border-radius: 0;
    text-overflow: ellipsis;
    color: var(${CssVars.fgColor});
    &::placeholder {
        color: ${import_colors4.giphyLightGrey};
    }
    min-width: 150px;
    flex: 1;
    ${(props) => props.$isUsernameSearch && import_styled_components6.css`
            color: ${import_colors4.giphyIndigo};
        `}
`;
var SearchBar = ({
  className,
  placeholder = `Search GIPHY`,
  clear = false,
  autoFocus,
  searchDebounce = SEARCH_DEBOUNCE,
  onEnter
}) => {
  const { activeChannel, setActiveChannel, term, setChannels } = (0, import_react8.useContext)(SearchContext);
  const { setIsFocused, _inputValOverride, _setSearch } = (0, import_react8.useContext)(_SearchContext);
  const [val, setVal] = (0, import_react8.useState)(term);
  (0, import_useDebounce.default)(() => _setSearch(val), searchDebounce, [val]);
  const inputRef = (0, import_react8.useRef)(null);
  const previousActiveChannel = usePrevious(activeChannel);
  const [isCleared, setCleared] = (0, import_react8.useState)(clear);
  (0, import_react8.useEffect)(() => {
    var _a;
    if (autoFocus) {
      (_a = inputRef.current) == null ? void 0 : _a.focus();
    }
  }, [autoFocus]);
  (0, import_react8.useEffect)(() => {
    var _a;
    if (activeChannel && !previousActiveChannel) {
      (_a = inputRef.current) == null ? void 0 : _a.focus();
      if (val === " ") {
        setVal("");
      } else {
        setVal(val.replace(/@?\w*\s?/, ""));
      }
    }
  }, [val, activeChannel, previousActiveChannel]);
  (0, import_react8.useEffect)(() => {
    setCleared(clear);
  }, [clear]);
  (0, import_react8.useEffect)(() => {
    setVal(_inputValOverride);
  }, [_inputValOverride, setVal]);
  const onKeyUp = (e) => {
    const key = e.keyCode || e.key;
    switch (key) {
      case 27:
      case `Escape`:
        setActiveChannel(void 0);
        setChannels([]);
        break;
      case 13:
      case `Enter`:
        onEnter == null ? void 0 : onEnter(val);
        break;
      default:
        break;
    }
  };
  const onKeyDown = (e) => {
    const key = e.keyCode || e.key;
    switch (key) {
      case 8:
      case "Backspace":
        if (val === "") {
          setActiveChannel(void 0);
          setChannels([]);
        }
        break;
      default:
        break;
    }
  };
  return /* @__PURE__ */ import_react8.default.createElement(Container3, { className: [SearchBar.className, className].join(" ") }, /* @__PURE__ */ import_react8.default.createElement(search_bar_channel_default, null), /* @__PURE__ */ import_react8.default.createElement(
    Input,
    {
      $isUsernameSearch: term.indexOf("@") === 0,
      onChange: ({ target: { value } }) => {
        if (!isCleared || value !== "") {
          setCleared(false);
          setVal(value);
        }
      },
      onFocus: () => {
        setIsFocused(true);
      },
      onBlur: () => {
        setIsFocused(false);
      },
      value: isCleared ? "" : val,
      placeholder: activeChannel ? `Search ${activeChannel.display_name}` : placeholder,
      autoCapitalize: "off",
      autoCorrect: "off",
      autoComplete: "off",
      ref: inputRef,
      onKeyUp,
      onKeyDown
    }
  ), /* @__PURE__ */ import_react8.default.createElement(cancel_icon_default, { setCleared: () => setVal("") }), /* @__PURE__ */ import_react8.default.createElement(search_button_default, { onClick: () => onEnter == null ? void 0 : onEnter(val) }));
};
SearchBar.className = "giphy-search-bar";
var search_bar_default = SearchBar;
