// @giphy/react-components/dist/components/search-bar/search-button.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/search-bar/search-button.tsx
var search_button_exports = {};
__export(search_button_exports, {
  default: () => search_button_default
});
module.exports = __toCommonJS(search_button_exports);
var import_react3 = __toESM(require("react"));
var import_useThrottle = __toESM(require("react-use/lib/useThrottle"));
var import_styled_components2 = __toESM(require("styled-components"));

// src/components/search-bar/context.tsx
var import_colors = require("@giphy/colors");
var import_js_fetch_api = require("@giphy/js-fetch-api");
var import_js_util = require("@giphy/js-util");
var import_react = __toESM(require("react"));
var import_styled_components = __toESM(require("styled-components"));
var CssVars = {
  searchbarHeight: `--searchbar-height`,
  bgColor: `--searchbar-bg-color`,
  bgColor2: `--searchbar-bg-color-2`,
  fgColor: `--searchbar-fg-color`,
  cancelButtonDisplay: `--searchbar-cancel-button-display`
};
var Container = import_styled_components.default.div`
    ${CssVars.searchbarHeight}: ${(props) => props.$searchbarHeight || 42}px;
    @media (${(props) => props.$mobileMediaQuery}) {
        ${CssVars.searchbarHeight}: ${(props) => props.$mobileSearchbarHeight || 35}px;
    }
    ${CssVars.bgColor}: ${import_colors.giphyWhite};
    ${CssVars.bgColor2}: ${import_colors.giphyWhite};
    ${CssVars.fgColor}: ${import_colors.giphyBlack};
    ${(props) => props.$darkMode && import_styled_components.css`
            ${CssVars.fgColor}: ${import_colors.giphyWhite};
            ${CssVars.bgColor}: ${import_colors.giphyBlack};
            ${CssVars.bgColor2}: ${import_colors.giphyCharcoal};
        `}
    ${CssVars.cancelButtonDisplay}: ${(props) => props.$hideCancelButton ? "none" : "block"};
`;
var SearchContext = (0, import_react.createContext)({});
var _SearchContext = (0, import_react.createContext)({});

// src/components/search-bar/search-icon.tsx
var import_react2 = __toESM(require("react"));
var SearchIcon = ({ className = "" }) => /* @__PURE__ */ import_react2.default.createElement("svg", { viewBox: "0 0 30 30", version: "1.1", className }, /* @__PURE__ */ import_react2.default.createElement("defs", null, /* @__PURE__ */ import_react2.default.createElement(
  "path",
  {
    d: "M11.5482521,20.4090671 L4.24727698,28.2009189 C3.68084207,28.8054377 2.73159653,28.8363108 2.12707771,28.2698759 C1.5225589,27.703441 1.4916858,26.7541954 2.0581207,26.1496766 L9.40599838,18.3077689 C7.95982241,16.4371424 7.0978836,14.0789715 7.0978836,11.5181818 C7.0978836,5.44914339 11.9392549,0.518181818 17.9252787,0.518181818 C23.9113026,0.518181818 28.7526738,5.44914339 28.7526738,11.5181818 C28.7526738,17.5872202 23.9113026,22.5181818 17.9252787,22.5181818 C15.539851,22.5181818 13.3361963,21.7351359 11.5482521,20.4090671 Z M17.9252787,19.5181818 C22.242011,19.5181818 25.7526738,15.9425536 25.7526738,11.5181818 C25.7526738,7.09381 22.242011,3.51818182 17.9252787,3.51818182 C13.6085464,3.51818182 10.0978836,7.09381 10.0978836,11.5181818 C10.0978836,15.9425536 13.6085464,19.5181818 17.9252787,19.5181818 Z",
    id: "giphy-search-icon-path-1"
  }
)), /* @__PURE__ */ import_react2.default.createElement("g", { id: "search", stroke: "none", strokeWidth: "1", fill: "none", fillRule: "evenodd" }, /* @__PURE__ */ import_react2.default.createElement("g", { id: "icons/search" }, /* @__PURE__ */ import_react2.default.createElement("mask", { id: "giphy-search-icon-mask-2", fill: "white" }, /* @__PURE__ */ import_react2.default.createElement("use", { xlinkHref: "#giphy-search-icon-path-1" })), /* @__PURE__ */ import_react2.default.createElement("use", { id: "Mask", fill: "#FFFFFF", fillRule: "nonzero", xlinkHref: "#giphy-search-icon-path-1" }), /* @__PURE__ */ import_react2.default.createElement("g", { mask: "url(#giphy-search-icon-mask-2)" }, /* @__PURE__ */ import_react2.default.createElement("g", { transform: "translate(0.250000, 0.250000)" }, /* @__PURE__ */ import_react2.default.createElement("g", null))))));
var search_icon_default = SearchIcon;

// src/components/search-bar/search-button.tsx
var time = "2s";
var purp = "#E646B6";
var pink = "#FF6666";
var strongEasing = "cubic-bezier(0.920, 0.240, 0.185, 0.730)";
var searchFx = import_styled_components2.keyframes`
    0% {
      transform: rotate(34deg) translate(-10px, 80px);
    };
    
    100% {
      transform: rotate(34deg) translate(-10px, -20px);
    }
`;
var plus = import_styled_components2.keyframes`
    0% { 
      transform: translate(0px, 0px);
      opacity: 0;
    }
    50% {
      opacity: 1;
    }
    100% {
      transform: translate(10px, -17px);
      opacity: 0;
    }
`;
var gradientFade = import_styled_components2.keyframes`
    0% {
      opacity: 0;
      transform: translateX(-400%);
    }
    50% {
      opacity: 1;
    }
    100% {
      opacity: 0;
      transform: translateX(0);
    }
`;
var Container2 = import_styled_components2.default.div`
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;
    cursor: pointer;
    @media screen and (-ms-high-contrast: active), screen and (-ms-high-contrast: none) {
        display: none;
    }
    height: var(${CssVars.searchbarHeight});
    width: var(${CssVars.searchbarHeight});
`;
var GradientBox = import_styled_components2.default.div`
    position: absolute;
    height: 100%;
    width: 100%;
    background: linear-gradient(45deg, ${purp} 0%, ${pink} 100%);
    border-radius: 0 4px 4px 0;
    overflow: hidden;
    &:before {
        animation: ${gradientFade} ${time} linear 0s infinite;
        background-image: linear-gradient(45deg, ${purp} 0%, ${pink} 50%, ${purp} 100%);
        background-size: 400%;
        background-position: 0% 100%;
        content: '';
        position: absolute;
        left: 0;
        top: 0;
        right: 0;
        bottom: 0;
        width: 400%;
    }
`;
var Fx = import_styled_components2.default.div`
    width: 100%;
    height: 100%;
    position: absolute;
    overflow: hidden;
    &::after {
        content: '+';
        color: white;
        font-family: 'SS Standard';
        font-size: 8px;
        position: absolute;
        top: 65%;
        left: 66%;
        animation: ${plus} 1s ${strongEasing} 0s 1 forwards;
    }
`;
var Scanner = import_styled_components2.default.div`
    position: absolute;
    width: 200%;
    height: 20px;
    background: rgba(255, 255, 255, 0.5);
    transform: rotate(34deg) translate(-10px, -20px);
    animation: ${searchFx} 1s ${strongEasing} 0s 1;
    filter: blur(1px);
`;
var SearchIcon2 = (0, import_styled_components2.default)(search_icon_default)`
    z-index: 1;
    display: flex;
    width: 50%;
    height: 50%;
`;
var SearchButton = ({ onClick }) => {
  const { isFetching } = (0, import_react3.useContext)(SearchContext);
  const throttledFetch = (0, import_useThrottle.default)(isFetching, 1e3);
  return /* @__PURE__ */ import_react3.default.createElement(Container2, { onClick: () => onClick == null ? void 0 : onClick() }, /* @__PURE__ */ import_react3.default.createElement(GradientBox, { suppressHydrationWarning: true }), /* @__PURE__ */ import_react3.default.createElement(SearchIcon2, null), throttledFetch && /* @__PURE__ */ import_react3.default.createElement(Fx, { suppressHydrationWarning: true }, /* @__PURE__ */ import_react3.default.createElement(Scanner, { suppressHydrationWarning: true })));
};
var search_button_default = SearchButton;
