// @replyke/comments-threaded-react-js/dist/esm/components/CommentsFeed/CommentThread/SingleComment/IndentationThreadingLines.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useThreadedStyleConfig } from "@replyke/comments-threaded-core";
function IndentationThreadingLines({ isLastReply }) {
    const { styleConfig } = useThreadedStyleConfig();
    const { threadingLineColor } = styleConfig.commentProps;
    return (_jsxs(_Fragment, { children: [!isLastReply && (_jsx("div", { style: {
                    position: "absolute",
                    width: "1px",
                    backgroundColor: threadingLineColor,
                    zIndex: 0,
                    left: "-12px",
                    top: "0px",
                    height: "100%",
                } })), _jsx("div", { style: {
                    position: "absolute",
                    top: "-8px",
                    width: "1px",
                    height: "18px",
                    backgroundColor: threadingLineColor,
                    zIndex: 0,
                    left: "-12px",
                } }), _jsx("div", { style: {
                    position: "absolute",
                    top: "10px",
                    width: "12px",
                    height: "16px",
                    borderLeft: `1.5px solid ${threadingLineColor}`,
                    borderBottom: `1.5px solid ${threadingLineColor}`,
                    borderTop: "1.5px solid transparent",
                    borderRight: "1.5px solid transparent",
                    borderBottomLeftRadius: "12px",
                    zIndex: 0,
                    left: "-12px",
                } })] }));
}
export default IndentationThreadingLines;
