// @replyke/comments-threaded-react-js/dist/esm/components/CommentsFeed/CommentThread/SingleComment/ReplyButtonAndInfo.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useThreadedStyleConfig } from "@replyke/comments-threaded-core";
function ReplyButtonAndInfo({ hasReplies, replyCount, setShowReplyForm, }) {
    const { styleConfig } = useThreadedStyleConfig();
    const { replyButtonFontSize, replyButtonFontWeight, replyButtonFontColor, actionsItemGap, fromNowFontColor, } = styleConfig.commentProps;
    return (_jsxs("div", { style: {
            display: "flex",
            alignItems: "center",
            gap: `${actionsItemGap}px`,
            fontSize: `${replyButtonFontSize}px`,
        }, children: [_jsx("button", { onClick: () => setShowReplyForm((prev) => !prev), style: {
                    color: replyButtonFontColor,
                    fontWeight: replyButtonFontWeight,
                    padding: "4px 8px",
                    borderRadius: "4px",
                    marginLeft: "-8px",
                    transition: "all 150ms ease-in-out",
                    background: "transparent",
                    border: "none",
                    cursor: "pointer",
                }, onMouseEnter: (e) => {
                    e.currentTarget.style.color = "#2563EB";
                    e.currentTarget.style.backgroundColor = "#EFF6FF";
                }, onMouseLeave: (e) => {
                    e.currentTarget.style.color = "#6B7280";
                    e.currentTarget.style.backgroundColor = "transparent";
                }, children: "Reply" }), hasReplies && (_jsxs("span", { style: { color: fromNowFontColor }, children: [replyCount, " ", replyCount === 1 ? "reply" : "replies"] }))] }));
}
export default ReplyButtonAndInfo;
