// @replyke/comments-threaded-react-js/dist/esm/components/CommentsFeed/CommentThread/SingleComment/SingleComment.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useState } from "react";
import { getUserName, useCommentSection, useUser, } from "@replyke/react-js";
import { parseContentWithMentions, UserAvatar, } from "@replyke/ui-core-react-js";
import { useThreadedStyleConfig } from "@replyke/comments-threaded-core";
import VoteButtons from "./VoteButtons";
import ActionMenu from "../ActionMenu";
import NewReplyForm from "../NewReplyForm";
import ToggleRepliesVisibilty from "./ToggleRepliesVisibilty";
import IndentationThreadingLines from "./IndentationThreadingLines";
import ReplyButtonAndInfo from "./ReplyButtonAndInfo";
function SingleComment({ comment: commentFromSection, depth, hasReplies, isCollapsed, replyCount, isLastReply = false, onToggleCollapse, }) {
    const { user } = useUser();
    const { callbacks, highlightedComment } = useCommentSection();
    const [comment, setComment] = useState(commentFromSection);
    const [showReplyForm, setShowReplyForm] = useState(false);
    const { styleConfig } = useThreadedStyleConfig();
    const { authorAvatarSize, authorFontSize, authorFontWeight, authorFontColor, fromNowFontSize, fromNowFontColor, commentBodyFontSize, commentBodyFontColor, horizontalItemsGap, verticalItemsGap, threadingLineColor, } = styleConfig.commentProps;
    const maxDepth = 6; // Limit visual nesting depth
    const actualDepth = Math.min(depth, maxDepth);
    // Calculate progressive indentation using inline styles for reliability
    const indentationPx = actualDepth * 24; // 24px per level (same as ml-10 = 2.5rem = 40px)
    return (_jsxs("div", { style: {
            position: "relative",
            // marginBottom: "8px",
            marginLeft: `${indentationPx}px`,
            backgroundColor: highlightedComment?.comment.id === comment.id
                ? "#dbeafe"
                : "transparent",
        }, children: [actualDepth > 0 && (_jsx(IndentationThreadingLines, { isLastReply: isLastReply })), _jsx("div", { style: {
                    padding: `${verticalItemsGap}px 0`,
                    borderRadius: "6px",
                    transition: "colors 150ms ease-in-out",
                }, children: _jsxs("div", { style: { display: "flex" }, children: [_jsxs("div", { style: {
                                flexShrink: 0,
                                marginRight: `${horizontalItemsGap}px`,
                                position: "relative",
                                marginTop: `${verticalItemsGap / 2}px`,
                            }, children: [_jsx("div", { style: { position: "relative", zIndex: 10 }, children: _jsx(UserAvatar, { user: comment.user, borderRadius: authorAvatarSize, size: authorAvatarSize }) }), hasReplies && !isCollapsed && (_jsx("div", { style: {
                                        position: "absolute",
                                        left: "50%",
                                        top: "20px",
                                        width: "1px",
                                        backgroundColor: threadingLineColor,
                                        zIndex: 0,
                                        height: "calc(100% + 10px)",
                                    } }))] }), _jsxs("div", { style: { flex: 1, minWidth: 0 }, children: [_jsxs("div", { style: {
                                        display: "flex",
                                        alignItems: "center",
                                        justifyContent: "space-between",
                                        marginBottom: `${verticalItemsGap / 2}px`,
                                    }, children: [_jsxs("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                gap: `${verticalItemsGap}px`,
                                                fontSize: `${fromNowFontSize}px`,
                                                color: fromNowFontColor,
                                            }, children: [_jsx("span", { style: {
                                                        fontWeight: authorFontWeight,
                                                        fontSize: authorFontSize,
                                                        color: authorFontColor,
                                                    }, children: getUserName(comment.user) }), _jsx("span", { children: "\u2022" }), _jsx("span", { children: new Date(comment.createdAt).toLocaleDateString() }), isCollapsed && hasReplies && (_jsxs("span", { className: "text-blue-600 text-xs", children: ["(", replyCount, " ", replyCount === 1 ? "reply" : "replies", ")"] })), hasReplies && (_jsx(ToggleRepliesVisibilty, { isCollapsed: isCollapsed, onToggleCollapse: onToggleCollapse }))] }), _jsx(ActionMenu, { comment: comment })] }), !isCollapsed && (_jsxs(_Fragment, { children: [comment.content && (_jsx("p", { style: {
                                                fontSize: `${commentBodyFontSize}px`,
                                                color: commentBodyFontColor,
                                                marginBottom: `${horizontalItemsGap}px`,
                                                lineHeight: "1.625",
                                            }, children: parseContentWithMentions(comment.content, comment.mentions, user?.id, callbacks?.currentUserClickCallback, callbacks?.otherUserClickCallback) })), comment.gif && (_jsx("img", { src: comment.gif.gifUrl, alt: comment.gif.altText, style: {
                                                width: comment.gif.aspectRatio > 1
                                                    ? 200
                                                    : 200 * comment.gif.aspectRatio,
                                                height: comment.gif.aspectRatio < 1
                                                    ? 200
                                                    : 200 / comment.gif.aspectRatio,
                                                borderRadius: "0.25rem",
                                                overflow: "hidden",
                                                objectFit: "cover",
                                                marginBottom: `${horizontalItemsGap}px`,
                                            } })), _jsxs("div", { style: {
                                                display: "flex",
                                                alignItems: "center",
                                                justifyContent: "space-between",
                                            }, children: [_jsx(ReplyButtonAndInfo, { hasReplies: hasReplies, replyCount: replyCount, setShowReplyForm: setShowReplyForm }), _jsx("div", { style: { flexShrink: 0 }, children: _jsx(VoteButtons, { comment: comment, setComment: setComment, size: "small" }) })] }), showReplyForm && (_jsx(NewReplyForm, { comment: comment, setShowReplyForm: setShowReplyForm }))] }))] })] }) })] }));
}
export default SingleComment;
